---
title: "SPEAQeasy: a Scalable Pipeline for Expression Analysis and Quantification, a demonstration of its usage on real data"
author: 
    
  - name: Daianna Gonzalez-Padilla
    affiliation:
    - &libd Lieber Institute for Brain Development
    email: glezdaianna@gmail.com
  - name: Renee Garcia-Flores
    affiliation:
    - &libd Lieber Institute for Brain Development
    email: renee.garciaflores@gmail.com
  - name: Nicholas J. Eagles
    affiliation:
    - &libd Lieber Institute for Brain Development, Johns Hopkins Medical Campus
    email: nick.eagles@libd.org
  - name: Leonardo Collado-Torres
    affiliation:
    - &libd Lieber Institute for Brain Development
    - &ccb Center for Computational Biology, Johns Hopkins University
    email: lcolladotor@gmail.com
output: 
  BiocStyle::html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
package: "`r pkg_ver('SPEAQeasyWorkshop2023')`"
vignette: >
  %\VignetteEncoding{UTF-8}  
  %\VignetteIndexEntry{SPEAQeasy: a Scalable Pipeline for Expression Analysis and Quantification for R/Bioconductor-powered RNA-seq analyses}
  %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  markdown: 
    wrap: 72
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r vignetteSetup, echo=FALSE, message=FALSE, warning = FALSE}
## Track time spent on making the vignette
startTime <- Sys.time()

## Bib setup
library(RefManageR)

## Write bibliography information
bib <- c(
    R = citation(),
    BiocStyle = citation("BiocStyle")[1],
    dplyr = citation("dplyr")[1],
    knitr = citation("knitr")[1],
    rmarkdown = citation("rmarkdown")[1],
    sessioninfo = citation("sessioninfo")[1],
    SPEAQeasyWorkshop2023 = citation("SPEAQeasyWorkshop2023")[1],
    SPEAQeasy = citation("SPEAQeasyWorkshop2023")[2],
    smokingMouse = citation("smokingMouse")[1],
    clusterProfiler = citation("clusterProfiler"),
    edgeR = citation("edgeR"),
    here = citation("here"),
    jaffelab = citation("jaffelab"),
    limma = citation("limma"),
    org.Hs.eg.db = citation("org.Hs.eg.db"),
    pheatmap = citation("pheatmap"),
    rmarkdown = citation("rmarkdown"),
    recount = citation("recount"),
    RefManageR = citation("RefManageR")[1],
    SummarizedExperiment = citation("SummarizedExperiment"),
    tidyr = citation("tidyr")[1],
    VariantAnnotation = citation("VariantAnnotation")[1],
    voom = RefManageR::BibEntry(
        "article",
        key = "voom",
        author = "CW Law and Y Chen and W Shi and GK Smyth",
        year = "2014",
        title = "Voom: precision weights unlock linear model analysis tools for RNA-seq read counts",
        journal = "Genome Biology",
        volume = "15",
        pages = "R29"
    )
)
```

------------------------------------------------------------------------

# Overview

This workshop aims to present the `SPEAQeasy`
`r Citep(bib[["SPEAQeasy"]])` RNA-seq processing pipeline, show how to
use it and demonstrate how its outputs can be analyzed for differential
expression analyses (DEA) using other Bioconductor R packages.

SPEAQeasy is a [Nextflow](https://www.nextflow.io/)-based **S**calable
RNA-seq processing **P**ipeline for **E**xpression **A**nalysis and
**Q**uantification that produces R objects ready for analysis with
Bioconductor tools. Partipants will become familiar with execution on
real data from
[smokingMouse](https://github.com/LieberInstitute/smokingMouse), which
contains gene expression data and sample information from an
RNA-sequencing. As well as practice configuring some common settings. We
will walk through a complete differential expression analysis, utilizing
popular packages such as
[limma](https://www.bioconductor.org/packages/limma),
[edgeR](http://bioconductor.org/packages/edgeR), and
[clusterProfiler](https://bioconductor.org/packages/clusterProfiler).

## Instructors

-   [Daianna Gonzalez-Padilla](https://daianna21.github.io)
-   [Renee Garcia-Flores](https://reneegf.github.io)
-   [Nick Eagles](https://nick-eagles.github.io)

## Pre-requisites

-   General familiarity with bulk RNA-seq data.
-   Basic knowledge of differential expression.
-   Basic experience handling `SummarizedExperiment` and `GenomicRanges`
    packages.
-   Previous review of <http://research.libd.org/smokingMouse/> for data
    explanation. (Optional)

## Participation

**TODO** Participants are expected to follow

## *R* / *Bioconductor* packages used

**TODO** List any *R* / *Bioconductor* packages that will be explicitly
covered.

## Time outline

**Check this with Leo**

| Activity                         | Time |
|----------------------------------|------|
| Introduction to SPEAQeasy        | 25m  |
| Understanding SPEAQeasy outputs  | 25m  |
| Differential Expression Analysis | 40m  |

Total: a 90 minute session.

## Workshop goals and objectives

**TODO** List "big picture" student-centered workshop goals and learning
objectives. Learning goals and objectives are related, but not the same
thing. These goals and objectives will help some people to decide
whether to attend the conference for training purposes, so please make
these as precise and accurate as possible.

*Learning goals* are high-level descriptions of what participants will
learn and be able to do after the workshop is over. *Learning
objectives*, on the other hand, describe in very specific and measurable
terms specific skills or knowledge attained. The [Bloom's
Taxonomy](#bloom) may be a useful framework for defining and describing
your goals and objectives, although there are others.

### Learning goals

Some examples:

-   Understand what SPEAQeasy is and how to include in a RNA-seq
    processing workflow
-   Become familiar with SPEAQeasy outputs
-   Perform a differential expression anlysis with real data after
    analyzed with SPEAQeasy

### Learning objectives

-   analyze xyz data to produce...
-   create xyz plots
-   evaluate xyz data for artifacts

## Citing `SPEAQeasy`

We hope that `SPEAQeasy` `r Citep(bib[["SPEAQeasy"]])` will be useful
for your research. Please use the following information to cite the
package and the overall approach. Thank you!

```{r "citation"}
## Citation info
citation("SPEAQeasyWorkshop2020")[2]
```

# Workshop

**TODO** Divide the workshop into sections (`## A Section`). Include
fully-evaluated *R* code chunks. Develop exercises and solutions, and
anticipate that your audience will walk through the code with you, or
work on the code idependently -- do not be too ambitious in the material
that you present.

## Introduction

### SPEAQeasy Overview

We introduce [SPEAQeasy](https://github.com/LieberInstitute/SPEAQeasy),
a **S**calable RNA-seq processing **P**ipeline for **E**xpression
**A**nalysis and **Q**uantification, that is **easy** to install, use,
and share with others. More detailed documentation is
[here](http://research.libd.org/SPEAQeasy/).

![Functional workflow. Yellow coloring denotes steps which are optional
or not always performed.](figs/SPEAQeasy_workflow.png)

SPEAQeasy starts with raw sequencing outputs in FASTQ format and
ultimately produces
[RangedSummarizedExperiment](https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html)
objects with raw and normalized counts for each feature type: genes,
exons, exon-exon junctions, and transcripts.

For human samples, variant calling is performed at a list of highly
variables sites. A single [VCF
file](https://www.internationalgenome.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-40/)
is generated, containing genotype calls for all samples. This
establishes a sufficiently unique profile for each sample which can be
compared against pre-sequencing genotype calls to resolve potential
identity issues.

Optionally,
[BigWig](https://genome.ucsc.edu/goldenPath/help/bigWig.html) coverage
files by strand, and averaged across strands, are produced. An
additional step utilizes the
[derfinder](https://bioconductor.org/packages/release/bioc/html/derfinder.html)
package to compute expressed regions, a precursor to analyses finding
differentially expressed regions (DERs).

### Installation and Set-up

First, clone the SPEAQeasy repository.

```{bash, eval=FALSE}
git clone git@github.com:LieberInstitute/SPEAQeasy.git
```

SPEAQeasy dependencies are recommended to be installed using docker or
singularity, involving a fairly quick installation that is reproducible
regardless of the computing environment. We'll install using
singularity, which is more frequently available in high-performance
computing (HPC) environments:

```{bash, eval=FALSE}
#  From within the SPEAQeasy repository:
bash install_software.sh "singularity"
```

If neither docker nor singularity are available, a more experimental
local installation of all dependencies is an option:

```{bash, eval=FALSE}
#  From within the SPEAQeasy repository:
bash install_software.sh "local"
```

Please note that installation may take some time-- **installing and
running SPEAQeasy is not required for this workshop**, but is
demonstrated to familiarize participants with the process. The workshop
files include outputs from SPEAQeasy, which can be used later for the
example differential expression analysis.

Choose a "main" script as appropriate for your particular set-up. "Main"
scripts and associated configuration files exist for SLURM-managed
computing clusters, SGE-managed clusters, local machines, and the
[JHPCE](https://jhpce.jhu.edu/) cluster.

| Environment   | "Main" script         | Config file                                   |
|------------------|----------------------|--------------------------------|
| SLURM cluster | run_pipeline_slurm.sh | conf/slurm.config or conf/docker_slurm.config |
| SGE cluster   | run_pipeline_sge.sh   | conf/sge.config or conf/docker_sge.config     |
| Local machine | run_pipeline_local.sh | conf/local.config or conf/docker_local.config |
| JHPCE cluster | run_pipeline_jhpce.sh | conf/jhpce.config                             |

Within the main script, you can configure arguments specific to the
experiment, such as the reference organism, pairing of samples, and
where to place output files, among other specifications.

When running SPEAQeasy on a cluster (i.e. SLURM, SGE, or JHPCE users),
it is recommended you submit the pipeline as a job, using the command
appropriate for your cluster. For those running SPEAQeasy locally, the
main script can be executed directly, ideally as a background task.

```{bash, eval=FALSE}
#  SLURM-managed clusters
sbatch run_pipeline_slurm.sh

#  SGE-managed clusters
qsub run_pipeline_sge.sh

#  local machines
bash run_pipeline_local.sh

#  The JHPCE cluster
qsub run_pipeline_jhpce.sh
```

Here's an example of such a file (TODO!)

```{r "example_run_file", eval = FALSE}
## Root for our example files
example_files <- system.file(
    "extdata",
    "SPEAQeasy-example",
    package = "SPEAQeasyWorkshop2020"
)

## Read in the actual bash script used to run SPEAQeasy with the
## data from SPEAQeasy-example
cat(paste0(readLines(
    file.path(example_files, "pipeline_outputs", "run_pipeline_jhpce.sh")
), "\n"))
```

## Data overview and download

The dataset that will be used for the DEA can be downloaded from the
`smokingMouse` package. Visit
[here](http://research.libd.org/smokingMouse/) for more details of this
package.

<img src="figs/Data_overview_experiments.jpg" width="750px"/>

<figcaption style="color: gray; line-height: 0.85; text-align: justify">

<font size="-2"><b>Figure 1</b>: <b>Experimental design of the study.
</b> The data was obtained from an experiment in which <b>A)</b> 6
pregnant dams and 17 non-pregnant female adult mice were administered
nicotine by intraperitoneal injection (IP; n=12) or controls (n=11). A
total of 42 pups were born to pregnant dams: 19 were born to mice that
were administered nicotine and 23 to control mice. Samples from frontal
cortices of P0 pups and adults were obtained. <b>B)</b> RNA was
extracted from those samples and then RNA-seq libraries were prepared
and sequenced to obtain the expression counts of the genes. At the end,
a dataset of 65 samples (either from adult or pup brain) and 55,401
genes was generated. </font>

</figcaption>

The data reside in a `RangedSummarizedExperiment` (RSE) object,
containing the following assays:

-   `counts`: original read counts of the 55,401 mouse genes across 208
    samples (inlcuding the 65 nicotine samples).
-   `logcounts`: normalized and scaled counts ($log_2(CPM + 0.5)$) of
    the same genes across the same samples; normalization was carried
    out applying TMM method with `cpm(calcNormFactors())` of
    `r BiocStyle::Biocpkg("edgeR")`.

The same RSE contains the sample information in `colData(RSE)` and the
gene information in `rowData(RSE)`.
<mark style= "background-color: #FCF3CF"> Yellow </mark> variables
correspond to SPEAQeasy outputs that are going to be used in downstream
analyses.

<mark style= "background-color: #FAECF8"> Pink </mark> variables are
specific of the study, such as sample metadata and some others
containing additional information of the genes.

<mark style= "background-color: #DFF0FE"> Blue </mark> variables are
Quality-Control metrics computed by `addPerCellQC()` of
`r BiocStyle::Biocpkg("scuttle")`.

#### Gene Information

Among the information in `rowData(RSE)` the next variables are of
interest for the analysis:

-   <mark style= "background-color: #FCF3CF"> [
    gencodeID]{style="font-family: monospace"} </mark>: GENCODE ID of
    each gene.
-   <mark style= "background-color: #FCF3CF"> [
    ensemblID]{style="font-family: monospace"} </mark>: gene ID in
    Ensembl database.
-   <mark style= "background-color: #FCF3CF"> [
    EntrezID]{style="font-family: monospace"} </mark>: identifier of
    each gene in NCBI Entrez database.
-   <mark style= "background-color: #FCF3CF"> [
    Symbol]{style="font-family: monospace"} </mark>: official gene
    symbol for each mouse gene.
-   <mark style= "background-color: #FAECF8"> [
    retained_after_feature_filtering]{style="font-family: monospace"}
    </mark>: Boolean variable that equals TRUE if the gene passed the
    gene filtering (with `filterByExpr()` of
    `r BiocStyle::Biocpkg("edgeR")`) based on its expression levels and
    FALSE if not.

#### Sample Information

Sample information in `colData(RSE)` contains the following variables:

-   <mark style= "background-color: #FCF3CF"> [
    SAMPLE_ID]{style="font-family: monospace"} </mark>: is the name of
    the sample.
-   <mark style= "background-color: #FCF3CF"> [
    ERCCsumLogErr]{style="font-family: monospace"} </mark>: a summary
    statistic quantifying overall difference of expected and actual ERCC
    concentrations for one sample.
-   <mark style= "background-color: #FCF3CF"> [
    overallMapRate]{style="font-family: monospace"} </mark>: the decimal
    fraction of reads which successfully mapped to the reference genome
    (i.e. *numMapped* / *numReads*).
-   <mark style= "background-color: #FCF3CF"> [
    mitoMapped]{style="font-family: monospace"} </mark>: the number of
    reads which successfully mapped to the mitochondrial chromosome.
-   <mark style= "background-color: #FCF3CF"> [
    totalMapped]{style="font-family: monospace"} </mark>: the number of
    reads which successfully mapped to the canonical sequences in the
    reference genome (excluding mitochondrial chromosomes).
-   <mark style= "background-color: #FCF3CF"> [
    mitoRate]{style="font-family: monospace"} </mark>: the decimal
    fraction of reads which mapped to the mitochondrial chromosome, of
    those which map at all (i.e. *mitoMapped* / (*totalMapped* +
    *mitoMapped*))
-   <mark style= "background-color: #FCF3CF"> [
    totalAssignedGene]{style="font-family: monospace"} </mark>: the
    decimal fraction of reads assigned unambiguously to a gene, with
    `featureCounts` (Liao et al. 2014), of those in total.
-   <mark style= "background-color: #FCF3CF"> [
    rRNA_rate]{style="font-family: monospace"} </mark>: the decimal
    fraction of reads assigned to a gene whose type is 'rRNA', of those
    assigned to any gene.
-   <mark style= "background-color: #FAECF8"> [
    Tissue]{style="font-family: monospace"} </mark>: tissue (mouse brain
    or blood) from which the sample comes.
-   <mark style= "background-color: #FAECF8"> [
    Age]{style="font-family: monospace"} </mark>: if the sample comes
    from an adult or a pup mouse.
-   <mark style= "background-color: #FAECF8"> [
    Sex]{style="font-family: monospace"} </mark>: if the sample comes
    from a female (F) or male (M) mouse.
-   <mark style= "background-color: #FAECF8"> [
    Expt]{style="font-family: monospace"} </mark>: the experiment
    (nicotine or smoking exposure) to which the sample mouse was
    subjected; it could be an exposed or a control mouse of that
    experiment.
-   <mark style= "background-color: #FAECF8"> [
    Group]{style="font-family: monospace"} </mark>: if the sample
    belongs to a nicotine/smoking-exposed mouse (Expt) or a
    nicotine/smoking control mouse (Ctrl).
-   <mark style= "background-color: #FAECF8"> [
    plate]{style="font-family: monospace"} </mark>: is the plate (1,2 or
    3) in which the sample library was prepared.
-   <mark style= "background-color: #FAECF8"> [
    Pregnancy]{style="font-family: monospace"} </mark>: if the sample
    comes from a pregnant (Yes) or not pregnant (No) mouse.
-   <mark style= "background-color: #FAECF8"> [
    medium]{style="font-family: monospace"} </mark>: is the medium in
    which the sample was treated: water for brain samples and an elution
    buffer (EB) for the blood ones.
-   <mark style= "background-color: #FAECF8"> [
    flowcell]{style="font-family: monospace"} </mark>: is the sequencing
    batch of each sample.
-   <mark style= "background-color: #DFF0FE"> [
    sum]{style="font-family: monospace"} </mark>: library size (total
    sum of counts across all genes for each sample).
-   <mark style= "background-color: #DFF0FE"> [
    detected]{style="font-family: monospace"} </mark>: number of
    non-zero expressed genes in each sample.
-   <mark style= "background-color: #DFF0FE"> [
    subsets_Mito_sum]{style="font-family: monospace"} </mark>: total sum
    of read counts of mt genes in each sample
-   <mark style= "background-color: #DFF0FE"> [
    subsets_Mito_detected]{style="font-family: monospace"} </mark>:
    total number of mt genes in each sample
-   <mark style= "background-color: #DFF0FE"> [
    subsets_Mito_percent]{style="font-family: monospace"} </mark>: % of
    mt genes' counts of the total counts of the sample.
-   <mark style= "background-color: #DFF0FE"> [
    subsets_Ribo_sum]{style="font-family: monospace"} </mark>: total sum
    of read counts of ribosomal genes in each sample.
-   <mark style= "background-color: #DFF0FE"> [
    subsets_Ribo_detected]{style="font-family: monospace"} </mark>:
    total number of ribosomal genes in each sample.
-   <mark style= "background-color: #DFF0FE"> [
    subsets_Ribo_percent]{style="font-family: monospace"} </mark>: % of
    ribosomal genes' counts of the total counts of the sample.

Note: in our case, we'll use samples from the nicotine experiment only,
so all samples come from brain and were treated in water.

```{r "load_pkgs", message=FALSE, warning=FALSE}
library("here")
library("SummarizedExperiment")
library("jaffelab")
```

```{r "examine_outputs"}

#################################
##         Retrive data
#################################
## Provisional data source until smokingMouse pkg is accepted
load(here('provisional_data/rse_gene_mouse_RNAseq_nic-smo.Rdata'), verbose=TRUE)
## Explore complete rse_gene object (including smoking-exposed samples as well)
rse_gene

#################################
##   Extract data of interest
#################################
## Keep samples from nicotine experiment only
rse_gene_nic <- rse_gene[,which(rse_gene$Expt=='Nicotine')]
## New dimensions
dim(rse_gene_nic)

#################################
##  Access gene expression data
#################################

## Raw counts for first 3 genes in the first 5 samples
assays(rse_gene_nic)$counts[1:3,1:5]
## Log-normalized counts for first 3 genes in the first 5 samples
assays(rse_gene_nic)$logcounts[1:3,1:5]

#################################
##     Access sample data
#################################

#  Check out the metrics SPEAQeasy collected for each sample, including stats
#  from FastQC, trimming, and alignment. View the corner
colnames(colData(rse_gene_nic))
corner(colData(rse_gene_nic))

#  Examine what data is provided in the RSE object for each gene ("row")
head(rowData(rse_gene))
```

## Differential Expression Analysis

**TODO**

<figure>

<img src="figs/Analyses.jpg" width="700px" align="center"/>

<figcaption style="color: gray; line-height: 0.87; text-align: justify">

<font size="-1.8"> <b>Figure 2</b>: <b> Summary of the analyses. </b>
Steps in yellow will be properly performed. <b> 1. Data Preparation </b>
<b> 2. Exploratory Data Analysis </b> <b> 3. Differential Expression
Analysis </b> <b> 4. GO & KEGG Analyses </b> <b> 5. DEG visualization
</b> </font> <font size="0.8"> Abbreviations: CPM: counts per million;
QC: quality control; PC: principal component; DEG: differentially
expressed genes; GO: Gene Ontology; KEGG: Kyoto Encyclopedia of Genes
and Genomes.</font>

</figcaption>

</figure>

### 1. Data preparation

### 2. Exploratory Data Analysis

#### 2.1 Quality Control Analysis

#### 2.2 Explore sample-level effects

#### 2.2 Explore gene-level effects

### 3. Differential Expression Analysis

#### 3.1 Modeling

#### 3.2 Comparisons

### 4. GO & KEGG analyses

### 5. DEG visualization
