---
title: "SPEAQeasy: a Scalable Pipeline for Expression Analysis and Quantification, a demonstration of its usage on real data"
author: 
  - name: Daianna Gonzalez-Padilla
    affiliation:
    - &libd Lieber Institute for Brain Development
    email: glezdaianna@gmail.com
  - name: Renee Garcia-Flores
    affiliation:
    - *libd
    email:  renee.garciaflores@gmail.com
  - name: Nicholas J. Eagles
    affiliation:
    - *libd
    email: nick.eagles@libd.org
  - name: Leonardo Collado-Torres
    affiliation:
    - *libd
    - &ccb Center for Computational Biology, Johns Hopkins University
    email: lcolladotor@gmail.com
output: 
  BiocStyle::html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
package: "`r pkg_ver('SPEAQeasyWorkshop2023')`"
vignette: >
  %\VignetteIndexEntry{An Example Workshop}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

```{r vignetteSetup, echo=FALSE, message=FALSE, warning = FALSE}
## For links
library(BiocStyle)

## Track time spent on making the vignette
startTime <- Sys.time()

## Bib setup
library(RefManageR)

## Write bibliography information
bib <- c(
    BiocStyle = citation("BiocStyle")[1],
    clusterProfiler = citation("clusterProfiler")[1],
    cowplot = citation("cowplot")[1],
    dplyr = citation("dplyr")[1],
    edgeR = citation("edgeR")[1],
    ggplot2 = citation("ggplot2")[1],
    ggrepel = citation("ggrepel")[1],
    ggpubr = citation("ggpubr")[1],
    here = citation("here")[1],
    Hmisc = citation("Hmisc")[1],
    jaffelab = citation("jaffelab")[1],
    knitr = citation("knitr")[1],
    limma = citation("limma")[1],
    org.Hs.eg.db = citation("org.Hs.eg.db")[1],
    pheatmap = citation("pheatmap")[1],
    RefManageR = citation("RefManageR")[1],
    rlang = citation("rlang")[1],
    rmarkdown = citation("rmarkdown")[1],
    rmarkdown = citation("rmarkdown")[1],
    scater = citation("scater")[1],
    sessioninfo = citation("sessioninfo")[1],
    smokingMouse = citation("smokingMouse")[1],
    SPEAQeasy = citation("SPEAQeasyWorkshop2023")[2],
    stringr = citation("stringr")[1],
    SummarizedExperiment = citation("SummarizedExperiment")[1],
    voom = RefManageR::BibEntry(
        "article",
        key = "voom",
        author = "CW Law and Y Chen and W Shi and GK Smyth",
        year = "2014",
        title = "Voom: precision weights unlock linear model analysis tools for RNA-seq read counts",
        journal = "Genome Biology",
        volume = "15",
        pages = "R29"
    )
)
```

# Overview

**TODO**
This workshop aims to present the `SPEAQeasy` `r Citep(bib[["SPEAQeasy"]])` RNA-seq processing pipeline, demonstrating how its outputs can be used in Differential Expression Analyses (DEA) and the type of results it enables to obtain using other Bioconductor R packages such as `r Biocpkg('limma')`, `r Biocpkg('edgeR')`, and `r Biocpkg('clusterProfiler')`.

Briefly describe what is it , etc ... 


## Description

**TODO**
In this demo, participants will be able to understand what the SPEAQeasy pipeline does 
and what it returns through a differential expression analysis performed on a 
dataset from the `smokingMouse` `r Citep(bib[["smokingMouse"]])` Bioconductor package, 
which contains gene expression data and sample information from an RNA-sequencing 
experiment in mice. 


## Instructors

* [Daianna Gonzalez](https://daianna21.github.io)
* [Renee Garcia](https://reneegf.github.io)
* [Nick Eagles](https://nick-eagles.github.io)


## Pre-requisites

* General familiarity with bulk RNA-seq data.
* Basic knowledge of differential expression. 
* Basic experience handling `r Biocpkg('SummarizedExperiment')` and `r Biocpkg('GenomicRanges')` packages.
* Previous review of http://research.libd.org/smokingMouse/ for data explanation. (Optional)

## Participation

**TODO**
Participants are expected to follow 


## _R_ / _Bioconductor_ packages used

**TODO**
List any _R_ / _Bioconductor_ packages that will be explicitly covered.

## Time outline

**TODO**

| Activity                     | Time |
|------------------------------|------|
| Introduction to SPEAQeasy               | 15m |
| Data overview                           | 20m |
| Differential Expression Analysis        | 55m |

Total: a 90 minute session.


## Workshop goals and objectives

**TODO**
List "big picture" student-centered workshop goals and learning
objectives. Learning goals and objectives are related, but not the
same thing. These goals and objectives will help some people to decide
whether to attend the conference for training purposes, so please make
these as precise and accurate as possible.

*Learning goals* 
are high-level descriptions of what
participants will learn and be able to do after the workshop is
over. 
*Learning objectives*, on the other hand, describe in very
specific and measurable terms specific skills or knowledge
attained. The [Bloom's Taxonomy](#bloom) may be a useful framework
for defining and describing your goals and objectives, although there
are others.

### Learning goals

Some examples:

* describe how to...
* identify methods for...
* understand the difference between...

### Learning objectives

* analyze xyz data to produce...
* create xyz plots
* evaluate xyz data for artifacts


# Workshop

**TODO**
Divide the workshop into sections (`## A Section`). Include
fully-evaluated _R_ code chunks. Develop exercises and solutions, and
anticipate that your audience will walk through the code with you, or
work on the code independently -- do not be too ambitious in the
material that you present.

## Introduction 

**TODO**

### SPEAQeasy overview

We introduce [SPEAQeasy](https://github.com/LieberInstitute/SPEAQeasy), a **S**calable RNA-seq processing **P**ipeline for **E**xpression **A**nalysis and **Q**uantification, that is **easy** to install, use, and share with others. More detailed documentation is [here](http://research.libd.org/SPEAQeasy/).

![Functional workflow. Yellow coloring denotes steps which are optional or not always performed.](figs/SPEAQeasy_workflow.png)

SPEAQeasy starts with raw sequencing outputs in FASTQ format and ultimately produces [RangedSummarizedExperiment](https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html) objects with raw and normalized counts for each feature type: genes, exons, exon-exon junctions, and transcripts.

For human samples, variant calling is performed at a list of highly variable sites. A single [VCF file](https://www.internationalgenome.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-40/) is generated, containing genotype calls for all samples. This establishes a sufficiently unique profile for each sample which can be compared against pre-sequencing genotype calls to resolve potential identity issues.

Optionally, [BigWig](https://genome.ucsc.edu/goldenPath/help/bigWig.html) coverage files by strand, and averaged across strands, are produced. An additional step utilizes the [derfinder](https://bioconductor.org/packages/release/bioc/html/derfinder.html) package to compute expressed regions, a precursor to analyses finding differentially expressed regions (DERs).

### Installation and set-up

SPEAQeasy is available from GitHub, and can be cloned:

```{bash, eval=FALSE}
git clone git@github.com:LieberInstitute/SPEAQeasy.git
```

SPEAQeasy dependencies are recommended to be installed using Docker or Singularity, involving a fairly quick installation that is reproducible regardless of the computing environment. We'll install using Singularity, which is more frequently available in high-performance computing (HPC) environments:

```{bash, eval=FALSE}
#  From within the SPEAQeasy repository:
bash install_software.sh "singularity"
```

If neither Docker nor Singularity are available, a more experimental local installation of all dependencies is an option:

```{bash, eval=FALSE}
#  From within the SPEAQeasy repository:
bash install_software.sh "local"
```

Please note that installation may take some time-- **installing and running SPEAQeasy is not required for this workshop**, but is demonstrated to familiarize participants with the process. The workshop files include outputs from SPEAQeasy, which can be used later for the example differential expression analysis.

Choose a "main" script as appropriate for your particular set-up. "Main" scripts and associated configuration files exist for SLURM-managed computing clusters, SGE-managed clusters, local machines, and the [JHPCE](https://jhpce.jhu.edu/) cluster.

| Environment  | "Main" script | Config file |
|----------|-------------|-------------------|
| SLURM cluster   | run_pipeline_slurm.sh      | conf/slurm.config or conf/docker_slurm.config  |
| SGE cluster     | run_pipeline_sge.sh        | conf/sge.config or conf/docker_sge.config      |
| Local machine   | run_pipeline_local.sh      | conf/local.config or conf/docker_local.config  |
| JHPCE cluster   | run_pipeline_jhpce.sh      | conf/jhpce.config                              |

Within the main script, you can configure arguments specific to the experiment, such as the reference organism, pairing of samples, and where to place output files, among other specifications.

When running SPEAQeasy on a cluster (i.e. SLURM, SGE, or JHPCE users), it is recommended you submit the pipeline as a job, using the command appropriate for your cluster. For those running SPEAQeasy locally, the main script can be executed directly, ideally as a background task.

```{bash, eval=FALSE}
#  SLURM-managed clusters
sbatch run_pipeline_slurm.sh

#  SGE-managed clusters
qsub run_pipeline_sge.sh

#  local machines
bash run_pipeline_local.sh

#  The JHPCE cluster
qsub run_pipeline_jhpce.sh
```

Here's an example of such a file:

```{r "example_run_file"}
example_main_script <- system.file(
    "extdata", "SPEAQeasy-example", "run_pipeline_jhpce.sh",
    package = "SPEAQeasyWorkshop2023"
)

cat(paste0(readLines(example_main_script), "\n"))
```


## Data overview and download 

The dataset that will be used for the DEA can be downloaded from the `smokingMouse` package. 
Visit [here](http://research.libd.org/smokingMouse/) for more details of this package.

We will download the `rse_gene` object which contains the data at the gene level. We'll use `r Biocpkg("BiocFileCache")` to keep the data in a local cache in case we want to run this example again and don't want to re-download the data from the web.

<img src="figs/Data_overview_experiments.png" width="750px"  />
    <figcaption style="color: gray; line-height: 0.85; text-align: justify"><font size="-2"><b>Figure 1</b>: <b>Experimental design of the study. </b> The data was obtained from an experiment in which <b>A)</b> 6 pregnant dams and 17 non-pregnant female adult mice were administered nicotine by intraperitoneal injection (IP; n=12) or controls (n=11). A total of 42 pups were born to pregnant dams: 19 were born to mice that were administered nicotine and 23 to control mice. Samples from frontal cortices of P0 pups and adults were obtained. <b>B)</b> RNA was extracted from those samples and then RNA-seq libraries were prepared and sequenced to obtain the expression counts of the genes. At the end, a dataset of 65 samples (either from adult or pup brain) and 55,401 genes was generated. </font> 

   </figcaption>



The data reside in a `RangedSummarizedExperiment` (RSE) object, containing the following assays:

* `counts`: original read counts of the 55,401 mouse genes across 208 samples (including the 65 nicotine samples).
* `logcounts`: normalized and scaled counts ($log_2(CPM + 0.5)$) of the same genes across the same samples; normalization was carried out applying TMM method with `cpm(calcNormFactors())` of `r BiocStyle::Biocpkg("edgeR")`.


The same RSE contains the sample information in `colData(RSE)` and the gene information in `rowData(RSE)`.

<mark style= "background-color: #FCF3CF"> Yellow </mark> variables correspond to SPEAQeasy outputs that are going to be used in downstream analyses. 

<mark style= "background-color: #FAECF8"> Pink </mark> variables are specific to the study, such as sample metadata and some others containing additional information about the genes. 

<mark style= "background-color: #DFF0FE"> Blue </mark> variables are quality-control metrics computed by `addPerCellQC()` of `r BiocStyle::Biocpkg("scuttle")`.


### Gene Information 

Among the information in `rowData(RSE)` the next variables are of interest for the analysis:

* <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> gencodeID</span> </mark>: GENCODE ID of each gene.
* <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> ensemblID</span> </mark>: gene ID in Ensembl database.
* <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> EntrezID</span> </mark>: identifier of each gene in NCBI Entrez database.
* <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> Symbol</span> </mark>:  official gene symbol for each mouse gene.
* <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> retained_after_feature_filtering</span> </mark>: Boolean variable that equals TRUE if the gene passed the gene filtering (with `filterByExpr()` of `r BiocStyle::Biocpkg("edgeR")`) based on its expression levels and FALSE if not.

### Sample Information 

Sample information in `colData(RSE)` contains the following variables:

* <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> SAMPLE_ID</span> </mark>: is the name of the sample.
* <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> ERCCsumLogErr</span> </mark>: a summary statistic quantifying overall difference of expected and actual ERCC concentrations for one sample. 
* <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> overallMapRate</span> </mark>: the decimal fraction of reads which successfully mapped to the reference genome (i.e. *numMapped* / *numReads*).
* <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> mitoMapped</span> </mark>: the number of reads which successfully mapped to the mitochondrial chromosome.
* <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> totalMapped</span> </mark>: the number of reads which successfully mapped to the canonical sequences in the reference genome (excluding mitochondrial chromosomes).
* <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> mitoRate</span> </mark>: the decimal fraction of reads which mapped to the mitochondrial chromosome, of those which map at all (i.e. *mitoMapped* / (*totalMapped* + *mitoMapped*))
* <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> totalAssignedGene</span> </mark>: the decimal fraction of reads assigned unambiguously to a gene (including mitochondrial genes), with `featureCounts` (Liao et al. 2014), of those in total.
* <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> rRNA_rate</span> </mark>: the decimal fraction of reads assigned to a gene whose type is ‘rRNA’, of those assigned to any gene.
* <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> Tissue</span> </mark>: tissue (mouse brain or blood) from which the sample comes.
* <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> Age</span> </mark>: if the sample comes from an adult or a pup mouse.
* <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> Sex</span> </mark>: if the sample comes from a female (F) or male (M)  mouse.
* <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> Expt</span> </mark>:  the experiment (nicotine or smoking exposure) to which the sample mouse was subjected; it could be an exposed or a control mouse of that experiment.
* <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> Group</span> </mark>: if the sample belongs to a nicotine/smoking-exposed mouse (Expt) or a nicotine/smoking control mouse (Ctrl).
* <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> plate</span> </mark>: is the plate (1,2 or 3) in which the sample library was prepared.
* <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> Pregnancy</span> </mark>:  if the sample comes from a pregnant (Yes) or not pregnant (No) mouse.
* <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> medium</span> </mark>: is the medium in which the sample was treated: water for brain samples and an elution buffer (EB) for the blood ones.
* <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> flowcell</span> </mark>: is the sequencing batch of each sample.
* <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> sum</span> </mark>: library size (total sum of counts across all genes for each sample).
* <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> detected</span> </mark>: number of non-zero expressed genes in each sample.
* <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> subsets_Mito_sum</span> </mark>: total sum of read counts of mt genes in each sample.
* <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> subsets_Mito_detected</span> </mark>: total number of mt genes in each sample.
* <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> subsets_Mito_percent</span> </mark>: % of mt genes' counts of the total counts of the sample. 
* <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> subsets_Ribo_sum</span> </mark>: total sum of read counts of ribosomal genes in each sample.
* <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> subsets_Ribo_detected</span> </mark>: total number of ribosomal genes in each sample.
* <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> subsets_Ribo_percent</span> </mark>: % of ribosomal genes' counts of the total counts of the sample.

<div class="alert alert-info">
  <strong>Note:</strong> in our case, we'll use samples from the nicotine experiment only, so all samples come from brain and were treated in water.
</div>

```{r "examine_outputs"}

library(here)
library(SummarizedExperiment)
library(smokingMouse)


################################
#         Retrieve data
################################
# Provisional data source until smokingMouse pkg is accepted
load(here('provisional_data/rse_gene_mouse_RNAseq_nic-smo.Rdata'), verbose=TRUE)
## Explore complete rse_gene object (including smoking-exposed samples as well)
rse_gene

#################################
##   Extract data of interest
#################################
## Keep samples from nicotine experiment only
rse_gene_nic <- rse_gene[,which(rse_gene$Expt=='Nicotine')]
## New dimensions
dim(rse_gene_nic)

#################################
##  Access gene expression data
#################################

## Raw counts for first 3 genes in the first 5 samples
assays(rse_gene_nic)$counts[1:3,1:5]
## Log-normalized counts for first 3 genes in the first 5 samples
assays(rse_gene_nic)$logcounts[1:3,1:5]

#################################
##     Access sample data
#################################

## Info for the first 2 samples
head(colData(rse_gene_nic), 2)

```

## Differential Expression Analysis

In this workshop we'll guide you through a differential expression analysis following the steps presented in **Figure 2**. The main objective is to determine which genes are affected by nicotine administration (in adults) or exposure (in pups) in the mouse brain. 

<figure>
    <img src="figs/Analyses.png" width="700px" align=center />
        <figcaption style="color: gray; line-height: 0.87; text-align: justify">
            <font size="-1.8">
                <b>Figure 2</b>: <b> Summary of the analyses. </b> Steps in yellow will be properly performed. 
                <b> 1. Data Preparation </b> 
                <b> 2. Exploratory Data Analysis </b> 
                <b> 3. Differential Expression Analysis </b> 
                <b> 4. GO & KEGG Analyses  </b> 
                <b> 5. DEG visualization  </b> </font> 
            <font size="0.8"> Abbreviations: CPM: counts per million; QC: quality control; PC: principal component; 
            DEG: differentially expressed genes; GO: Gene Ontology; KEGG: Kyoto Encyclopedia of Genes and Genomes.</font>
        </figcaption>
</figure>

### 1. Data preparation {-}
Before exploring a dataset, normalization and filtering steps are generally appropriate. In our case, SPEAQeasy outputs contain normalized counts, and the example data we'll use already has non-expressed genes filtered out. Regardless, let's review why these steps are important and how to extract the data we need for downstream analyses.

**1.1 Data normalization**

Data normalization is a preliminary step when working with expression data because raw counts do not necessarily reflect real expression measures of the genes, since there are technical differences in the way the libraries are prepared and sequenced, as well as intrinsic differences in the genes that are translated into more or less mapped reads. Particularly, there are *within-sample effects* that are the differences between genes in the same sample, such as their **length** (the longer the gene, the more reads it will have) and **GC content**, factors that contribute to variations in their counts. On the other hand, *between-sample effects* are differences between samples such as the **sequencing depth**, i.e., the total number of molecules sequenced, and the **library size**, i.e., the total number of reads of each sample [1].

These variables lead to significantly different mRNA amounts but of course are not due to the biological or treatment conditions of interest (such as nicotine administration in this example) so in order to remove, or at least, to minimize this technical bias and obtain measures comparable and consistent across samples, raw counts must be normalized by these factors. The data that we'll use in this case are already normalized in `assays(rse_gene)$logcounts`. Specifically, the assay contains counts per million (CPM), also known as reads per million (RPM), one basic gene expression unit that only normalizes by the sequencing depth and is computed by dividing the read counts of a gene in a sample by a scaling factor given by the total mapping reads of the sample per million [2]:

$$CPM = \frac{read \ \ counts \ \ of \ \ gene \ \ \times \ \ 10^6 }{Total \ \ mapping \ \ reads \ \ of \ \ sample}$$ 


As outlined in [**Data overview and download**](##Data overview and download), the scaling factors were obtained with `calcNormFactors()` applying the Trimmed Mean of M-Values (TMM) method, the `r BiocStyle::Biocpkg("edgeR")` package's default normalization method that assumes that most genes are not differentially expressed. The effective library sizes of the samples and the CPM of each observation were computed with the `r BiocStyle::Biocpkg("edgeR")` function `cpm()` setting the `log` argument to `TRUE` and `prior.count` to 0.5 to receive values in $log_2(CPM+0.5)$.

After data normalization and scaling, we’d expect the read counts to follow a normal distribution, something we can confirm by comparing the counts' distribution before and after the normalization. Note that both datasets contain the exact same genes.

```{r "Data preparation", message=FALSE, warning=FALSE}

library(ggplot2)

## Histogram and density plot of read counts before and after normalization 

## Raw counts
counts_data <- data.frame(counts=as.vector(assays(rse_gene_nic)$counts))
plot <- ggplot(counts_data, aes(x=counts)) +
    geom_histogram(colour="black", fill="lightgray") +
    labs(x='read counts', y='Frecuency') +
    theme_classic() +
    theme(plot.margin = unit(c(0.5,1,0.2,1), "cm")) 
plot

## Normalized counts
logcounts_data <- data.frame(logcounts=as.vector(assays(rse_gene_nic)$logcounts))
plot <- ggplot(logcounts_data, aes(x=logcounts)) +
    geom_histogram(aes(y=..density..), colour="darkgray", fill="lightgray") +
    theme_classic() +
    geom_density(fill="#69b3a2", alpha=0.3) +
    labs(x='log(CPM+0.5)', y='Frecuency') +
    theme(plot.margin = unit(c(0.5,1,0.2,1), "cm")) 
plot

```

As presented, after data transformation, we can now see a more widespread distribution of the counts, but note that most of them are zeros in the first plot (the one with the raw counts) and those zeros remain after normalization, corresponding to counts below 0 in the second plot. That is because we haven't filtered the low and zero-expressed genes. 


**1.2 Gene filtering**

Low-expressed or non-expressed genes in many samples are not of biological interest in a study of differential expression because they don't inform about the gene expression changes and they are, by definition, not differentially expressed, so we have to drop them using `filterByExpr()` from `r BiocStyle::Biocpkg("edgeR")` that only keeps genes with at least *K* CPM in *n* samples and with a minimum total number of counts across all samples.

```{r,  message=FALSE, warning=FALSE}

## Retain genes that passed filtering step
rse_gene_filt <- rse_gene_nic[rowData(rse_gene_nic)$retained_after_feature_filtering==TRUE,]

## Normalized counts and filtered genes
filt_logcounts_data <- data.frame(logcounts=as.vector(assays(rse_gene_filt)$logcounts))

## Plot
plot <- ggplot(filt_logcounts_data, aes(x=logcounts)) +
    geom_histogram(aes(y=..density..), colour="darkgray", fill="lightgray") +
    theme_classic() +
    geom_density(fill="#69b3a2", alpha=0.3) +
    labs(x='log(CPM+0.5)', y='Frecuency') +
    theme(plot.margin = unit(c(0.5,1,0.2,1), "cm")) 
plot


```

In this third plot we can observe a curve that is closer (though not completely) to a normal distribution and with less lowly-expressed genes. 

With the object `rse_gene_filt` we can proceed with downstream analyses. 


### 2. Exploratory Data Analysis {-}
The first formal step that we will be performing is the sample exploration. This crucial initial part of the analysis consists of an examination of differences and relationships between Quality-Control (QC) metrics of the samples from different groups in order to identify poor-quality samples that must be removed before DEA. After that, we will evaluate the similarities and grouping (not clustering) of the samples by their gene expression variation through a dimensionality reduction analysis. Finally, the sample variables in the metadata also need to be analyzed and filtered based on the percentage of gene expression variance that they explain for each gene. 


**2.1 Quality Control Analysis**

First we have to explore and compare the quality-control metrics of the samples in the different groups given by covariates such as age, sex, pregnancy state, group, plate and flowcell. See [**Sample Information**](###Sample Information) for a description of these variables.  

<style>
p.comment {
background-color: #DBDBDB;
padding: 10px;
border: 1px solid black;
margin-left: 25px;
border-radius: 5px;
font-family: sans-serif;
}

</style>

<p class="comment">
❓ ***Why is that relevant?*** As you could imagine, technical and methodological differences in all the steps that were carried out during the experimental stages are potential sources of variation in the quality of the samples. Just imagine all that could have been gone wrong or unexpected while experimenting with mice, during the sampling, in the RNA extraction using different batches, when treating samples in different mediums, when preparing libraries in different plates and sequencing in different flowcells. Moreover, the inherent features of the mice from which the samples come from such as age, tissue, sex and pregnancy state could also affect the samples' metrics if, for example, they were separately analyzed and processed. 
</p>


<p class="comment">
❓***But why do we care about mitochondrial and ribosomal counts as QC metrics?*** In the process of mRNA extraction, either by mRNA enrichment (capturing polyadenylated mRNAs) or rRNA-depletion (removing rRNA), we’d expect to have a low number of ribosomal counts, i.e., counts that map to rDNA, and if we don’t, something must have gone wrong with the procedures. In the case of mitochondrial counts something similar occurs: higher mitochondrial rates will be obtained if the cytoplasmic mRNA capture was deficient or if the transcripts were lost by some technical issue, increasing the proportion of mitochondrial mRNAs. As a result, high <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> mitoRate</span> </mark> and <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> rRNA_rate</span> </mark> imply poor quality in the samples.
</p>


<div class="alert alert-info">
  <strong>Note:</strong> the QC metrics were computed with the unprocessed datasets (neither filtered nor normalized) to preserve the original estimates of the samples.
</div>


*2.1.1 Evaluate QC metrics for groups of samples*

Fortunately, we can identify to some extent possible factors that could have influenced on the quality of the samples, as well as isolated samples that are problematic. To do that, we will create boxplots that present the distribution of the samples' metrics separating them by sample variables. 

```{r "QC_boxplots",  message=FALSE, warning=FALSE}
library(Hmisc)
library(stringr)
library(cowplot)

## Define QC metrics of interest
qc_metrics <- c('mitoRate', 'overallMapRate', 'totalAssignedGene', 'rRNA_rate', 'sum', 'detected', 'ERCCsumLogErr')

## Define sample variables of interest
sample_variables <- c('Group', 'Age', 'Sex', 'Pregnancy', 'plate', 'flowcell')


## Function to create boxplots of QC metrics for groups of samples

QC_boxplots <- function(qc_metric, sample_var){
    
    ## Define sample colors depending on the sample variable
    if (sample_var=="Group"){
        colors=c('Control'='brown2', 'Experimental'='deepskyblue3')
    }
    else if (sample_var=="Age"){
        colors=c('Adult'='slateblue3', 'Pup'='yellow3')
    }
    else if (sample_var=="Sex"){
        colors=c('F'='hotpink1', 'M'='dodgerblue')
    }
    else if (sample_var=='Pregnancy'){
        colors=c('Yes'='darkorchid3', 'No'='darkolivegreen4')
    }
    else if (sample_var=='plate'){
        colors=c('Plate1'='darkorange', 'Plate2'='lightskyblue', 'Plate3'='deeppink1')
    }
    else if (sample_var=='flowcell'){
        colors=c('HKCG7DSXX'='chartreuse2','HKCMHDSXX'='magenta', 'HKCNKDSXX'='turquoise3', 
                 'HKCTMDSXX'='tomato')
    }
    
    ## Axis labels
    x_label=capitalize(sample_var)
    y_label=str_replace_all(qc_metric, c("_"=""))
    
    ## x-axis text angle and position
    if (sample_var=='flowcell'){
        x_axis_angle = 18
        x_axis_hjust = 0.5
        x_axis_vjust = 0.7
        x_axis_size = 4
    }
    else {
        x_axis_angle = 0
        x_axis_hjust = 0.5
        x_axis_vjust = 0.5
        x_axis_size = 6
    }
    
    ## Extract sample data in colData(rse_gene_filt)
    data <- data.frame(colData(rse_gene_filt))
    
    ## Sample variable separating samples in x-axis and QC metric in y-axis
    ## (Coloring by sample variable)
    plot <- ggplot(data = data, mapping = aes(x = !! rlang::sym(sample_var), y = !! rlang::sym(qc_metric), color = !! rlang::sym(sample_var))) +
                ## Add violin plots
                geom_violin(alpha = 0, size = 0.4, color='black', width=0.7)+
                ## Spread dots
                geom_jitter(width = 0.08, alpha = 0.7, size = 1.3) +
                ## Add boxplots
                geom_boxplot(alpha = 0, size = 0.4, width=0.1, color='black') +
                ## Set colors
                scale_color_manual(values = colors) +
                ## Define axis labels
                labs(y= y_label, x = x_label) +
                ## Get rid of the background
                theme_bw() +
                ## Hide legend and define plot margins and size of axis title and text
                theme(legend.position = "none",
                      plot.margin = unit(c(0.5,0.4,0.5,0.4), "cm"),
                      axis.title = element_text(size = 7),
                      axis.text = element_text(size = x_axis_size),
                      axis.text.x = element_text(angle = x_axis_angle, hjust = x_axis_hjust, vjust = x_axis_vjust)) 
        

    return(plot)
}



## Plots of all QC metrics for each sample variable
multiple_QC_boxplots <- function(sample_var){
    
    i=1
    plots = list()
    for (qc_metric in qc_metrics) {
        ## Call function to create each individual plot
        plots[[i]]<- QC_boxplots(qc_metric, sample_var)
        i=i+1
    }
    ## Arrange multiple plots into a grid
    print(plot_grid(plots[[1]], plots[[2]], plots[[3]], plots[[4]], plots[[5]], plots[[6]], plots[[7]], nrow = 2))
}

```


```{r message=FALSE, warning=FALSE}
multiple_QC_boxplots('Age')
```

Initially, when we separate samples by Age, we can appreciate a clear segregation of adult and pup samples in <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> mitoRate</span> </mark>, with higher mitochondrial rates for adult samples and thus, being lower quality samples than the pup ones. We can also see that pup samples have higher <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> totalAssignedGene</span> </mark>, again being higher quality. The samples are very similar in the rest of the QC metrics.
The former differences must be taken into account because they guide further sample separation by Age, which is necessary to avoid dropping most of the adult samples (that are lower quality) in the **QC-based sample filtering** (see below) and to prevent misinterpreting sample variation given by quality and not by mouse age itself. 

```{r message=FALSE, warning=FALSE}
multiple_QC_boxplots('Sex')
```

With Sex, female and male samples have similar QC metrics but there are some female samples that have high <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> mitoRate</span> </mark> and low <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> totalAssignedGene</span> </mark>, but males don´t. That is consistent with the fact that all males are pups and as we previously observed, pup samples present better QC metrics, but since not all pups are males, some female pups also have smaller <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> mitoRate</span> </mark> and greater <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> totalAssignedGene</span> </mark>. So in this case samples' differences are actually dictated by Age and not by Sex.


```{r message=FALSE, warning=FALSE}
multiple_QC_boxplots('Group')
```

Notably, for Group no evident contrasts are seen in the quality of the samples, which means that both control and exposed samples have similar metrics and therefore the differences between them won't be determined by technical factors but effectively by gene expression changes. 

```{r message=FALSE, warning=FALSE}
multiple_QC_boxplots('Pregnancy')
```

The Pregnancy variable is interesting because pregnant dams were obviously females and adults, which we already noted are lower quality. Accordingly, pregnant samples present overall higher <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> mitoRate</span> </mark> and lower <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> totalAssignedGene</span> </mark>, <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> sum </span> </mark>  (library size) and <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> detected </span> </mark> (number of expressed genes) than some samples coming from non-pregnant mice. However, these samples have smaller |<mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> ERCCsumLogErr</span> </mark>|, which means that they had smaller differences between expected and observed concentrations of control transcripts. 

We must clarify one more time that as in Sex, the trends observed in Pregnancy are all given by Age: it is Age that fully segregates samples in <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> mitoRate</span> </mark> and almost completely in </mark>, lower <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> totalAssignedGene</span> </mark>. Because all pregnant dams were adults, their metrics will be positioned where the adult samples were, but there were also not-pregnant adults that share similar QC values. 


```{r message=FALSE, warning=FALSE}
multiple_QC_boxplots('plate')
```

In plate, no alarming differences are presented but some samples from the 1st plate
have low (though not much lower) <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> overallMapRate</span> </mark>,  <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> totalAssignedGene</span> </mark> and <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> sum</span> </mark>. 

```{r message=FALSE, warning=FALSE}
multiple_QC_boxplots('flowcell')
```

For the flowcell, again no worrying distinctions are seen, with the exception of a few individual samples far from the rest. 

In all the previous plots we can appreciate a group of samples placed below in <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> mitoRate</span> </mark> that correspond to pup samples. The relationship is more fuzzy in <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> totalAssignedGene</span> </mark>.


*2.1.2 Examine relationships between QC metrics of samples*

Secondly, we want to analyze if there are relationships between QC variables that explain the presence of samples with two or more particular QC metrics. For that, scatterplots are useful because they enable the evaluation of pairs of samples' QC metrics. 

```{r "QC scatterplots", message=FALSE, warning=FALSE}

## Scatterplots for a pair of QC metrics

QC_scatterplots <- function(sample_var, qc_metric1, qc_metric2){

    ## Define sample colors
    if (sample_var=="Group"){
        colors=c('Control'='brown2', 'Experimental'='deepskyblue3')
    }
    else if (sample_var=="Age"){
        colors=c('Adult'='slateblue3', 'Pup'='yellow3')
    }
    else if (sample_var=="Sex"){
        colors=c('F'='hotpink1', 'M'='dodgerblue')
    }
    else if (sample_var=='Pregnancy'){
        colors=c('Yes'='darkorchid3', 'No'='darkolivegreen4')
    }
    else if (sample_var=='plate'){
        colors=c('Plate1'='darkorange', 'Plate2'='lightskyblue', 'Plate3'='deeppink1')
    }
    else if (sample_var=='flowcell'){
        colors=c('HKCG7DSXX'='chartreuse2','HKCMHDSXX'='magenta', 'HKCNKDSXX'='turquoise3', 
                 'HKCTMDSXX'='tomato')
    }

    data <- colData(rse_gene_filt)

    ## Scatterplots for continuous variable vs continuous variable
    ## First QC metric in x-axis and second QC metric in y-axis
    plot <- ggplot(as.data.frame(data), aes(x=!! rlang::sym(qc_metric1),
                                            y=!! rlang::sym(qc_metric2),
                                            ## Color samples by a variable
                                           color=!! rlang::sym(sample_var))) +
        ## Add scatterplot
        geom_point(size=1) +
        ## Add regression line
        stat_smooth (geom="line", alpha=0.4, size=0.4, span=0.25, method = lm, color='orangered3') +
        ## Colors
        scale_color_manual(name = sample_var, values = colors) +
        theme_bw() +
        ## Add Pearson correlation coefficient between the metrics as subtitle
        labs(subtitle = paste0("Corr: ", signif(cor(data[,qc_metric1], data[,qc_metric2], method = "pearson"), digits=3)), 
             ## Add axis labels
             y= gsub('_', ' ', qc_metric2), 
             x = gsub('_', ' ', qc_metric1)) +
        ## Plot margins and text size
        theme(plot.margin=unit (c (0.4,0.1,0.4,0.1), 'cm'),
              axis.title = element_text(size = (7)),
              axis.text = element_text(size = (6)),
              plot.subtitle = element_text(size = 7, color='gray40'),
              legend.text = element_text(size=6),
              legend.title = element_text(size=7))
    return(plot)
    
}



## QC scatterplots coloring by all sample variables
multiple_QC_scatterplots <- function(qc_metric1, qc_metric2){
    
    sample_variables <- c('Age', 'Sex', 'Group', 'plate', 'Pregnancy', 'flowcell')
    
    i=1
    plots = list()
    for (sample_var in sample_variables) {
        plots[[i]]<- QC_scatterplots(sample_var, qc_metric1, qc_metric2)
        i=i+1
    }
    print(plot_grid(plots[[1]], plots[[2]], plots[[3]], plots[[4]], plots[[5]], plots[[6]], nrow = 2, rel_widths = c(1.1,1,1.2) ))
}



```

```{r message=FALSE, warning=FALSE}
multiple_QC_scatterplots('mitoRate', 'rRNA_rate')
```

For <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> mitoRate</span> </mark> vs <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> rRNA_rate</span> </mark> plots there's a negligible positive correlation which was anticipated since no obvious relationship exists between them, though if cytoplasmic mRNAs weren't well captured (lowering the total number of mapping reads) we could expect an increase in both mitochondrial and ribosomal rates. 

Note the complete separation of samples in <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> mitoRate</span> </mark> (but not in <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> rRNA_rate</span> </mark>) by Age, which in turn causes the grouping of male samples (all pups). 


```{r message=FALSE, warning=FALSE}
multiple_QC_scatterplots('mitoRate', 'totalAssignedGene')
```

<mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> mitoRate</span> </mark>  and <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> totalAssignedGene</span> </mark> have a strong negative correlation, which reveals the existence of very bad quality samples (with small <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> totalAssignedGene</span> </mark> and high <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> mitoRate</span> </mark>).

```{r message=FALSE, warning=FALSE}
multiple_QC_scatterplots('sum', 'detected')
```
Noticeably, a positive correlation is present for <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> sum</span> </mark> and <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> detected</span> </mark> and that is also expected because a higher number of expressed genes implies bigger library sizes, but the converse is not necessarily true: just a few highly expressed genes (small number of <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> detected</span> </mark> genes) could increase the library size. Interestingly, in this case samples are diagonally separated by age. 


```{r message=FALSE, warning=FALSE}
multiple_QC_scatterplots('sum', 'totalAssignedGene')
```

Contrary to expectations, <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> sum</span> </mark> is not correlated to <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> totalAssignedGene</span> </mark> and this is unexpected because higher proportions of genes' reads could initially increase library size. However, since <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> totalAssignedGene</span> </mark> is a fraction of the total reads, it could be that a high fraction of reads mapping to genes reflects a small number of total reads and does not necessarily imply more expressed genes (and therefore bigger libraries). In other words, even when <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> totalAssignedGene</span> </mark> equals 1, if the total number of reads is small, the genes won't be widely covered by them and the libraries won't be sizable. Observe the separation of samples by Age. 


```{r message=FALSE, warning=FALSE}
multiple_QC_scatterplots('detected', 'totalAssignedGene')
```

Lastly, there's a slight positive correlation between <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> totalAssignedGene</span> </mark> and <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> detected</span> </mark>, which makes sense because reads assigned to genes rise the number of non-zero expressed genes but as mentioned, those genes are not necessarily highly expressed and thus, library sizes are not perforce so much bigger. Also note that the correlation wouldn't be conserved within pup samples. 


Remarkably, samples were not differentiated by any other variable (including Group!) but the differences between adult and pup metrics suggest that they must be split into two different groups for the following analyses.


*2.1.3 QC sample filtering*

After assessing how different or similar are the QC values between samples, we can now proceed to sample filtering based precisely, on these metrics. For that, we will use `isOutlier()` from `r BiocStyle::Biocpkg("scater")` to identify outlier samples only at the lower end or the higher end, depending on the QC metric.

```{r "QC sample filtering", message=FALSE, warning=FALSE}

library(scater)
library(rlang)
library(ggrepel)

## Separate data by Age
rse_gene_pups<- rse_gene_filt[,which(rse_gene_filt$Age=="Pup")]
rse_gene_adults <- rse_gene_filt[,which(rse_gene_filt$Age=="Adult")]


## Find outlier samples based on their QC metrics (samples that are 3 median-absolute-deviations away from the median)


## Filter all samples together

## Drop samples with lower library sizes (sum), detected number of genes and totalAssignedGene 
outliers_library_size <- isOutlier(rse_gene_filt$sum, nmads = 3, type="lower")
outliers_detected_num <- isOutlier(rse_gene_filt$detected, nmads = 3, type="lower")
outliers_totalAssignedGene <- isOutlier(rse_gene_filt$totalAssignedGene, nmads = 3, type="lower")
## Drop samples with higher mitoRates and rRNA rates 
outliers_mito<-isOutlier(rse_gene_filt$mitoRate, nmads = 3, type="higher")
outliers_rRNArate <- isOutlier(rse_gene_filt$rRNA_rate, nmads = 3, type="higher")

## Keep not outlier samples
not_outliers<-which(! (outliers_library_size | outliers_detected_num | outliers_totalAssignedGene | outliers_mito | outliers_rRNArate))
rse_gene_qc<-rse_gene_filt[,not_outliers]

## Number of samples retained
dim(rse_gene_qc)[2]

## Add new variables to rse_gene_filt with info of samples retained/dropped
rse_gene_filt$Retention_after_QC_filtering <- as.vector(sapply(rse_gene_filt$SAMPLE_ID, function(x){if (x %in% rse_gene_qc$SAMPLE_ID){'Retained'} else{'Dropped'}}))



## Filter adult samples

outliers_library_size <- isOutlier(rse_gene_adults$sum, nmads = 3, type="lower")
outliers_detected_num <- isOutlier(rse_gene_adults$detected, nmads = 3, type="lower")
outliers_totalAssignedGene <- isOutlier(rse_gene_adults$totalAssignedGene, nmads = 3, type="lower")
outliers_mito<-isOutlier(rse_gene_adults$mitoRate, nmads = 3, type="higher")
outliers_rRNArate <- isOutlier(rse_gene_adults$rRNA_rate, nmads = 3, type="higher")

not_outliers<-which(! (outliers_library_size | outliers_detected_num | outliers_totalAssignedGene | outliers_mito | outliers_rRNArate))
rse_gene_adults_qc<-rse_gene_adults[,not_outliers]

## Number of samples retained
dim(rse_gene_adults_qc)[2]

rse_gene_adults$Retention_after_QC_filtering <- as.vector(sapply(rse_gene_adults$SAMPLE_ID, function(x){if (x %in% rse_gene_adults_qc$SAMPLE_ID){'Retained'} else{'Dropped'}}))



## Filter pup samples

outliers_library_size <- isOutlier(rse_gene_pups$sum, nmads = 3, type="lower")
outliers_detected_num <- isOutlier(rse_gene_pups$detected, nmads = 3, type="lower")
outliers_totalAssignedGene <- isOutlier(rse_gene_pups$totalAssignedGene, nmads = 3, type="lower")
outliers_mito<-isOutlier(rse_gene_pups$mitoRate, nmads = 3, type="higher")
outliers_rRNArate <- isOutlier(rse_gene_pups$rRNA_rate, nmads = 3, type="higher")

not_outliers<-which(! (outliers_library_size | outliers_detected_num | outliers_totalAssignedGene | outliers_mito | outliers_rRNArate))
rse_gene_pups_qc<-rse_gene_pups[,not_outliers]

## Number of samples retained
dim(rse_gene_pups_qc)[2]

rse_gene_pups$Retention_after_QC_filtering <- as.vector(sapply(rse_gene_pups$SAMPLE_ID, function(x){if (x %in% rse_gene_pups_qc$SAMPLE_ID){'Retained'} else{'Dropped'}}))


```

We already filtered outlier samples ... but what have we removed?
It is always important to trace the QC metrics of the filtered samples to verify that they really are poor quality. We don't want to get rid of useful samples! So let's go back to the QC boxplots but now color samples according to whether they passed the filtering step, and also distinguishing samples' groups by shape. 

```{r message=FALSE, warning=FALSE}

## Boxplots of QC metrics after sample filtering

## Boxplots
boxplots_after_QC_filtering <- function(rse_gene, qc_metric, sample_var){

    ## Color samples 
    colors=c('Retained'='deepskyblue', 'Dropped'='brown2')

    ## Sample shape by sample variables 
    if (sample_var=="Group"){
        shapes=c('Control'=0, 'Experimental'=15)
    }
    else if (sample_var=="Age"){
        shapes=c('Adult'=16, 'Pup'=1)
    }
    else if (sample_var=="Sex"){
        shapes=c('F'=11, 'M'=19)
    }
    else if (sample_var=='Pregnancy'){
        shapes=c('Yes'=10, 'No'=1)
    }
    else if (sample_var=='plate'){
        shapes=c('Plate1'=12, 'Plate2'=5, 'Plate3'=4)
    }
    else if (sample_var=='flowcell'){
        shapes=c('HKCG7DSXX'=3,'HKCMHDSXX'=8, 'HKCNKDSXX'=14, 
                 'HKCTMDSXX'=17)
    }


    y_label=str_replace_all(qc_metric, c("_"=" "))
    
    data <- data.frame(colData(rse_gene))

    ## Median of the QC var values
    median<-median(eval(parse_expr(paste("rse_gene$", qc_metric, sep=""))))
    ## Median-absolute-deviation of the QC var values
    mad<-mad(eval(parse_expr(paste("rse_gene$", qc_metric, sep=""))))

    plot <- ggplot(data = data, mapping = aes(x = '', y = !! rlang::sym(qc_metric), 
                                              color = !! rlang::sym('Retention_after_QC_filtering'))) +
        geom_jitter(alpha = 1, size = 2, aes(shape=eval(parse_expr((sample_var))))) +
        geom_boxplot(alpha = 0, size = 0.15, color='black') +
        scale_color_manual(values = colors) +
        scale_shape_manual(values = shapes) +
        labs(x="", y = y_label, color='Retention after QC filtering', shape=sample_var) +
        theme_classic() +
        ## Median line
        geom_hline(yintercept = median, size=0.5) +
        ## Line of median + 3 MADs
        geom_hline(yintercept = median+(3*mad), size=0.5, linetype=2) +
        ## Line of median - 3 MADs
        geom_hline(yintercept = median-(3*mad), size=0.5, linetype=2) +
        theme(axis.title = element_text(size = (9)),
              axis.text = element_text(size = (8)),
              legend.position="right",
              legend.text = element_text(size=8),
              legend.title = element_text(size=9))

    return(plot)
}


```

In the following plot we can confirm that taking all samples together (both adults and pups) all adult samples are dropped by their high <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> mitoRate</span> </mark>, but that is not desirable because we want to analyze these samples too, so we need to analyze pup and adult samples separately. 

```{r message=FALSE, warning=FALSE}

## Plots

## All samples together
p <- boxplots_after_QC_filtering(rse_gene_filt, 'mitoRate', 'Age')
p + theme(plot.margin=unit (c (2,4,2,4), 'cm'))

```
For adults, only 3 controls are dropped, 3 of them by their <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> mitoRate</span> </mark>. 
```{r message=FALSE, warning=FALSE}
## Adult samples
p <- boxplots_after_QC_filtering(rse_gene_adults, 'mitoRate', 'Group')
p + theme(plot.margin=unit (c (2,4,2,4), 'cm'))
```

For pups, 1 experimental sample was dropped by its <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> rRNA_rate</span> </mark>.

```{r message=FALSE, warning=FALSE}
## Pup samples
p <- boxplots_after_QC_filtering(rse_gene_pups, 'rRNA_rate', 'Group')
p + theme(plot.margin=unit (c (2,4,2,4), 'cm'))
```
<div class="alert alert-info">
   Try the function yourself with different QC metrics and sample variables! 
</div>




**2.2 Explore sample-level effects**

```{r}
library("jaffelab")

## Obtain PCAs
pca_all <- prcomp(t(assays(rse_gene_filt)$logcounts))
 
## % of the variance explained by each PC
pca_vars <- getPcaVars(pca_all)
pca_vars_labs <- paste0(
    "PC", seq(along = pca_vars), ": ",
    pca_vars, "% Var Expl")

## Join PCs and samples' info 
pca_data <- cbind(pca_all$x, colData(rse_gene_filt))

sample_variables <- c('Group', 'Age', 'Sex', 'Pregnancy', 'plate', 'flowcell')
```


```{r}
library("ggpubr")

plot_PCAs <- function(sample_var) {
    
    ## Define sample colors depending on the sample variable
    if (sample_var == "Group") {
        colors <- c("Control" = "brown2", "Experimental" = "deepskyblue3")
    } else if (sample_var == "Age") {
        colors <- c("Adult" = "slateblue3", "Pup" = "yellow3")
    } else if (sample_var == "Sex") {
        colors <- c("F" = "hotpink1", "M" = "dodgerblue")
    } else if (sample_var == "Pregnancy") {
        colors <- c("Yes" = "darkorchid3", "No" = "darkolivegreen4")
    } else if (sample_var == "plate") {
        colors <- c("Plate1" = "darkorange", "Plate2" = "lightskyblue", "Plate3" = "deeppink1")
    } else if (sample_var == "flowcell") {
        colors <- c(
            "HKCG7DSXX" = "chartreuse2", "HKCMHDSXX" = "magenta", "HKCNKDSXX" = "turquoise3",
            "HKCTMDSXX" = "tomato"
        )
    }


    plot1 <- ggplot(as.data.frame(pca_data), aes(PC1, PC2, color = !! rlang::sym(sample_var))) +
        geom_point() +
        theme_bw() +
        scale_color_manual(values = colors)
    
    plot2 <- ggplot(as.data.frame(pca_data), aes(PC2, PC3, color = !! rlang::sym(sample_var))) +
        geom_point() +
        theme_bw() +
        scale_color_manual(values = colors)
    
    plot3 <- ggplot(as.data.frame(pca_data), aes(PC3, PC4, color = !! rlang::sym(sample_var))) +
        geom_point() +
        theme_bw() +
        scale_color_manual(values = colors)


    wplot <- ggarrange(plot1, plot2, plot3, ncol = 3, common.legend = TRUE, legend = "right")
    
    return(wplot)
}

pca_plots <- lapply(sample_variables, plot_PCAs)
ggarrange(plotlist = pca_plots, nrow = length(pca_plots))
```



**2.3 Explore gene-level effects**

Once the quality and the variability of the samples have been evaluated, the next step is to explore the differences in the expression of the genes themselves in the sample groups, or in other words, to quantify the contribution of the multiple sample variables of the study in the gene expression variation, which constitutes one of the fundamental challenges when analyzing complex RNA-seq datasets [3].

To determine which variables are the major drivers of expression variability, and importantly to define if the technical variability of RNA-seq data is low enough to study nicotine effects, we can implement an analysis of variance partition. `variancePartition` is a package that decomposes for each gene the expression variation into fractions of variance explained (FVE) by the sample variables of the experimental design of high-throughput genomics studies [3].


*2.3.1 Canonical Correlation Analysis*

Before the analysis itself, we need to measure the correlation between the sample variables. This is an important step because highly correlated variables can produce unstable estimates of the variance fractions and hinder the identification of the variables that really contribute to the expression variation. There are at least two problems with that:

* If two variables are correlated, we could incorrectly determine that one of them contributes to gene expression changes when in reality it was just correlated with a real contributory variable.
* The part of variance explained by a biologically relevant variable can be reduced by the apparent contributions of correlated variables, if for example, they contain very similar information.

Additionally, the analysis is better performed with simpler models, specially when we have a limited number of samples in the study. 
 
Hence, to drop such variables we must identify them first. Pearson correlation can be used for comparing continuous variables but the models can contain categorical variables as well, so in order to obtain the correlation between a continuous and a categorical variable, or between two categorical variables, we will perform a Canonical Correlation Analysis (CCA) with `canCorPairs()` that assesses the degree to which the variables co-vary and contain the same information. This function returns rho / sum(rho), the fraction of the maximum possible correlation. Note that CCA returns correlations values between 0 and 1 [4].



```{r message=FALSE, warning=FALSE}

library(variancePartition)
library(pheatmap)

#######################   Variance Partition Analysis   #######################

## Fraction of variation attributable to each variable after correcting for all other variables


## 1. Canonical Correlation Analysis (CCA)

## Asses the correlation between each pair of sample variables

## Plot heatmap of correlations
plot_CCA<- function(age){
    
    ## Data
    rse_gene <- eval(parse_expr(paste0('rse_gene_', age, '_qc')))

    
    ## Define variables to examine: remove those with single values
    ## For adults: all are females (so we drop 'Sex' variable)
    
    if (age == 'adults'){
        formula = ~ Group + Pregnancy + plate + flowcell + mitoRate + overallMapRate + totalAssignedGene + rRNA_rate + sum + detected + ERCCsumLogErr

    }
    ## For pups: none is pregnant (so 'Pregnancy' variable is not considered)
    else {
      formula = ~ Group + Sex + plate + flowcell + mitoRate + overallMapRate + totalAssignedGene + rRNA_rate + sum + detected + ERCCsumLogErr
    }

    ## Measure correlations
    C=canCorPairs(formula, colData(rse_gene))
    ## Heatmap
    pheatmap(
        C, ## data
        color = hcl.colors(50, "YlOrRd", rev = TRUE), ## color scale
        fontsize=8,  ## text size
        border_color = "black",  ## border color for heatmap cells
        cellwidth = unit(0.4, 'cm'),  ## height of cells
        cellheight = unit(0.4, 'cm') ## width of cells
    )

    return(C)
}


```


As you can see, and in agreement to what we observed in the previous correlation plots, for adults there's a strong correlation between  <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> mitoRate</span> </mark>  and <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> totalAssignedGene</span> </mark>; we already know it's a negative correlation. There's also a strong correlation between <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> plate</span> </mark> and <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> flowcell </span> </mark> and between <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> plate</span> </mark> and <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> overallMapRate</span> </mark>. 

```{r message=FALSE, warning=FALSE}
## Heatmap for adult samples
CCA_adults <- plot_CCA('adults')
```

In pups <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> overallMapRate</span> </mark> has a relationship with <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> rRNA_rate</span> </mark>, <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> plate</span> </mark> and <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> flowcell</span> </mark>.

```{r message=FALSE, warning=FALSE}
## Heatmap for pup samples
CCA_pups <- plot_CCA('pups')
```

Importantly, in both groups of age <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> Group</span> </mark> is not highly correlated with any other variable. This is desirable because if this variable, the one that separates experimental and control samples, were correlated with another, then its individual contribution to gene expression changes would be diminished, affecting the results of the differential expression analysis: we would obtain gene differences driven by other variables rather than <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> Group</span> </mark>. Also, <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> sum</span> </mark> and <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> detected</span> </mark> are correlated. 

Let's explore why these variables are correlated in the following plots. 

```{r message=FALSE, warning=FALSE}

## 1.1  Barplots/Boxplots/Scatterplots for each pair of correlated variables

corr_plots <- function(age, sample_var1, sample_var2, sample_color){

    ## Data
    rse_gene <- eval(parse_expr(paste("rse_gene", age, 'qc', sep="_")))
    CCA <- eval(parse_expr(paste0('CCA_', age)))

    ## Sample color by one variable
    colors <- list(
        "Group" = c('Control'='brown2', 'Experimental'='deepskyblue3'),
        "Age" = c('Adult'='slateblue3', 'Pup'='yellow3'),
        "Sex" = c('F'='hotpink1', 'M'='dodgerblue'),
        "Pregnancy" = c('Yes'='darkorchid3', 'No'='darkolivegreen4'),
        "plate" = c('Plate1'='darkorange', 'Plate2'='lightskyblue', 'Plate3'='deeppink1'),
        "flowcell" = c('HKCG7DSXX'='chartreuse2','HKCMHDSXX'='magenta', 'HKCNKDSXX'='turquoise3', 
                 'HKCTMDSXX'='tomato')
    )

    data <- colData(rse_gene)
    
    ## Barplots for categorical variable vs categorical variable
     if ( class(data[,sample_var1]) ==  "character" &  class(data[,sample_var2]) ==  "character") {
         
         ## y-axis label 
         if (sample_var2=='Pregnancy'){
             y_label <- paste('Number of samples from each ', sample_var2, ' group', sep='')
         }
         else{
             y_label <- paste('Number of samples from each ', sample_var2, sep='')
         }
         
         
         # Stacked barplot with counts for 2nd variable
         plot <- ggplot(data=as.data.frame(data), aes(x = !! rlang::sym(sample_var1),
                                                      fill = !! rlang::sym(sample_var2))) + 
                        geom_bar(position = "stack") + 
                        ## Colors by 2nd variable
                        scale_fill_manual(values = colors[[sample_var2]]) + 
                        ## Show sample counts on stacked bars
                        geom_text(aes(label = after_stat(count)), stat = "count", 
                                  position=position_stack(vjust = 0.5),  colour='gray20', size=3) +
                        theme_bw() +
                        labs(
                            subtitle = paste0("Corr: ", signif(CCA[sample_var1, sample_var2], digits=3)), 
                            y = y_label) +
                        theme(axis.title = element_text(size = (7)),
                              axis.text = element_text(size = (6)),
                              plot.subtitle = element_text(size = 7, color='gray40'),
                              legend.text = element_text(size=6),
                              legend.title = element_text(size=7))
     }
    
    
    ## Boxplots for categorical variable vs continuous variable 
    else if ( class(data[,sample_var1]) ==  "character" &  class(data[,sample_var2]) ==  "numeric") {
        plot <- ggplot(data = as.data.frame(data), mapping = aes(x = !! rlang::sym(sample_var1),
                                                                 y = !! rlang::sym(sample_var2),
                                                                 color = !! rlang::sym(sample_var1))) +
            geom_boxplot(size = 0.25, width=0.32, color='black', outlier.color = "#FFFFFFFF") +
            geom_jitter( width = 0.15, alpha = 1, size = 1) +
            stat_smooth (geom="line", alpha=0.6, size=0.4, span=0.3, method = lm, aes(group=1), color='orangered3') +
            scale_color_manual(values = colors[[sample_var1]]) +
            theme_bw() +
            guides(color="none") +
            labs(subtitle = paste0("Corr: ", signif(CCA[sample_var1, sample_var2], digits=3)), y = gsub('_', ' ', sample_var2), 
                 x = sample_var1) +
            theme(axis.title = element_text(size = (7)),
                  axis.text = element_text(size = (6)),
                  plot.subtitle = element_text(size = 7, color='gray40'),
                  legend.text = element_text(size=6),
                  legend.title = element_text(size=7))
    }

    
    ## Scatterplots for continuous variable vs continuous variable
    else if ( class(data[,sample_var1]) ==  "numeric" &  class(data[,sample_var2]) ==  "numeric") {

        plot <- ggplot(as.data.frame(data), aes(x=!! rlang::sym(sample_var1),
                                                y=!! rlang::sym(sample_var2),
                                                color=!! rlang::sym(sample_color))) +
            geom_point(size=2) +
            stat_smooth (geom="line", alpha=0.4, size=0.4, span=0.25, method = lm, color='orangered3') +
            ## Color by sample_color variable
            scale_color_manual(name=sample_color, values = colors[[sample_color]]) +
            theme_bw() +
            labs(subtitle = paste0("Corr: ", signif(CCA[sample_var1, sample_var2], digits=3)), y= gsub('_', ' ', sample_var2), x = gsub('_', ' ', sample_var1)) +
            theme(axis.title = element_text(size = (7)),
                  axis.text = element_text(size = (6)),
                  plot.subtitle = element_text(size = 7, color='gray40'),
                  legend.text = element_text(size=6),
                  legend.title = element_text(size=7))
    }

    return(plot)
}

```

As shown above and as previously demonstrated, the mitochondrial rate and the fraction of reads that mapped to genes are negatively correlated in adults but control and experimental samples are evenly distributed. 

```{r message=FALSE, warning=FALSE}
## Heatmap for pup samples
p <- corr_plots('adults', 'mitoRate', 'totalAssignedGene', 'Group')
p + theme(plot.margin=unit (c(2,4,2,4), 'cm'))
```
Although not expected, the flowcell and the plate of the adult samples were correlated, but that is due to the fact that all samples from the first flowcell (HKCG7DSXX) were in the 1st plate, and almost all samples from the second flowcell were in the 3rd plate.  

```{r message=FALSE, warning=FALSE}
## Heatmap for pup samples
p <- corr_plots('adults', 'flowcell', 'plate',  NULL)
p + theme(plot.margin=unit (c(1.5,4.5,1.5,4.5), 'cm'))
```
Plate was also slightly correlated with the overall mapping rate, but if we look closely, the trend is given by the plate 1 samples that have lower rates; the rates of samples from the 2nd and 3rd plates are similar. 

```{r message=FALSE, warning=FALSE}
## Heatmap for pup samples
p <- corr_plots('adults', 'plate', 'overallMapRate', NULL)
p + theme(plot.margin=unit (c(2,5.3,2,5.3), 'cm'))
```

We also appreciated and described already that the library size and the number of detected of genes are correlated in pups and adults. Note however that in pups control samples have bigger numbers of expressed genes. 
```{r message=FALSE, warning=FALSE}
## Heatmap for pup samples
p <- corr_plots('adults', 'sum', 'detected', 'Group')
p + theme(plot.margin=unit (c(2,4,2,4), 'cm'))

## Heatmap for pup samples
p <- corr_plots('pups', 'sum', 'detected', 'Group')
p + theme(plot.margin=unit (c(2,4,2,4), 'cm'))
```

In pups, the overall mapping rate was slightly and negatively correlated with the ribosomal rate, but again controls and experimental samples are equally distributed and the <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> rRNA_rate</span> </mark> is pretty small for all of them. 

```{r message=FALSE, warning=FALSE}
## Heatmap for pup samples
p <- corr_plots('pups', 'rRNA_rate', 'overallMapRate', 'Group')
p + theme(plot.margin=unit (c(2,4,2,4), 'cm'))
```

Overall mapping rate is also positively correlated with plate and flowcell, relationships mainly given by samples from the 1st plate with the lowest <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> overallMapRate</span> </mark>, most of which were also in the 1st flowcell.

```{r message=FALSE, warning=FALSE}
## Heatmap for pup samples
p <- corr_plots('pups', 'plate', 'overallMapRate', NULL)
p + theme(plot.margin=unit (c(2,5.3,2,5.3), 'cm'))
```

```{r message=FALSE, warning=FALSE}
## Heatmap for pup samples
p <- corr_plots('pups', 'flowcell', 'overallMapRate', NULL)
p + theme(plot.margin=unit (c(2,5.3,2,5.3), 'cm'))
```



<div class="alert alert-info">
Now look at the following plots. **Why is it important that experimental and control samples are distributed throughout all the plates and flowcells?** 
```{r message=FALSE, warning=FALSE}

p1 <- corr_plots('adults', 'Group', 'plate', NULL)
p2 <- corr_plots('pups', 'Group', 'plate', NULL)
p3 <- corr_plots('adults', 'Group', 'flowcell', NULL)
p4 <- corr_plots('pups', 'Group', 'flowcell', NULL)


plots <- plot_grid(p1, p2, p3, p4, ncol=2)
plots + theme(plot.margin=unit (c(1,2.5,1,2.5), 'cm'))
```
***Hint***: What would happen if all experimental samples were in one plate or flowcell and controls in another?
</div>

After identifying which variables are correlated and exploring the metrics of control and experimental samples the next step is to determine which of these variables must be removed. How do we discern which of the correlated variables to keep and which to drop? As recommended in the `variancePartition` user’s guide [4], initially we can fit a linear model for each gene taking all sample variables and then define which ones explain a higher percentage of variance in many genes. But first let's review how `variancePartition` works. 



*2.3.2 Fit model and extract fraction of variance explained*

Briefly, `variancePartition` fits a linear model for each gene separately and `calcVarPart()` computes the fraction of variance in gene expression that is explained by each variable of the study design, plus the residual variation. The effect of each variable is assessed while jointly accounting for all others [3].

Basically what it does is calculate the data variance given by each variable and that of the total model fit, summarizing the contribution of each variable in terms of the fraction of variation explained (FVE). Since it calculates the fraction of total variation attributable to each aspect of the study design, these fractions naturally sum to 1 [3].

`variancePartition` fits two types of models:

1. **Linear mixed model (LMM)** where all categorical variables are modeled as **random effects** and all continuous variables are **fixed effects**. The function `lmer()` from `lme4` is used to fit this model.

```r
## Fit LMM specifying the existence of random effects with '(1| )' 
fit <- lmer(expr ~ a + b + (1|c), data=data)
```

2. Fixed effects model, which is basically the **standard linear model (LM)**, where all variables are modeled as fixed effects. The function `lm()` is used to fit this model.

```r
## Fit LM modeling all variables as fixed effects
fit <- lm(expr ~ a + b + c, data=data)
```

In our case, the function will be modeled a mixed model since we have both effects.

<p class="comment">
❓***What are random and fixed effects? How to determine if a variable is one or the other?*** **Categorical variables** are usually modeled as **random effects**, i.e., traits such as the batch, sex, flowcell, plate, individual, variables 'randomly chosen or selected from a population' and whose specific levels are not of particular interest, only the grouping of the samples by those variables. These are control variables/factors that vary randomly across individuals or groups and we use them because we must control for these effects. Think of them as having different effects on gene expression (the dependent variable) depending on their values. 
**Continuous variables** must be modeled as **fixed effects**; they cannot be modeled as random effects. These correspond to variables that can be measured somehow and whose levels are themselves of interest (the QC metrics, for instance); these effects would be the same for all genes. 
</p>

<p class="comment">
❓***Why is this effect distinction important?***
Because when we have clustered data, like gene expression values grouped by sex, plate, etc. we are violating the relevant assumption of independence, making an incorrect inference when using a **general linear model (GLM)**. If we have clustered data where the variables' values have distinct effects on gene expression, we must work with an extension of GLM, with the **linear mixed model (LMM)** that contains a mix of both fixed and random effects [5].
</p>


**Linear mixed model framework**

<aside>
1️⃣ For a single gene, after fitting the linear model we obtain the predicted expression of the genes given by the predicted coefficients of the variables:

$\hat y =\sum_{1}^j\beta_{j}X_{j} + \sum_{1}^k\alpha_{k}Z_{k}$  ← the expression of a gene across all samples is given by the samples' values in the $j$ fixed effects and $k$ random effects.  So basically the gene expression in the sample $i$ is given by $\hat y_i =\sum_{1}^j\beta_{j}X_{ij} + \sum_{1}^k\alpha_{k}Z_{ik}$.

Then $y=\hat y+\epsilon$, which means that the real (observed) expression value is given by the predicted value plus an error term ($\epsilon$), also called noise or residual:

$$y =\sum_{1}^j\beta_{j}X_{j} + \sum_{1}^k\alpha_{k}Z_{k} + \epsilon$$  

- $X_j$ is the vector of the values of the samples in the $j$th fixed effect.
- $\beta_j$ is the predicted coefficient of the fixed effect $j$.

$$
X_j\beta_j=  \ \ _{n \ \ samples}\stackrel{j^{th}\ \ fixed\ \ effect }{\begin{bmatrix} X_{1j} \\ ... \\ X_{(n-1)j} \\ X_{nj}  \end{bmatrix}}\beta_j = {\begin{bmatrix}  X_{1j}\beta_j \\ ... \\ X_{(n-1)j}\beta_j \\ X_{nj} \beta_j \end{bmatrix}}
$$

- $Z_k$ is the vector of values the samples have for the $k$th random effect.
- $\alpha_k$ is the predicted coefficient of the random effect $k$. These are drawn from a normal distribution   $∼N(0, \sigma_{\alpha_k}^2 )$.

$$
Z_k\alpha_k= \ \ _{n \ \ samples}\stackrel{{k^{th}\ \ random\ \ effect }}{\begin{bmatrix}  Z_{1k} \\ ... \\ Z_{(n-1)k}\\ Z_{nk} \end{bmatrix}}\alpha_k = {\begin{bmatrix} Z_{1k}\alpha_k \\ ... \\ Z_{(n-1)k}\alpha_k\\ Z_{nk}\alpha_k \end{bmatrix}}
$$

$\epsilon$ is the noise term which is  $y-\hat y$, the difference between the observed and predicted expression and is also drawn from a  $∼N(0, \sigma_{\epsilon}^2 )$.

Expanding, 

$$
y= {\begin{bmatrix}  X_{11}\beta_1 \\ ... \\ X_{(n-1)1}\beta_1 \\ X_{n1} \beta_1 \end{bmatrix}} + {\begin{bmatrix}  X_{12}\beta_2 \\ ... \\ X_{(n-1)2}\beta_2 \\ X_{n2} \beta_2 \end{bmatrix}} + ... +{\begin{bmatrix}  X_{1j}\beta_j \\ ... \\ X_{(n-1)j}\beta_j \\ X_{nj} \beta_j \end{bmatrix}} + {\begin{bmatrix} Z_{11}\alpha_1 \\ ... \\ Z_{(n-1)1}\alpha_1\\ Z_{n1}\alpha_1 \end{bmatrix}} +{\begin{bmatrix} Z_{12}\alpha_2 \\ ... \\ Z_{(n-1)2}\alpha_2\\ Z_{n2}\alpha_2 \end{bmatrix}} + ... + {\begin{bmatrix} Z_{1k}\alpha_k \\ ... \\ Z_{(n-1)k}\alpha_k\\ Z_{nk}\alpha_k \end{bmatrix}} + {\begin{bmatrix} \epsilon_1 \\ ... \\ \epsilon_{(n-1)} \\ \epsilon_n \end{bmatrix}}
$$

All parameters are estimated with maximum likelihood, the default method in the `variancePartition` software when random effects are specified because it performs best in simulations.
</aside>




<aside>
2️⃣ Then, `calcVarPart()` computes for each fixed effect

$\sum_{i=1}^n(\beta_{j}X_{ij}-\bar{\beta_{j}X_{j}})^2=var(\beta_{j}X_{j})(n-1)$, the squared sum of the predicted expression values of a gene in the samples ($i=1, ..., n$), only taking into account the variable $j$ in the regression model: $\hat y = \beta_{j}X_{j}$ .

Each of these squared sums is scaled by the total sum of squares of the predicted values taking all fixed effects: $var(\hat y)\times (n-1)$ and by the sum of the squared sums for all variables: $\sum_1^j var(\beta_jX_j)(n-1)$.

- To simplify let's just take the variance of the $j$th fixed effect because the squared sum is proportional to it:  <mark style= "background-color: #FEF5E7">$\sigma_{\beta_j}^2=var(X_j{\beta_j})$</mark>

For random effects the variances are computed by variance component estimates with `VarCorr()` from `nmle`:

- The variance of the $k$th random effect is  <mark style= "background-color: #D6EAF8">$\sigma_{\alpha_k}^2=var(Z_k{\alpha_k})$</mark>

After that it calculates the total variance of the expression values $\sum_{i=1}^n(y_i - \bar { y})^2=var(y)(n-1)$, where $y_i = \sum_{1}^j\beta_{j}X_{ij} + \sum_{1}^k\alpha_{k}Z_{ik} + \epsilon_i$, considering all variables in the model and the error:

- The total variance is:

$var(y)= \sigma_{Total}^2= var(X_{1}\beta_1)+var(X_{2}\beta_2)+...+var(X_{j}\beta_j)+var(Z_{1}\alpha_1)+var(Z_{2}\alpha_2)+...+var(Z_{k}\alpha_k)+var(\epsilon)=$
$\sum_1^jvar(X_j\beta_j)+\sum_1^kvar(Z_k\alpha_k)+var(\epsilon)=$

<mark style= "background-color: #F9EBEA">$\sigma_{Total}^2=\sum_1^j{ \sigma_{\beta_j}^2} + \sum_1^k{ \sigma_{\alpha_k}^2} + \sigma_{\epsilon}^2$</mark>

</aside>



<aside>
3️⃣ Finally, it computes:

- The fraction of the total data variance explained by the $j$th fixed effect is  <mark style= "background-color: #FEF5E7">$\sigma_{\beta_j}^2$</mark> /   <mark style= "background-color: #F9EBEA">$\sigma_{Total}^2$</mark>
- The fraction of the total data variance explained by the $k$th random effect is  <mark style= "background-color: #D6EAF8">$\sigma_{\alpha_k}^2$</mark> /    <mark style= "background-color: #F9EBEA">$\sigma_{Total}^2$</mark>

Note that $y=\hat y+\epsilon$ because the expression can't be completely described by a straight line, so not all the variation of  $y$ can be explained by the variation of the sample variables, instead $var(y)=var(\hat y)+var(\epsilon)=var(\hat y) + \sigma_{\epsilon}^2$, where  $\sigma_{\epsilon}^2=\sum_{i=1}^n(\hat y_i -y_i)^2/n-1$.

- The residual variance is  <mark style= "background-color: #E8DAEF">$\sigma_{\epsilon}^2$</mark> /    <mark style= "background-color: #F9EBEA">$\sigma_{Total}^2$</mark> ; this is the variance that the model couldn't explain (we don’t know where these residuals come from).
</aside>



👉🏼Source code of `calcVarPart()` [here](https://rdrr.io/github/GabrielHoffman/variancePartition/src/R/calcVarPart.R#sym-getVarianceComponents).


Once we understood what it computes, we can use `variancePartition` to quantify the FVE of each variable. 

```{r message=FALSE, warning=FALSE}
## 2. Fit model

## Fit a linear mixed model (LMM) that takes continuous variables as fixed effects and categorical variables as random effects

varPartAnalysis <- function(age, formula){

    RSE<-eval(parse_expr(paste("rse_gene", age, 'qc', sep="_")))

    ## Ignore genes with variance 0
    genes_var_zero<-which(apply(assays(RSE)$logcounts, 1, var)==0)
    if (length(genes_var_zero)>0){
        RSE <- RSE[-genes_var_zero, ]
    }

    ## Loop over each gene to fit model and extract variance explained by each variable
    varPart <- fitExtractVarPartModel(assays(RSE)$logcounts, formula, colData(RSE))

    # Sort variables by median fraction of variance explained
    vp <- sortCols(varPart)
    p <- plotVarPart(vp)
    
    return(list(p, vp))
}

```


```{r message=FALSE, warning=FALSE, eval=FALSE}

## Violin plots

#####  Model with all variables  #####

## Adults 
## Define variables; random effects indicated with (1| )
formula <-  ~ (1|Group) + (1|Pregnancy) + (1|plate) + (1|flowcell) + mitoRate + overallMapRate + 
                totalAssignedGene + rRNA_rate + sum + detected + ERCCsumLogErr
plot <- varPartAnalysis('adults', formula)[[1]]
plot + theme(plot.margin=unit (c(1,1,1,1), 'cm'), 
              axis.text.x = element_text(size = (7)),
              axis.text.y = element_text(size = (7.5)))
```

As presented above, in adults we can notice that <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> totalAssignedGene</span> </mark> has a larger mean FVE than <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> mitoRate</span> </mark> so we keep the former. Same reason to remove <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> plate</span> </mark> that is correlated with <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> flowcell</span> </mark> and with <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> overallMapRate</span> </mark>, and to drop <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> sum</span> </mark> that goes after <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> detected</span> </mark>.

```{r message=FALSE, warning=FALSE, eval=FALSE}

#####  Model without correlated variables  #####

## Adult plots without mitoRate, plate and sum
formula <-  ~ (1|Group) + (1|Pregnancy) + (1|flowcell) + overallMapRate + totalAssignedGene + rRNA_rate + detected + ERCCsumLogErr
varPart <- varPartAnalysis('adults', formula)
varPart_data_adults <- varPart[[2]]
plot <- varPart[[1]]
plot + theme(plot.margin=unit (c(1,1,1,1), 'cm'), 
              axis.text.x = element_text(size = (7)),
              axis.text.y = element_text(size = (7.5)))
```
Notwithstanding, in this new reduced model Group contribution doesn't increment much in comparison with the complete model and also note that the % of variance explained by the residuals, i.e., the % of gene expression variance that the model couldn't explain, increments in this model compared to the previous; by removing independent variables to a regression equation, we can explain less of the variance of the dependent variable [5]. That's the price to pay when dropping variables but it is convenient in this case since we don't have many samples for the model to determine their real unique contributions.  

In pups, based on the model with all variables, again <mark style= "background-color: #DFF0FE"> <span style="font-family: monospace"> sum</span> </mark> must be removed, as well as <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> rRNA_rate</span> </mark> and <mark style= "background-color: #FAECF8"> <span style="font-family: monospace"> plate</span> </mark> that are correlated with <mark style= "background-color: #FCF3CF"> <span style="font-family: monospace"> overallMapRate</span> </mark>.

```{r message=FALSE, warning=FALSE, eval=FALSE}
#####  Model with all variables  #####

## Pups
formula <-  ~ (1|Group) + (1|Sex) + (1|plate) + (1|flowcell) + mitoRate + overallMapRate + 
                totalAssignedGene + rRNA_rate + sum + detected + ERCCsumLogErr
plot <- varPartAnalysis('pups', formula)[[1]]
plot + theme(plot.margin=unit (c(1,1,1,1), 'cm'), 
              axis.text.x = element_text(size = (7)),
              axis.text.y = element_text(size = (7.5)))
```

Without correlated variables, Group's contribution increases but so does the residual source.
```{r message=FALSE, warning=FALSE, eval=FALSE}
#####  Model without correlated variables  #####

## Pup plots without sum, rRNA_rate and plate
formula <-  ~ (1|Group) + (1|Sex) + (1|flowcell) + mitoRate + overallMapRate + totalAssignedGene + detected + ERCCsumLogErr
varPart <- varPartAnalysis('pups', formula)
varPart_data_pups <- varPart[[2]]
plot <- varPart[[1]]
plot + theme(plot.margin=unit (c(1,1,1,1), 'cm'), 
              axis.text.x = element_text(size = (7)),
              axis.text.y = element_text(size = (7.5)))
```

But what does it mean that a variable explains a high percentage of the expression variation of a gene? Look at the following plots and try the function with other variables to observe the relationships that exists between the gene expression values in the samples and samples' variables. 

*2.3.3 Examine expression of most affected genes by each sample variable*

In the following plots you can appreciate the expression values in the samples of the most affected genes by a certain variable (i.e., the genes for which the variable explains the highest percentages of variance) vs. the values of the samples for the same variable. In simple words, each dot is a sample, in the y-axis the expression of the gene in each sample and in the x-axis the samples' values for the variable. The idea is to observe the strong relationships that exists for the sample variables and the gene expression of such affected genes, which ends up causing the variables to explain high percentages of gene expression variation and which obligate us to adjust for them the models of DEA so that we don´t get DEGs driven by sample quality differences.  

```{r message=FALSE, warning=FALSE}

## Plot of gene expression lognorm counts vs. sample variable
plot_gene_expr <- function(age, sample_var, gene_id){

   rse_gene <- eval(parse_expr(paste("rse_gene", age, 'qc', sep="_")))
   varPart_data <- eval(parse_expr(paste0('varPart_data_', age)))

   colors <- list(
     "Group" = c('Control'='brown2', 'Experimental'='deepskyblue3'),
     "Age" = c('Adult'='slateblue3', 'Pup'='yellow3'),
     "Sex" = c('F'='hotpink1', 'M'='dodgerblue'),
     "Pregnancy" = c('Yes'='darkorchid3', 'No'='darkolivegreen4'),
     "plate" = c('Plate1'='darkorange', 'Plate2'='lightskyblue', 'Plate3'='deeppink1'),
     "flowcell" = c('HKCG7DSXX'='chartreuse2','HKCMHDSXX'='magenta', 'HKCNKDSXX'='turquoise3', 
                 'HKCTMDSXX'='tomato')
    )

    ## Lognorm counts of the gene across samples
    data <- colData(rse_gene)
    data$gene_expr <- assays(rse_gene)$logcounts[gene_id,]

    ## Percentage of variance explained by the variable
    percentage <- 100*signif(varPart_data[gene_id, sample_var], digits=3)

    ## Boxplots for discrete variables
    if (class(data[,sample_var])=='character') {
        plot <- ggplot(data = as.data.frame(data), mapping = aes(x = !! rlang::sym(sample_var),
                                                                 y = gene_expr, color = !! rlang::sym(sample_var))) +
            geom_boxplot(size = 0.25, width=0.32, color='black', outlier.color = "#FFFFFFFF") +
            geom_jitter(width = 0.15, alpha = 1, size = 1) +
            stat_smooth (geom="line", alpha=0.6, size=0.4, span=0.3, method = lm, aes(group=1), color='orangered3') +
            scale_color_manual(values = colors[[sample_var]]) +
            theme_bw() +
            guides(color="none") +
            labs(title = gene_id,
                 subtitle = paste0("Variance explained: ", percentage, '%'),
                 y= 'lognorm counts', x = sample_var) +
            theme(axis.title = element_text(size = (7)),
                  axis.text = element_text(size = (6)),
                  plot.title = element_text(hjust=0.5, size=7.5, face="bold"),
                  plot.subtitle = element_text(size = 7, color='gray40'),
                  legend.text = element_text(size=6),
                  legend.title = element_text(size=7))
    }

    ## Scatterplots for continuous variables
    else {
        colors <- c('mitoRate'='khaki3', 'overallMapRate'='turquoise', 'totalAssignedGene'='plum2', 'rRNA_rate'='orange3', 
                    'sum'='palegreen3', 'detected'='skyblue2', 'ERCCsumLogErr'='slateblue1')

        plot <- ggplot(as.data.frame(data), aes(x=eval(parse_expr(sample_var)), y=gene_expr)) +
            geom_point( color=colors[[sample_var]], size=2) +
            stat_smooth (geom="line", alpha=0.4, size=0.4, span=0.25, method = lm, color='orangered3') +
            theme_bw() +
            guides(color="none") +
            labs(title = gene_id,
                 subtitle = paste0("Variance explained: ", percentage, '%'),
                 y= 'lognorm counts', x = gsub('_', ' ', sample_var)) +
            theme(plot.margin=unit (c (0.4,0.1,0.4,0.1), 'cm'),
                  axis.title = element_text(size = (7)),
                  axis.text = element_text(size = (6)),
                  plot.title = element_text(hjust=0.5, size=7.5, face="bold"),
                  plot.subtitle = element_text(size = 7, color='gray40'),
                  legend.text = element_text(size=6),
                  legend.title = element_text(size=7))
    }

    return(plot)
}
```



```{r message=FALSE, warning=FALSE}

## Function to plot gene expression vs sample variable data for top 3 most affected genes

plot_gene_expr_sample <- function(age, sample_var){
    
   rse_gene <- eval(parse_expr(paste("rse_gene", age, 'qc', sep="_")))
   varPart_data <- eval(parse_expr(paste0('varPart_data_', age)))
   
   ## Top 3 genes most affected by sample variable
   affected_genes <- rownames(varPart_data[order(varPart_data[,sample_var], decreasing =  TRUE),][1:3,])
   
   ## Plots
   plots <- list()
   for (i in 1:length(affected_genes)){
      plots[[i]] <- plot_gene_expr(age, sample_var, affected_genes[i])
   }
   plot_grid(plots[[1]], plots[[2]], plots[[3]], ncol=3)
}

```


```{r message=FALSE, warning=FALSE, eval=FALSE}
## Adults

## Plots for top affected genes by 'totalAssignedGene'
plots <- plot_gene_expr_sample('adults', 'totalAssignedGene')
plots + theme(plot.margin=unit (c(3,1,2,3), 'cm'))

## Plots for top affected genes by 'overallMapRate'
plots <- plot_gene_expr_sample('adults', 'overallMapRate')
plots + theme(plot.margin=unit (c(3,1,2,3), 'cm'))

## Plots for top affected genes by 'Group'
plots <- plot_gene_expr_sample('adults', 'Group')
plots + theme(plot.margin=unit (c(3,1,2,3), 'cm'))
```

```{r message=FALSE, warning=FALSE, eval=FALSE}
## Pups

## Plots for top affected genes by 'overallMapRate'
plots <- plot_gene_expr_sample('pups', 'overallMapRate')
plots + theme(plot.margin=unit (c(3,1,2,3), 'cm'))

## Plots for top affected genes by 'totalAssignedGene'
plots <- plot_gene_expr_sample('pups', 'totalAssignedGene')
plots + theme(plot.margin=unit (c(3,1,2,3), 'cm'))

## Plots for top affected genes by 'Group'
plots <- plot_gene_expr_sample('pups', 'Group')
plots + theme(plot.margin=unit (c(3,1,2,3), 'cm'))
```

At this point we have extracted normalized counts and filtered genes, we have taken out poor-quality samples and we have decided which sample variables to retain for future analyses. Now we are ready for the DEA!


### 3. Differential Expression Analysis {-}

**3.1 Modeling**
ToDO Daianna & Renee

**3.2 Comparisons**
ToDO Daianna & Renee





### 4. DEG visualization {-}
ToDO  Renee 







### References {-}
1. Evans, C., Hardin, J., & Stoebel, D. M. (2018). Selecting between-sample RNA-Seq normalization methods from the perspective of their assumptions. *Briefings in bioinformatics*, 19(5), 776-792.

2. Bedre, R. (2023). Gene expression units explained: RPM, RPKM, FPKM, TPM, *DESeq*, TMM, SCnorm, GeTMM, and ComBat-Seq. Web site: https://www.reneshbedre.com/blog/expression_units.html

3. Hoffman, G. E., & Schadt, E. E. (2016). variancePartition: interpreting drivers of variation in complex gene expression studies. BMC bioinformatics, 17(1), 1-13.

4. Hoffman, G. (2022). variancePartition: Quantifying and interpreting drivers of variation in multilevel gene expression experiments.

5. van den Berg, S. M. (2022). Analysing data using linear models. Web site: https://bookdown.org/pingapang9/linear_models_bookdown/



# Reproducibility

Code for creating the vignette

```{r createVignette, eval=FALSE}
## Create the vignette
library("rmarkdown")
system.time(render("SPEAQeasyWorkshop2023.Rmd"))

## Extract the R code
library("knitr")
knit("SPEAQeasyWorkshop2023.Rmd", tangle = TRUE)
```


Date the vignette was generated.

```{r reproduce1, echo=FALSE}
## Date the vignette was generated
Sys.time()
```

Wallclock time spent generating the vignette.

```{r reproduce2, echo=FALSE}
## Processing time in seconds
totalTime <- diff(c(startTime, Sys.time()))
round(totalTime, digits = 3)
```

`R` session information.

```{r reproduce3, echo=FALSE}
## Session info
library("sessioninfo")
options(width = 120)
session_info()
```

# Bibliography

This vignette was generated using `r Biocpkg('BiocStyle')` `r Citep(bib[['BiocStyle']])`, `r CRANpkg('knitr')` `r Citep(bib[['knitr']])` and `r CRANpkg('rmarkdown')` `r Citep(bib[['rmarkdown']])` running behind the scenes.

Citations made with `r CRANpkg('RefManageR')` `r Citep(bib[['RefManageR']])`.

```{r vignetteBiblio, results = 'asis', echo = FALSE, warning = FALSE, message = FALSE}
## Print bibliography
PrintBibliography(bib, .opts = list(hyperlink = "to.doc", style = "html"))
```
