<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SPEAQeasy: a Scalable Pipeline for Expression Analysis and Quantification, a demonstration of its usage on real data • SPEAQeasyWorkshop2023</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="SPEAQeasy: a Scalable Pipeline for Expression Analysis and Quantification, a demonstration of its usage on real data">
<meta property="og:description" content="SPEAQeasyWorkshop2023">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPEAQeasyWorkshop2023</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/SPEAQeasyWorkshop2023.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>SPEAQeasy: a Scalable Pipeline for Expression
Analysis and Quantification, a demonstration of its usage on real
data</h1>
                        <h4 data-toc-skip class="author">Daianna
Gonzalez-Padilla</h4>
            <address class="author_afil">
      Lieber Institute for Brain
Development<br><a class="author_email" href="mailto:#"></a><a href="mailto:glezdaianna@gmail.com" class="email">glezdaianna@gmail.com</a>
      </address>
                              <h4 data-toc-skip class="author">Renee
Garcia-Flores</h4>
            <address class="author_afil">
      Lieber Institute for Brain
Development<br><a class="author_email" href="mailto:#"></a><a href="mailto:renee.garciaflores@gmail.com" class="email">renee.garciaflores@gmail.com</a>
      </address>
                              <h4 data-toc-skip class="author">Nicholas
J. Eagles</h4>
            <address class="author_afil">
      Lieber Institute for Brain
Development<br><a class="author_email" href="mailto:#"></a><a href="mailto:nick.eagles@libd.org" class="email">nick.eagles@libd.org</a>
      </address>
                              <h4 data-toc-skip class="author">Leonardo
Collado-Torres</h4>
            <address class="author_afil">
      Lieber Institute for Brain DevelopmentCenter for Computational
Biology, Johns Hopkins
University<br><a class="author_email" href="mailto:#"></a><a href="mailto:lcolladotor@gmail.com" class="email">lcolladotor@gmail.com</a>
      </address>
                  
            <h4 data-toc-skip class="date">9 June 2023</h4>
      
      
      <div class="hidden name"><code>SPEAQeasyWorkshop2023.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p><strong>TODO</strong> This workshop aims to present the
<code>SPEAQeasy</code> (Eagles, Burke, Leonard, Barry, Stolz, Huuki,
Phan, Serrato, Gutiérrez-Millán, Aguilar-Ordoñez, Jaffe, and
Collado-Torres, 2021) RNA-seq processing pipeline, demonstrating how its
outputs can be used in Differential Expression Analyses (DEA) and the
type of results it enables to obtain using other Bioconductor R packages
such as <em><a href="https://bioconductor.org/packages/3.17/limma" class="external-link">limma</a></em>,
<em><a href="https://bioconductor.org/packages/3.17/edgeR" class="external-link">edgeR</a></em>, and
<em><a href="https://bioconductor.org/packages/3.17/clusterProfiler" class="external-link">clusterProfiler</a></em>.</p>
<p>Briefly describe what is it , etc …</p>
<div class="section level3">
<h3 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h3>
<p><strong>TODO</strong> In this demo, participants will be able to
understand what the pipeline does and what it returns through a
Differential Expression Analysis performed on a SPEAQeasy dataset from
the <code>smokingMouse</code> (Gonzalez-Padilla and Collado-Torres,
2023) Bioconductor package which contains gene expression data and
sample information from an RNA-sequencing experiment in mice.</p>
</div>
<div class="section level3">
<h3 id="instructors">Instructors<a class="anchor" aria-label="anchor" href="#instructors"></a>
</h3>
<ul>
<li><a href="https://daianna21.github.io" class="external-link">Daianna Gonzalez</a></li>
<li><a href="https://reneegf.github.io" class="external-link">Renee Garcia</a></li>
<li><a href="https://nick-eagles.github.io" class="external-link">Nick Eagles</a></li>
</ul>
</div>
<div class="section level3">
<h3 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h3>
<ul>
<li>General familiarity with bulk RNA-seq data.</li>
<li>Basic knowledge of differential expression.</li>
<li>Basic experience handling <em><a href="https://bioconductor.org/packages/3.17/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>
and <em><a href="https://bioconductor.org/packages/3.17/GenomicRanges" class="external-link">GenomicRanges</a></em>
packages.</li>
<li>Previous review of <a href="http://research.libd.org/smokingMouse/" class="external-link uri">http://research.libd.org/smokingMouse/</a> for data
explanation. (Optional)</li>
</ul>
</div>
<div class="section level3">
<h3 id="participation">Participation<a class="anchor" aria-label="anchor" href="#participation"></a>
</h3>
<p><strong>TODO</strong> Participants are expected to follow</p>
</div>
<div class="section level3">
<h3 id="r-bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r-bioconductor-packages-used"></a>
</h3>
<p><strong>TODO</strong> List any <em>R</em> / <em>Bioconductor</em>
packages that will be explicitly covered.</p>
</div>
<div class="section level3">
<h3 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h3>
<p><strong>TODO</strong></p>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction to SPEAQeasy</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Data overview</td>
<td>20m</td>
</tr>
<tr class="odd">
<td>Differential Expression Analysis</td>
<td>55m</td>
</tr>
</tbody>
</table>
<p>Total: a 90 minute session.</p>
</div>
<div class="section level3">
<h3 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h3>
<p><strong>TODO</strong> List “big picture” student-centered workshop
goals and learning objectives. Learning goals and objectives are
related, but not the same thing. These goals and objectives will help
some people to decide whether to attend the conference for training
purposes, so please make these as precise and accurate as possible.</p>
<p><em>Learning goals</em> are high-level descriptions of what
participants will learn and be able to do after the workshop is over.
<em>Learning objectives</em>, on the other hand, describe in very
specific and measurable terms specific skills or knowledge attained. The
<a href="#bloom">Bloom’s Taxonomy</a> may be a useful framework for
defining and describing your goals and objectives, although there are
others.</p>
<div class="section level4">
<h4 id="learning-goals">Learning goals<a class="anchor" aria-label="anchor" href="#learning-goals"></a>
</h4>
<p>Some examples:</p>
<ul>
<li>describe how to…</li>
<li>identify methods for…</li>
<li>understand the difference between…</li>
</ul>
</div>
<div class="section level4">
<h4 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h4>
<ul>
<li>analyze xyz data to produce…</li>
<li>create xyz plots</li>
<li>evaluate xyz data for artifacts</li>
</ul>
</div>
</div>
</div>
<div class="section level2">
<h2 id="workshop">Workshop<a class="anchor" aria-label="anchor" href="#workshop"></a>
</h2>
<p><strong>TODO</strong> Divide the workshop into sections
(<code>## A Section</code>). Include fully-evaluated <em>R</em> code
chunks. Develop exercises and solutions, and anticipate that your
audience will walk through the code with you, or work on the code
idependently – do not be too ambitious in the material that you
present.</p>
<div class="section level3">
<h3 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h3>
<p><strong>TODO</strong></p>
<div class="section level4">
<h4 id="speaqeasy-overview">SPEAQeasy Overview<a class="anchor" aria-label="anchor" href="#speaqeasy-overview"></a>
</h4>
<p><strong>TODO</strong></p>
</div>
</div>
<div class="section level3">
<h3 id="data-overview-and-download">Data overview and download<a class="anchor" aria-label="anchor" href="#data-overview-and-download"></a>
</h3>
<p>The dataset that will be used for the DEA can be downloaded from the
<code>smokingMouse</code> package. Visit <a href="http://research.libd.org/smokingMouse/" class="external-link">here</a> for more details
of this package.</p>
<img src="figs/Data_overview_experiments.jpg" width="750px"><figcaption style="color: gray; line-height: 0.85; text-align: justify"><p><font size="-2"><b>Figure 1</b>: <b>Experimental design of the study.
</b> The data was obtained from an experiment in which <b>A)</b> 6
pregnant dams and 17 non-pregnant female adult mice were administered
nicotine by intraperitoneal injection (IP; n=12) or controls (n=11). A
total of 42 pups were born to pregnant dams: 19 were born to mice that
were administered nicotine and 23 to control mice. Samples from frontal
cortices of P0 pups and adults were obtained. <b>B)</b> RNA was
extracted from those samples and then RNA-seq libraries were prepared
and sequenced to obtain the expression counts of the genes. At the end,
a dataset of 65 samples (either from adult or pup brain) and 55,401
genes was generated. </font></p>
</figcaption><p>The data reside in a <code>RangedSummarizedExperiment</code> (RSE)
object, containing the following assays:</p>
<ul>
<li>
<code>counts</code>: original read counts of the 55,401 mouse genes
across 208 samples (inlcuding the 65 nicotine samples).</li>
<li>
<code>logcounts</code>: normalized and scaled counts (<span class="math inline">\(log_2(CPM + 0.5)\)</span>) of the same genes
across the same samples; normalization was carried out applying TMM
method with <code>cpm(calcNormFactors())</code> of <em><a href="https://bioconductor.org/packages/3.17/edgeR" class="external-link">edgeR</a></em>.</li>
</ul>
<p>The same RSE contains the sample information in
<code>colData(RSE)</code> and the gene information in
<code>rowData(RSE)</code>. <mark style="background-color: #FCF3CF">
Yellow </mark> variables correspond to SPEAQeasy outputs that are going
to be used in downstream analyses.</p>
<p><mark style="background-color: #FAECF8"> Pink </mark> variables are
specific of the study, such as sample metadata and some others
containing additional information of the genes.</p>
<p><mark style="background-color: #DFF0FE"> Blue </mark> variables are
Quality-Control metrics computed by <code>addPerCellQC()</code> of
<em><a href="https://bioconductor.org/packages/3.17/scuttle" class="external-link">scuttle</a></em>.</p>
<div class="section level5">
<h5 id="gene-information">Gene Information<a class="anchor" aria-label="anchor" href="#gene-information"></a>
</h5>
<p>Among the information in <code>rowData(RSE)</code> the next variables
are of interest for the analysis:</p>
<ul>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> gencodeID</span> </mark>: GENCODE ID of
each gene.</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> ensemblID</span> </mark>: gene ID in
Ensembl database.</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> EntrezID</span> </mark>: identifier of
each gene in NCBI Entrez database.</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> Symbol</span> </mark>: official gene
symbol for each mouse gene.</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> retained_after_feature_filtering</span>
</mark>: Boolean variable that equals TRUE if the gene passed the gene
filtering (with <code>filterByExpr()</code> of <em><a href="https://bioconductor.org/packages/3.17/edgeR" class="external-link">edgeR</a></em>)
based on its expression levels and FALSE if not.</li>
</ul>
</div>
<div class="section level5">
<h5 id="sample-information">Sample Information<a class="anchor" aria-label="anchor" href="#sample-information"></a>
</h5>
<p>Sample information in <code>colData(RSE)</code> contains the
following variables:</p>
<ul>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> SAMPLE_ID</span> </mark>: is the name of
the sample.</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> ERCCsumLogErr</span> </mark>: a summary
statistic quantifying overall difference of expected and actual ERCC
concentrations for one sample.</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> overallMapRate</span> </mark>: the
decimal fraction of reads which successfully mapped to the reference
genome (i.e. <em>numMapped</em> / <em>numReads</em>).</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> mitoMapped</span> </mark>: the number of
reads which successfully mapped to the mitochondrial chromosome.</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> totalMapped</span> </mark>: the number
of reads which successfully mapped to the canonical sequences in the
reference genome (excluding mitochondrial chromosomes).</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> mitoRate</span> </mark>: the decimal
fraction of reads which mapped to the mitochondrial chromosome, of those
which map at all (i.e. <em>mitoMapped</em> / (<em>totalMapped</em> +
<em>mitoMapped</em>))</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> totalAssignedGene</span> </mark>: the
decimal fraction of reads assigned unambiguously to a gene, with
<code>featureCounts</code> (Liao et al. 2014), of those in total.</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> rRNA_rate</span> </mark>: the decimal
fraction of reads assigned to a gene whose type is ‘rRNA’, of those
assigned to any gene.</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> Tissue</span> </mark>: tissue (mouse
brain or blood) from which the sample comes.</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> Age</span> </mark>: if the sample comes
from an adult or a pup mouse.</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> Sex</span> </mark>: if the sample comes
from a female (F) or male (M) mouse.</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> Expt</span> </mark>: the experiment
(nicotine or smoking exposure) to which the sample mouse was subjected;
it could be an exposed or a control mouse of that experiment.</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> Group</span> </mark>: if the sample
belongs to a nicotine/smoking-exposed mouse (Expt) or a nicotine/smoking
control mouse (Ctrl).</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> plate</span> </mark>: is the plate (1,2
or 3) in which the sample library was prepared.</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> Pregnancy</span> </mark>: if the sample
comes from a pregnant (Yes) or not pregnant (No) mouse.</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> medium</span> </mark>: is the medium in
which the sample was treated: water for brain samples and an elution
buffer (EB) for the blood ones.</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> flowcell</span> </mark>: is the
sequencing batch of each sample.</li>
<li>
<mark style="background-color: #DFF0FE"><span style="font-family: monospace"> sum</span> </mark>: library size (total
sum of counts across all genes for each sample).</li>
<li>
<mark style="background-color: #DFF0FE"><span style="font-family: monospace"> detected</span> </mark>: number of
non-zero expressed genes in each sample.</li>
<li>
<mark style="background-color: #DFF0FE"><span style="font-family: monospace"> subsets_Mito_sum</span> </mark>: total
sum of read counts of mt genes in each sample</li>
<li>
<mark style="background-color: #DFF0FE"><span style="font-family: monospace"> subsets_Mito_detected</span> </mark>:
total number of mt genes in each sample</li>
<li>
<mark style="background-color: #DFF0FE"><span style="font-family: monospace"> subsets_Mito_percent</span> </mark>: %
of mt genes’ counts of the total counts of the sample.</li>
<li>
<mark style="background-color: #DFF0FE"><span style="font-family: monospace"> subsets_Ribo_sum</span> </mark>: total
sum of read counts of ribosomal genes in each sample.</li>
<li>
<mark style="background-color: #DFF0FE"><span style="font-family: monospace"> subsets_Ribo_detected</span> </mark>:
total number of ribosomal genes in each sample.</li>
<li>
<mark style="background-color: #DFF0FE"><span style="font-family: monospace"> subsets_Ribo_percent</span> </mark>: %
of ribosomal genes’ counts of the total counts of the sample.</li>
</ul>
<p>Note: in our case, we’ll use samples from the nicotine experiment
only, so all samples come from brain and were treated in water.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://here.r-lib.org/" class="external-link">"here"</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">"SummarizedExperiment"</a></span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#################################</span></span>
<span><span class="co">##         Retrive data</span></span>
<span><span class="co">#################################</span></span>
<span><span class="co">## Provisional data source until smokingMouse pkg is accepted</span></span>
<span><span class="co"># load(here('provisional_data/rse_gene_mouse_RNAseq_nic-smo.Rdata'), verbose=TRUE)</span></span>
<span><span class="co"># ## Explore complete rse_gene object (including smoking-exposed samples as well)</span></span>
<span><span class="co"># rse_gene</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># #################################</span></span>
<span><span class="co"># ##   Extract data of interest</span></span>
<span><span class="co"># #################################</span></span>
<span><span class="co"># ## Keep samples from nicotine experiment only</span></span>
<span><span class="co"># rse_gene_nic &lt;- rse_gene[,which(rse_gene$Expt=='Nicotine')]</span></span>
<span><span class="co"># ## New dimensions</span></span>
<span><span class="co"># dim(rse_gene_nic)</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># #################################</span></span>
<span><span class="co"># ##  Access gene expression data</span></span>
<span><span class="co"># #################################</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># ## Raw counts for first 3 genes in the first 5 samples</span></span>
<span><span class="co"># assays(rse_gene_nic)$counts[1:3,1:5]</span></span>
<span><span class="co"># ## Log-normalized counts for first 3 genes in the first 5 samples</span></span>
<span><span class="co"># assays(rse_gene_nic)$logcounts[1:3,1:5]</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># #################################</span></span>
<span><span class="co"># ##     Access sample data</span></span>
<span><span class="co"># #################################</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># ## Info for the first 2 samples</span></span>
<span><span class="co"># head(colData(rse_gene_nic), 2)</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="differential-expression-analysis">Differential Expression Analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"></a>
</h3>
<p><strong>TODO</strong></p>
<figure><img src="figs/Analyses.jpg" width="700px" align="center"><figcaption style="color: gray; line-height: 0.87; text-align: justify"><font size="-1.8"> <b>Figure 2</b>: <b> Summary of the analyses. </b>
Steps in yellow will be properly performed. <b> 1. Data Preparation </b>
<b> 2. Exploratory Data Analysis </b> <b> 3. Differential Expression
Analysis </b> <b> 4. GO &amp; KEGG Analyses </b> <b> 5. DEG
visualization </b> </font> <font size="0.8"> Abbreviations: CPM: counts
per million; QC: quality control; PC: principal component; DEG:
differentially expressed genes; GO: Gene Ontology; KEGG: Kyoto
Encyclopedia of Genes and Genomes.</font>
</figcaption></figure><div class="section level4">
<h4 id="data-preparation">1. Data preparation<a class="anchor" aria-label="anchor" href="#data-preparation"></a>
</h4>
</div>
<div class="section level4">
<h4 id="exploratory-data-analysis">2. Exploratory Data Analysis<a class="anchor" aria-label="anchor" href="#exploratory-data-analysis"></a>
</h4>
<div class="section level5">
<h5 id="quality-control-analysis">2.1 Quality Control Analysis<a class="anchor" aria-label="anchor" href="#quality-control-analysis"></a>
</h5>
</div>
<div class="section level5">
<h5 id="explore-sample-level-effects">2.2 Explore sample-level effects<a class="anchor" aria-label="anchor" href="#explore-sample-level-effects"></a>
</h5>
</div>
<div class="section level5">
<h5 id="explore-gene-level-effects">2.2 Explore gene-level effects<a class="anchor" aria-label="anchor" href="#explore-gene-level-effects"></a>
</h5>
</div>
</div>
<div class="section level4">
<h4 id="differential-expression-analysis-1">3. Differential Expression Analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis-1"></a>
</h4>
<div class="section level5">
<h5 id="modeling">3.1 Modeling<a class="anchor" aria-label="anchor" href="#modeling"></a>
</h5>
</div>
<div class="section level5">
<h5 id="comparisons">3.2 Comparisons<a class="anchor" aria-label="anchor" href="#comparisons"></a>
</h5>
</div>
</div>
<div class="section level4">
<h4 id="go-kegg-analyses">4. GO &amp; KEGG analyses<a class="anchor" aria-label="anchor" href="#go-kegg-analyses"></a>
</h4>
</div>
<div class="section level4">
<h4 id="deg-visualization">5. DEG visualization<a class="anchor" aria-label="anchor" href="#deg-visualization"></a>
</h4>
</div>
</div>
</div>
<div class="section level2">
<h2 id="reproducibility">Reproducibility<a class="anchor" aria-label="anchor" href="#reproducibility"></a>
</h2>
<p>Code for creating the vignette</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Create the vignette</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/rstudio/rmarkdown" class="external-link">"rmarkdown"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/render.html" class="external-link">render</a></span><span class="op">(</span><span class="st">"SPEAQeasyWorkshop2023.Rmd"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Extract the R code</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://yihui.org/knitr/" class="external-link">"knitr"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/knit.html" class="external-link">knit</a></span><span class="op">(</span><span class="st">"SPEAQeasyWorkshop2023.Rmd"</span>, tangle <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Date the vignette was generated.</p>
<pre><code><span><span class="co">#&gt; [1] "2023-06-09 23:19:21 UTC"</span></span></code></pre>
<p>Wallclock time spent generating the vignette.</p>
<pre><code><span><span class="co">#&gt; Time difference of 7.644 secs</span></span></code></pre>
<p><code>R</code> session information.</p>
<pre><code><span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">setting </span> <span style="color: #555555; font-style: italic;">value</span></span></span>
<span><span class="co">#&gt;  version  R version 4.3.0 (2023-04-21)</span></span>
<span><span class="co">#&gt;  os       Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt;  system   x86_64, linux-gnu</span></span>
<span><span class="co">#&gt;  ui       X11</span></span>
<span><span class="co">#&gt;  language en</span></span>
<span><span class="co">#&gt;  collate  en_US.UTF-8</span></span>
<span><span class="co">#&gt;  ctype    en_US.UTF-8</span></span>
<span><span class="co">#&gt;  tz       UTC</span></span>
<span><span class="co">#&gt;  date     2023-06-09</span></span>
<span><span class="co">#&gt;  pandoc   2.19.2 @ /usr/local/bin/ (via rmarkdown)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">package             </span> <span style="color: #555555; font-style: italic;">*</span> <span style="color: #555555; font-style: italic;">version  </span> <span style="color: #555555; font-style: italic;">date (UTC)</span> <span style="color: #555555; font-style: italic;">lib</span> <span style="color: #555555; font-style: italic;">source</span></span></span>
<span><span class="co">#&gt;  backports              1.4.1     <span style="color: #555555;">2021-12-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  bibtex                 0.5.1     <span style="color: #555555;">2023-01-26</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  Biobase              * 2.60.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  BiocGenerics         * 0.46.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  BiocManager            1.30.20   <span style="color: #555555;">2023-02-24</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  BiocStyle            * 2.28.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  bitops                 1.0-7     <span style="color: #555555;">2021-04-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  bookdown               0.34      <span style="color: #555555;">2023-05-09</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  bslib                  0.5.0     <span style="color: #555555;">2023-06-09</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  cachem                 1.0.8     <span style="color: #555555;">2023-05-01</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  cli                    3.6.1     <span style="color: #555555;">2023-03-23</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  crayon                 1.5.2     <span style="color: #555555;">2022-09-29</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  DelayedArray           0.26.3    <span style="color: #555555;">2023-05-22</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  desc                   1.4.2     <span style="color: #555555;">2022-09-08</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  digest                 0.6.31    <span style="color: #555555;">2022-12-11</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  evaluate               0.21      <span style="color: #555555;">2023-05-05</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  fastmap                1.1.1     <span style="color: #555555;">2023-02-24</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  fs                     1.6.2     <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  generics               0.1.3     <span style="color: #555555;">2022-07-05</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  GenomeInfoDb         * 1.36.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  GenomeInfoDbData       1.2.10    <span style="color: #555555;">2023-05-17</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  GenomicRanges        * 1.52.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  glue                   1.6.2     <span style="color: #555555;">2022-02-24</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  here                 * 1.0.1     <span style="color: #555555;">2020-12-13</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  htmltools              0.5.5     <span style="color: #555555;">2023-03-23</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  httr                   1.4.6     <span style="color: #555555;">2023-05-08</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  IRanges              * 2.34.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  jquerylib              0.1.4     <span style="color: #555555;">2021-04-26</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  jsonlite               1.8.5     <span style="color: #555555;">2023-06-05</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  knitr                  1.43      <span style="color: #555555;">2023-05-25</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  lattice                0.21-8    <span style="color: #555555;">2023-04-05</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  lifecycle              1.0.3     <span style="color: #555555;">2022-10-07</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  lubridate              1.9.2     <span style="color: #555555;">2023-02-10</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  magrittr               2.0.3     <span style="color: #555555;">2022-03-30</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  Matrix                 1.5-4.1   <span style="color: #555555;">2023-05-18</span> <span style="color: #555555;">[3]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  MatrixGenerics       * 1.12.1    <span style="color: #555555;">2023-06-08</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  matrixStats          * 1.0.0     <span style="color: #555555;">2023-06-02</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  memoise                2.0.1     <span style="color: #555555;">2021-11-26</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  pkgdown                2.0.7     <span style="color: #555555;">2022-12-14</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  plyr                   1.8.8     <span style="color: #555555;">2022-11-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  purrr                  1.0.1     <span style="color: #555555;">2023-01-10</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  R6                     2.5.1     <span style="color: #555555;">2021-08-19</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  ragg                   1.2.5     <span style="color: #555555;">2023-01-12</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  Rcpp                   1.0.10    <span style="color: #555555;">2023-01-22</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  RCurl                  1.98-1.12 <span style="color: #555555;">2023-03-27</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  RefManageR           * 1.4.0     <span style="color: #555555;">2022-09-30</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  rlang                  1.1.1     <span style="color: #555555;">2023-04-28</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  rmarkdown              2.22      <span style="color: #555555;">2023-06-01</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  rprojroot              2.0.3     <span style="color: #555555;">2022-04-02</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  S4Arrays               1.0.4     <span style="color: #555555;">2023-05-14</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  S4Vectors            * 0.38.1    <span style="color: #555555;">2023-05-02</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  sass                   0.4.6     <span style="color: #555555;">2023-05-03</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  sessioninfo          * 1.2.2     <span style="color: #555555;">2021-12-06</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  stringi                1.7.12    <span style="color: #555555;">2023-01-11</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  stringr                1.5.0     <span style="color: #555555;">2022-12-02</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  SummarizedExperiment * 1.30.2    <span style="color: #555555;">2023-06-06</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  systemfonts            1.0.4     <span style="color: #555555;">2022-02-11</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  textshaping            0.3.6     <span style="color: #555555;">2021-10-13</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  timechange             0.2.0     <span style="color: #555555;">2023-01-11</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  vctrs                  0.6.2     <span style="color: #555555;">2023-04-19</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  xfun                   0.39      <span style="color: #555555;">2023-04-20</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  xml2                   1.3.4     <span style="color: #555555;">2023-04-27</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  XVector                0.40.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  yaml                   2.3.7     <span style="color: #555555;">2023-01-23</span> <span style="color: #555555;">[2]</span> <span style="color: #BB00BB; font-weight: bold;">RSPM (R 4.3.0)</span></span></span>
<span><span class="co">#&gt;  zlibbioc               1.46.0    <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [1] /__w/_temp/Library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [2] /usr/local/lib/R/site-library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [3] /usr/local/lib/R/library</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
</div>
<div class="section level2">
<h2 id="bibliography">Bibliography<a class="anchor" aria-label="anchor" href="#bibliography"></a>
</h2>
<p>This vignette was generated using <em><a href="https://bioconductor.org/packages/3.17/BiocStyle" class="external-link">BiocStyle</a></em>
(Oleś, 2023), <em><a href="https://CRAN.R-project.org/package=knitr" class="external-link">knitr</a></em> (Xie,
2023) and <em><a href="https://CRAN.R-project.org/package=rmarkdown" class="external-link">rmarkdown</a></em>
(Allaire, Xie, Dervieux, McPherson, Luraschi, Ushey, Atkins, Wickham,
Cheng, Chang, and Iannone, 2023) running behind the scenes.</p>
<p>Citations made with <em><a href="https://CRAN.R-project.org/package=RefManageR" class="external-link">RefManageR</a></em>
(McLean, 2017).</p>
<p>
<a id="bib-allaire2023rmarkdown"></a><a href="#cite-allaire2023rmarkdown">[1]</a><cite>
J. Allaire, Y. Xie, C. Dervieux, et al. <em>rmarkdown: Dynamic Documents
for R</em>. R package version 2.22. 2023. URL:
<a href="https://github.com/rstudio/rmarkdown" class="external-link">https://github.com/rstudio/rmarkdown</a>.</cite>
</p>
<p>
<a id="bib-eagles2021speaqeasy"></a><a href="#cite-eagles2021speaqeasy">[2]</a><cite>
N. J. Eagles, E. E. Burke, J. Leonard, et al. “SPEAQeasy: a scalable
pipeline for expression analysis and quantification for
R/bioconductor-powered RNA-seq analyses”. In: <em>BMC
Bioinformatics</em> (2021). DOI:
<a href="https://doi.org/10.1186/s12859-021-04142-3" class="external-link">10.1186/s12859-021-04142-3</a>.
URL:
<a href="https://doi.org/10.1186/s12859-021-04142-3" class="external-link">https://doi.org/10.1186/s12859-021-04142-3</a>.</cite>
</p>
<p>
<a id="bib-gonzalezpadilla2023provides"></a><a href="#cite-gonzalezpadilla2023provides">[3]</a><cite>
D. Gonzalez-Padilla and L. Collado-Torres. <em>Provides access to
smokingMouse project data </em>. <a href="https://github.com/LieberInstitute/smokingMouse/smokingMouse" class="external-link uri">https://github.com/LieberInstitute/smokingMouse/smokingMouse</a>
- R package version 0.99.3. 2023. DOI:
<a href="https://doi.org/10.18129/B9.bioc.smokingMouse" class="external-link">10.18129/B9.bioc.smokingMouse</a>.
URL:
<a href="http://www.bioconductor.org/packages/smokingMouse" class="external-link">http://www.bioconductor.org/packages/smokingMouse</a>.</cite>
</p>
<p>
<a id="bib-mclean2017refmanager"></a><a href="#cite-mclean2017refmanager">[4]</a><cite>
M. W. McLean. “RefManageR: Import and Manage BibTeX and BibLaTeX
References in R”. In: <em>The Journal of Open Source Software</em>
(2017). DOI:
<a href="https://doi.org/10.21105/joss.00338" class="external-link">10.21105/joss.00338</a>.</cite>
</p>
<p>
<a id="bib-ole2023biocstyle"></a><a href="#cite-ole2023biocstyle">[5]</a><cite>
A. Oleś. <em>BiocStyle: Standard styles for vignettes and other
Bioconductor documents</em>. R package version 2.28.0. 2023. DOI:
<a href="https://doi.org/10.18129/B9.bioc.BiocStyle" class="external-link">10.18129/B9.bioc.BiocStyle</a>.
URL:
<a href="https://bioconductor.org/packages/BiocStyle" class="external-link">https://bioconductor.org/packages/BiocStyle</a>.</cite>
</p>
<p>
<a id="bib-xie2023knitr"></a><a href="#cite-xie2023knitr">[6]</a><cite>
Y. Xie. <em>knitr: A General-Purpose Package for Dynamic Report
Generation in R</em>. R package version 1.43. 2023. URL:
<a href="https://yihui.org/knitr/" class="external-link">https://yihui.org/knitr/</a>.</cite>
</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick J. Eagles, Renee Garcia-Flores, Daianna Gonzalez-Padilla.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
