<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>SPEAQeasy: a Scalable Pipeline for Expression Analysis and Quantification, a demonstration of its usage on real data • SPEAQeasyWorkshop2023</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="SPEAQeasy: a Scalable Pipeline for Expression Analysis and Quantification, a demonstration of its usage on real data">
<meta property="og:description" content="SPEAQeasyWorkshop2023">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">SPEAQeasyWorkshop2023</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/SPEAQeasyWorkshop2023.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>SPEAQeasy: a Scalable Pipeline for Expression
Analysis and Quantification, a demonstration of its usage on real
data</h1>
                        <h4 data-toc-skip class="author">Daianna
Gonzalez-Padilla</h4>
            <address class="author_afil">
      Lieber Institute for Brain
Development<br><a class="author_email" href="mailto:#"></a><a href="mailto:glezdaianna@gmail.com" class="email">glezdaianna@gmail.com</a>
      </address>
                              <h4 data-toc-skip class="author">Renee
Garcia-Flores</h4>
            <address class="author_afil">
      Lieber Institute for Brain
Development<br><a class="author_email" href="mailto:#"></a><a href="mailto:renee.garciaflores@gmail.com" class="email">renee.garciaflores@gmail.com</a>
      </address>
                              <h4 data-toc-skip class="author">Nicholas
J. Eagles</h4>
            <address class="author_afil">
      Lieber Institute for Brain
Development<br><a class="author_email" href="mailto:#"></a><a href="mailto:nick.eagles@libd.org" class="email">nick.eagles@libd.org</a>
      </address>
                              <h4 data-toc-skip class="author">Leonardo
Collado-Torres</h4>
            <address class="author_afil">
      Lieber Institute for Brain DevelopmentCenter for Computational
Biology, Johns Hopkins
University<br><a class="author_email" href="mailto:#"></a><a href="mailto:lcolladotor@gmail.com" class="email">lcolladotor@gmail.com</a>
      </address>
                  
            <h4 data-toc-skip class="date">24 July 2023</h4>
      
      
      <div class="hidden name"><code>SPEAQeasyWorkshop2023.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h2>
<p>This workshop aims to present the <code>SPEAQeasy</code> (Eagles,
Burke, Leonard, Barry, Stolz, Huuki, Phan, Serrato, Gutiérrez-Millán,
Aguilar-Ordoñez, Jaffe, and Collado-Torres, 2021) RNA-seq processing
pipeline, demonstrating how its outputs can be used in Differential
Expression Analyses (DEA) and the type of results it enables to obtain
using other Bioconductor R packages such as <em><a href="https://bioconductor.org/packages/3.18/limma" class="external-link">limma</a></em> and
<em><a href="https://bioconductor.org/packages/3.18/edgeR" class="external-link">edgeR</a></em>.</p>
<div class="section level3">
<h3 id="description">Description<a class="anchor" aria-label="anchor" href="#description"></a>
</h3>
<p>In this demo, participants will be able to understand what the
SPEAQeasy pipeline does and what it returns through a differential
expression analysis performed on a dataset from the
<code>smokingMouse</code> (Gonzalez-Padilla and Collado-Torres, 2023)
Bioconductor package, which contains gene expression data and sample
information from multiple RNA-sequencing experiments in mice.</p>
</div>
<div class="section level3">
<h3 id="instructors">Instructors<a class="anchor" aria-label="anchor" href="#instructors"></a>
</h3>
<ul>
<li><a href="https://daianna21.github.io" class="external-link">Daianna Gonzalez</a></li>
<li><a href="https://reneegf.github.io" class="external-link">Renee Garcia</a></li>
<li><a href="https://nick-eagles.github.io" class="external-link">Nick Eagles</a></li>
</ul>
</div>
<div class="section level3">
<h3 id="pre-requisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#pre-requisites"></a>
</h3>
<ul>
<li>General familiarity with bulk RNA-seq data.</li>
<li>Basic knowledge of differential expression.</li>
<li>Basic experience handling <em><a href="https://bioconductor.org/packages/3.18/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>
and <em><a href="https://bioconductor.org/packages/3.18/GenomicRanges" class="external-link">GenomicRanges</a></em>
packages.</li>
<li>Previous review of <a href="http://research.libd.org/smokingMouse/" class="external-link uri">http://research.libd.org/smokingMouse/</a> for data
explanation. (Optional)</li>
</ul>
</div>
<div class="section level3">
<h3 id="participation">Participation<a class="anchor" aria-label="anchor" href="#participation"></a>
</h3>
<p>Participants are expected to follow us on the theoretical
explanations and presentation of the code logic, but are not really
expected to run code since some analyses take plenty of time. An active
participation and opportune interventions are always welcome and
enriching for the discussion.</p>
</div>
<div class="section level3">
<h3 id="r-bioconductor-packages-used">
<em>R</em> / <em>Bioconductor</em> packages used<a class="anchor" aria-label="anchor" href="#r-bioconductor-packages-used"></a>
</h3>
<ul>
<li><em><a href="https://bioconductor.org/packages/3.18/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.18/smokingMouse" class="external-link">smokingMouse</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.18/limma" class="external-link">limma</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.18/edgeR" class="external-link">edgeR</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.18/scater" class="external-link">scater</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.18/variancePartition" class="external-link">variancePartition</a></em></li>
<li><em><a href="https://bioconductor.org/packages/3.18/ComplexHeatmap" class="external-link">ComplexHeatmap</a></em></li>
</ul>
</div>
<div class="section level3">
<h3 id="time-outline">Time outline<a class="anchor" aria-label="anchor" href="#time-outline"></a>
</h3>
<table class="table">
<thead><tr class="header">
<th>Activity</th>
<th>Time</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Introduction to SPEAQeasy</td>
<td>15m</td>
</tr>
<tr class="even">
<td>Data overview</td>
<td>10m</td>
</tr>
<tr class="odd">
<td>Differential Expression Analysis</td>
<td>65m</td>
</tr>
</tbody>
</table>
<p>Total: a 90 minute session.</p>
</div>
</div>
<div class="section level2">
<h2 id="workshop">Workshop<a class="anchor" aria-label="anchor" href="#workshop"></a>
</h2>
<div class="section level3">
<h3 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h3>
<div class="section level4">
<h4 id="learning-goals">Learning goals<a class="anchor" aria-label="anchor" href="#learning-goals"></a>
</h4>
<p>The global goals of this workshop are:</p>
<ul>
<li>to understand the purpose of <a href="https://github.com/LieberInstitute/SPEAQeasy" class="external-link">SPEAQeasy</a> and
the type of data it produces;</li>
<li>to learn what type of analyses can be done with <a href="https://github.com/LieberInstitute/SPEAQeasy" class="external-link">SPEAQeasy</a>
outputs;</li>
<li>to understand techniques used to filter and normalize
gene-expression data;</li>
<li>to learn of quality control analyses and other type of exploratory
data analyses;</li>
<li>to understand how to identify suitable covariates for differential
expression analyses;</li>
<li>to learn how to perform a differential expression analysis and how
to visualize expression of DEGs.</li>
</ul>
</div>
<div class="section level4">
<h4 id="learning-objectives">Learning objectives<a class="anchor" aria-label="anchor" href="#learning-objectives"></a>
</h4>
<p>As specific objectives we have:</p>
<ul>
<li>to use <em><a href="https://bioconductor.org/packages/3.18/ggplot2" class="external-link">ggplot2</a></em>
to assess bulk RNA-seq quality metrics for the <em><a href="https://bioconductor.org/packages/3.18/smokingMouse" class="external-link">smokingMouse</a></em>
dataset stored in <em><a href="https://bioconductor.org/packages/3.18/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>
objects output by the <a href="https://github.com/LieberInstitute/SPEAQeasy" class="external-link">SPEAQeasy</a>
pipeline;</li>
<li>use <em><a href="https://bioconductor.org/packages/3.18/variancePartition" class="external-link">variancePartition</a></em>
to perform a Canonical Correlation Analysis (CCA) on gene-expression
data with associated quality metrics;</li>
<li>to use <em><a href="https://bioconductor.org/packages/3.18/limma" class="external-link">limma</a></em>
functions such as <code><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">voom()</a></code>, <code><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit()</a></code>,
<code><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes()</a></code> and <code>topTable()</code> to find DEGs.</li>
<li>to use <em><a href="https://bioconductor.org/packages/3.18/ComplexHeatmap" class="external-link">ComplexHeatmap</a></em>
to create heatmaps of expression levels of DEGs, identifying clusters of
up and down regulated genes.</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="speaqeasy-overview">SPEAQeasy overview<a class="anchor" aria-label="anchor" href="#speaqeasy-overview"></a>
</h3>
<p>We introduce <a href="https://github.com/LieberInstitute/SPEAQeasy" class="external-link">SPEAQeasy</a>, a
<strong>S</strong>calable RNA-seq processing <strong>P</strong>ipeline
for <strong>E</strong>xpression <strong>A</strong>nalysis and
<strong>Q</strong>uantification, that is <strong>easy</strong> to
install, use, and share with others. More detailed documentation is <a href="http://research.libd.org/SPEAQeasy/" class="external-link">here</a>.</p>
<div class="float">
<img src="figs/SPEAQeasy_workflow.png" alt="Functional workflow. Yellow coloring denotes steps which are optional or not always performed."><div class="figcaption">Functional workflow. Yellow coloring denotes
steps which are optional or not always performed.</div>
</div>
<p>SPEAQeasy starts with raw sequencing outputs in FASTQ format and
ultimately produces <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html" class="external-link">RangedSummarizedExperiment</a>
objects with raw and normalized counts for each feature type: genes,
exons, exon-exon junctions, and transcripts.</p>
<p>For human samples, variant calling is performed at a list of highly
variable sites. A single <a href="https://www.internationalgenome.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-40/" class="external-link">VCF
file</a> is generated, containing genotype calls for all samples. This
establishes a sufficiently unique profile for each sample which can be
compared against pre-sequencing genotype calls to resolve potential
identity issues.</p>
<p>Optionally, <a href="https://genome.ucsc.edu/goldenPath/help/bigWig.html" class="external-link">BigWig</a>
coverage files by strand, and averaged across strands, are produced. An
additional step utilizes the <a href="https://bioconductor.org/packages/release/bioc/html/derfinder.html" class="external-link">derfinder</a>
package to compute expressed regions, a precursor to analyses finding
differentially expressed regions (DERs).</p>
<div class="section level4">
<h4 id="installation-and-set-up">Installation and set-up<a class="anchor" aria-label="anchor" href="#installation-and-set-up"></a>
</h4>
<p>SPEAQeasy is available from GitHub, and can be cloned:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">git</span> clone git@github.com:LieberInstitute/SPEAQeasy.git</span></code></pre></div>
<p>SPEAQeasy dependencies are recommended to be installed using Docker
or Singularity, involving a fairly quick installation that is
reproducible regardless of the computing environment. We’ll install
using Singularity, which is more frequently available in
high-performance computing (HPC) environments:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="co">#  From within the SPEAQeasy repository:</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">bash</span> install_software.sh <span class="st">"singularity"</span></span></code></pre></div>
<p>If neither Docker nor Singularity are available, a more experimental
local installation of all dependencies is an option:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="co">#  From within the SPEAQeasy repository:</span></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a><span class="fu">bash</span> install_software.sh <span class="st">"local"</span></span></code></pre></div>
<p>Please note that installation may take some time– <strong>installing
and running SPEAQeasy is not required for this workshop</strong>, but is
demonstrated to familiarize participants with the process. The workshop
files include outputs from SPEAQeasy, which can be used later for the
example differential expression analysis.</p>
<p>Choose a “main” script as appropriate for your particular set-up.
“Main” scripts and associated configuration files exist for
SLURM-managed computing clusters, SGE-managed clusters, local machines,
and the <a href="https://jhpce.jhu.edu/" class="external-link">JHPCE</a> cluster.</p>
<table class="table">
<colgroup>
<col width="23%">
<col width="30%">
<col width="45%">
</colgroup>
<thead><tr class="header">
<th>Environment</th>
<th>“Main” script</th>
<th>Config file</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>SLURM cluster</td>
<td>run_pipeline_slurm.sh</td>
<td>conf/slurm.config or conf/docker_slurm.config</td>
</tr>
<tr class="even">
<td>SGE cluster</td>
<td>run_pipeline_sge.sh</td>
<td>conf/sge.config or conf/docker_sge.config</td>
</tr>
<tr class="odd">
<td>Local machine</td>
<td>run_pipeline_local.sh</td>
<td>conf/local.config or conf/docker_local.config</td>
</tr>
<tr class="even">
<td>JHPCE cluster</td>
<td>run_pipeline_jhpce.sh</td>
<td>conf/jhpce.config</td>
</tr>
</tbody>
</table>
<p>Within the main script, you can configure arguments specific to the
experiment, such as the reference organism, pairing of samples, and
where to place output files, among other specifications.</p>
<p>When running SPEAQeasy on a cluster (i.e. SLURM, SGE, or JHPCE
users), it is recommended you submit the pipeline as a job, using the
command appropriate for your cluster. For those running SPEAQeasy
locally, the main script can be executed directly, ideally as a
background task.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="co">#  SLURM-managed clusters</span></span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="ex">sbatch</span> run_pipeline_slurm.sh</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a><span class="co">#  SGE-managed clusters</span></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="ex">qsub</span> run_pipeline_sge.sh</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a><span class="co">#  local machines</span></span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a><span class="fu">bash</span> run_pipeline_local.sh</span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a></span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a><span class="co">#  The JHPCE cluster</span></span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a><span class="ex">qsub</span> run_pipeline_jhpce.sh</span></code></pre></div>
<p>Here’s an example of such a file:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">example_main_script</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"extdata"</span>, <span class="st">"SPEAQeasy-example"</span>, <span class="st">"run_pipeline_jhpce.sh"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"SPEAQeasyWorkshop2023"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/cat.html" class="external-link">cat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/readLines.html" class="external-link">readLines</a></span><span class="op">(</span><span class="va">example_main_script</span><span class="op">)</span>, <span class="st">"\n"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; #!/bin/bash</span></span>
<span><span class="co">#&gt;  #$ -l mem_free=40G,h_vmem=40G,h_fsize=800G</span></span>
<span><span class="co">#&gt;  #$ -o ../../processed-data/01_SPEAQeasy/SPEAQeasy_output.log</span></span>
<span><span class="co">#&gt;  #$ -e ../../processed-data/01_SPEAQeasy/SPEAQeasy_output.log</span></span>
<span><span class="co">#&gt;  #$ -cwd</span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt;  #  Get absolute path to git repo</span></span>
<span><span class="co">#&gt;  base_dir=$(git rev-parse --show-toplevel)</span></span>
<span><span class="co">#&gt;  </span></span>
<span><span class="co">#&gt;  nextflow SPEAQeasy/main.nf \</span></span>
<span><span class="co">#&gt;      --sample "paired" \</span></span>
<span><span class="co">#&gt;      --reference "mm10" \</span></span>
<span><span class="co">#&gt;      --strand "reverse" \</span></span>
<span><span class="co">#&gt;      --ercc \</span></span>
<span><span class="co">#&gt;      --experiment "smoking_mouse" \</span></span>
<span><span class="co">#&gt;      --annotation "/dcl01/lieber/ajaffe/Nick/SPEAQeasy/Annotation" \</span></span>
<span><span class="co">#&gt;      -with-report "${base_dir}/processed-data/01_SPEAQeasy/execution_reports/JHPCE_run.html" \</span></span>
<span><span class="co">#&gt;      -w "${base_dir}/processed-data/01_SPEAQeasy/work" \</span></span>
<span><span class="co">#&gt;      --input "${base_dir}/processed-data/01_SPEAQeasy" \</span></span>
<span><span class="co">#&gt;      --output "${base_dir}/processed-data/01_SPEAQeasy/pipeline_output" \</span></span>
<span><span class="co">#&gt;      -profile jhpce</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="data-overview-and-download">Data overview and download<a class="anchor" aria-label="anchor" href="#data-overview-and-download"></a>
</h3>
<p>The dataset that will be used for the DEA can be downloaded from the
<code>smokingMouse</code> package. We will download the
<code>rse_gene</code> object which contains the data at the gene level.
Visit <a href="http://research.libd.org/smokingMouse/" class="external-link">here</a> for more
details of this package.</p>
<img src="figs/Data_overview_experiments.png" width="750px"><figcaption style="color: gray; line-height: 0.85; text-align: justify"><p><font size="-2"><b>Figure 1</b>: <b>Experimental design of the study.
</b> The data was obtained from an experiment in which <b>A)</b> 6
pregnant dams and 17 non-pregnant female adult mice were administered
nicotine by intraperitoneal injection (IP; n=12) or controls (n=11). A
total of 42 pups were born to pregnant dams: 19 were born to mice that
were administered nicotine and 23 to control mice. Samples from frontal
cortices of P0 pups and adults were obtained. <b>B)</b> RNA was
extracted from those samples and then RNA-seq libraries were prepared
and sequenced to obtain the expression counts of the genes. At the end,
a dataset of 65 samples (either from adult or pup brain) and 55,401
genes was generated. </font></p>
</figcaption><p>The data reside in a <code>RangedSummarizedExperiment</code> (RSE)
object, containing the following assays:</p>
<ul>
<li>
<code>counts</code>: original read counts of the 55,401 mouse genes
across 208 samples (including the 65 nicotine samples).</li>
<li>
<code>logcounts</code>: normalized and scaled counts (<span class="math inline">\(log_2(CPM + 0.5)\)</span>) of the same genes
across the same samples; normalization was carried out applying TMM
method with <code>cpm(calcNormFactors())</code> of <em><a href="https://bioconductor.org/packages/3.18/edgeR" class="external-link">edgeR</a></em>.</li>
</ul>
<p>The same RSE contains the sample information in
<code>colData(RSE)</code> and the gene information in
<code>rowData(RSE)</code>.</p>
<p><mark style="background-color: #FCF3CF"> Yellow </mark> variables
correspond to SPEAQeasy outputs that are going to be used in downstream
analyses.</p>
<p><mark style="background-color: #FAECF8"> Pink </mark> variables are
specific to the study, such as sample metadata and some others
containing additional information about the genes.</p>
<p><mark style="background-color: #DFF0FE"> Blue </mark> variables are
quality-control metrics computed by <code><a href="https://rdrr.io/pkg/scuttle/man/addPerCellQCMetrics.html" class="external-link">addPerCellQC()</a></code> of
<em><a href="https://bioconductor.org/packages/3.18/scuttle" class="external-link">scuttle</a></em>.</p>
<div class="section level4">
<h4 id="gene-information">Gene Information<a class="anchor" aria-label="anchor" href="#gene-information"></a>
</h4>
<p>Among the information in <code>rowData(RSE)</code> the next variables
are of interest for the analysis:</p>
<ul>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> gencodeID</span> </mark>: GENCODE ID of
each gene.</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> ensemblID</span> </mark>: gene ID in
Ensembl database.</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> EntrezID</span> </mark>: identifier of
each gene in NCBI Entrez database.</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> Symbol</span> </mark>: official gene
symbol for each mouse gene.</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> retained_after_feature_filtering</span>
</mark>: Boolean variable that equals TRUE if the gene passed the gene
filtering (with <code>filterByExpr()</code> of <em><a href="https://bioconductor.org/packages/3.18/edgeR" class="external-link">edgeR</a></em>)
based on its expression levels and FALSE if not.</li>
</ul>
</div>
<div class="section level4">
<h4 id="sample-information">Sample Information<a class="anchor" aria-label="anchor" href="#sample-information"></a>
</h4>
<p>Sample information in <code>colData(RSE)</code> contains the
following variables:</p>
<ul>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> SAMPLE_ID</span> </mark>: is the name of
the sample.</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> ERCCsumLogErr</span> </mark>: a summary
statistic quantifying overall difference of expected and actual ERCC
concentrations for one sample.</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> overallMapRate</span> </mark>: the
decimal fraction of reads which successfully mapped to the reference
genome (i.e. <em>numMapped</em> / <em>numReads</em>).</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> mitoMapped</span> </mark>: the number of
reads which successfully mapped to the mitochondrial chromosome.</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> totalMapped</span> </mark>: the number
of reads which successfully mapped to the canonical sequences in the
reference genome (excluding mitochondrial chromosomes).</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> mitoRate</span> </mark>: the decimal
fraction of reads which mapped to the mitochondrial chromosome, of those
which map at all (i.e. <em>mitoMapped</em> / (<em>totalMapped</em> +
<em>mitoMapped</em>))</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> totalAssignedGene</span> </mark>: the
decimal fraction of reads assigned unambiguously to a gene (including
mitochondrial genes), with <code>featureCounts</code> (Liao et
al. 2014), of those in total.</li>
<li>
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> rRNA_rate</span> </mark>: the decimal
fraction of reads assigned to a gene whose type is ‘rRNA’, of those
assigned to any gene.</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> Tissue</span> </mark>: tissue (mouse
brain or blood) from which the sample comes.</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> Age</span> </mark>: if the sample comes
from an adult or a pup mouse.</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> Sex</span> </mark>: if the sample comes
from a female (F) or male (M) mouse.</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> Expt</span> </mark>: the experiment
(nicotine or smoking exposure) to which the sample mouse was subjected;
it could be an exposed or a control mouse of that experiment.</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> Group</span> </mark>: if the sample
belongs to a nicotine/smoking-exposed mouse (Expt) or a nicotine/smoking
control mouse (Ctrl).</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> plate</span> </mark>: is the plate (1,2
or 3) in which the sample library was prepared.</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> Pregnancy</span> </mark>: if the sample
comes from a pregnant (Yes) or not pregnant (No) mouse.</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> medium</span> </mark>: is the medium in
which the sample was treated: water for brain samples and an elution
buffer (EB) for the blood ones.</li>
<li>
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> flowcell</span> </mark>: is the
sequencing batch of each sample.</li>
<li>
<mark style="background-color: #DFF0FE"><span style="font-family: monospace"> sum</span> </mark>: library size (total
sum of counts across all genes for each sample).</li>
<li>
<mark style="background-color: #DFF0FE"><span style="font-family: monospace"> detected</span> </mark>: number of
non-zero expressed genes in each sample.</li>
</ul>
<div class="alert alert-info">
<p><strong>Note:</strong> in our case, we’ll use samples from the
nicotine experiment only, so all samples come from brain and were
treated in water.</p>
</div>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/SummarizedExperiment" class="external-link">SummarizedExperiment</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: MatrixGenerics</span></span>
<span><span class="co">#&gt; Loading required package: matrixStats</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'MatrixGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">#&gt;     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">#&gt;     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">#&gt;     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">#&gt;     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">#&gt;     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">#&gt;     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">#&gt;     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">#&gt;     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">#&gt;     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">#&gt;     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">#&gt;     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">#&gt;     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">#&gt;     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">#&gt;     rowWeightedSds, rowWeightedVars</span></span>
<span><span class="co">#&gt; Loading required package: GenomicRanges</span></span>
<span><span class="co">#&gt; Loading required package: stats4</span></span>
<span><span class="co">#&gt; Loading required package: BiocGenerics</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'BiocGenerics'</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:stats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     IQR, mad, sd, var, xtabs</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">#&gt;     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">#&gt;     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">#&gt;     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">#&gt;     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">#&gt;     table, tapply, union, unique, unsplit, which.max, which.min</span></span>
<span><span class="co">#&gt; Loading required package: S4Vectors</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'S4Vectors'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:utils':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     findMatches</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:base':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     expand.grid, I, unname</span></span>
<span><span class="co">#&gt; Loading required package: IRanges</span></span>
<span><span class="co">#&gt; Loading required package: GenomeInfoDb</span></span>
<span><span class="co">#&gt; Loading required package: Biobase</span></span>
<span><span class="co">#&gt; Welcome to Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     Vignettes contain introductory material; view with</span></span>
<span><span class="co">#&gt;     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">#&gt;     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'Biobase'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:MatrixGenerics':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     rowMedians</span></span>
<span><span class="co">#&gt; The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     anyMissing, rowMedians</span></span>
<span></span>
<span><span class="co">## Connect to ExperimentHub</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ExperimentHub</span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: AnnotationHub</span></span>
<span><span class="co">#&gt; Loading required package: BiocFileCache</span></span>
<span><span class="co">#&gt; Loading required package: dbplyr</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attaching package: 'AnnotationHub'</span></span>
<span><span class="co">#&gt; The following object is masked from 'package:Biobase':</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     cache</span></span>
<span><span class="va">eh</span> <span class="op">&lt;-</span> <span class="fu">ExperimentHub</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ExperimentHub/man/ExperimentHub-class.html" class="external-link">ExperimentHub</a></span><span class="op">(</span><span class="op">)</span></span>
<span></span>
<span><span class="co">################################</span></span>
<span><span class="co">#         Retrieve data</span></span>
<span><span class="co">################################</span></span>
<span><span class="co">## Load the datasets of the package</span></span>
<span><span class="va">myfiles</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/AnnotationHub/man/AnnotationHub-class.html" class="external-link">query</a></span><span class="op">(</span><span class="va">eh</span>, <span class="st">"smokingMouse"</span><span class="op">)</span></span>
<span>  </span>
<span><span class="co">## Download the mouse gene data</span></span>
<span><span class="va">rse_gene</span> <span class="op">&lt;-</span> <span class="va">myfiles</span><span class="op">[[</span><span class="st">"EH8313"</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; see ?smokingMouse and browseVignettes('smokingMouse') for documentation</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">## This is a RangedSummarizedExperiment object</span></span>
<span><span class="va">rse_gene</span></span>
<span><span class="co">#&gt; class: RangedSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 55401 208 </span></span>
<span><span class="co">#&gt; metadata(1): Obtained_from</span></span>
<span><span class="co">#&gt; assays(2): counts logcounts</span></span>
<span><span class="co">#&gt; rownames(55401): ENSMUSG00000102693.1 ENSMUSG00000064842.1 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000064371.1 ENSMUSG00000064372.1</span></span>
<span><span class="co">#&gt; rowData names(13): Length gencodeID ... DE_in_pup_brain_nicotine</span></span>
<span><span class="co">#&gt;   DE_in_pup_brain_smoking</span></span>
<span><span class="co">#&gt; colnames: NULL</span></span>
<span><span class="co">#&gt; colData names(71): SAMPLE_ID FQCbasicStats ...</span></span>
<span><span class="co">#&gt;   retained_after_QC_sample_filtering</span></span>
<span><span class="co">#&gt;   retained_after_manual_sample_filtering</span></span>
<span></span>
<span><span class="co">## Check sample info </span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse_gene</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 5 rows and 5 columns</span></span>
<span><span class="co">#&gt;     SAMPLE_ID FQCbasicStats perBaseQual perTileQual  perSeqQual</span></span>
<span><span class="co">#&gt;   &lt;character&gt;   &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1 Sample_2914          PASS        PASS        PASS        PASS</span></span>
<span><span class="co">#&gt; 2 Sample_4042          PASS        PASS        PASS        PASS</span></span>
<span><span class="co">#&gt; 3 Sample_4043          PASS        PASS        PASS        PASS</span></span>
<span><span class="co">#&gt; 4 Sample_4044          PASS        PASS        PASS        PASS</span></span>
<span><span class="co">#&gt; 5 Sample_4045          PASS        PASS        PASS        PASS</span></span>
<span><span class="co">## Check gene info</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">rse_gene</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt; DataFrame with 5 rows and 5 columns</span></span>
<span><span class="co">#&gt;                         Length            gencodeID          ensemblID</span></span>
<span><span class="co">#&gt;                      &lt;integer&gt;          &lt;character&gt;        &lt;character&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000102693.1      1070 ENSMUSG00000102693.1 ENSMUSG00000102693</span></span>
<span><span class="co">#&gt; ENSMUSG00000064842.1       110 ENSMUSG00000064842.1 ENSMUSG00000064842</span></span>
<span><span class="co">#&gt; ENSMUSG00000051951.5      6094 ENSMUSG00000051951.5 ENSMUSG00000051951</span></span>
<span><span class="co">#&gt; ENSMUSG00000102851.1       480 ENSMUSG00000102851.1 ENSMUSG00000102851</span></span>
<span><span class="co">#&gt; ENSMUSG00000103377.1      2819 ENSMUSG00000103377.1 ENSMUSG00000103377</span></span>
<span><span class="co">#&gt;                                 gene_type    EntrezID</span></span>
<span><span class="co">#&gt;                               &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; ENSMUSG00000102693.1                  TEC       71042</span></span>
<span><span class="co">#&gt; ENSMUSG00000064842.1                snRNA          NA</span></span>
<span><span class="co">#&gt; ENSMUSG00000051951.5       protein_coding      497097</span></span>
<span><span class="co">#&gt; ENSMUSG00000102851.1 processed_pseudogene   100418032</span></span>
<span><span class="co">#&gt; ENSMUSG00000103377.1                  TEC          NA</span></span>
<span><span class="co">## Access the original counts</span></span>
<span><span class="va">original_counts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">rse_gene</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span></span>
<span><span class="co">## Access the log normalized counts</span></span>
<span><span class="va">logcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">rse_gene</span><span class="op">)</span><span class="op">$</span><span class="va">logcounts</span></span>
<span></span>
<span><span class="co">## Explore complete rse_gene object (including smoking-exposed samples as well)</span></span>
<span><span class="va">rse_gene</span></span>
<span><span class="co">#&gt; class: RangedSummarizedExperiment </span></span>
<span><span class="co">#&gt; dim: 55401 208 </span></span>
<span><span class="co">#&gt; metadata(1): Obtained_from</span></span>
<span><span class="co">#&gt; assays(2): counts logcounts</span></span>
<span><span class="co">#&gt; rownames(55401): ENSMUSG00000102693.1 ENSMUSG00000064842.1 ...</span></span>
<span><span class="co">#&gt;   ENSMUSG00000064371.1 ENSMUSG00000064372.1</span></span>
<span><span class="co">#&gt; rowData names(13): Length gencodeID ... DE_in_pup_brain_nicotine</span></span>
<span><span class="co">#&gt;   DE_in_pup_brain_smoking</span></span>
<span><span class="co">#&gt; colnames: NULL</span></span>
<span><span class="co">#&gt; colData names(71): SAMPLE_ID FQCbasicStats ...</span></span>
<span><span class="co">#&gt;   retained_after_QC_sample_filtering</span></span>
<span><span class="co">#&gt;   retained_after_manual_sample_filtering</span></span>
<span></span>
<span><span class="co">#################################</span></span>
<span><span class="co">##   Extract data of interest</span></span>
<span><span class="co">#################################</span></span>
<span><span class="co">## Keep samples from nicotine experiment only</span></span>
<span><span class="va">rse_gene_nic</span> <span class="op">&lt;-</span> <span class="va">rse_gene</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">rse_gene</span><span class="op">$</span><span class="va">Expt</span> <span class="op">==</span> <span class="st">"Nicotine"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co">## New dimensions</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">rse_gene_nic</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 55401    65</span></span>
<span></span>
<span><span class="co">#################################</span></span>
<span><span class="co">##  Access gene expression data</span></span>
<span><span class="co">#################################</span></span>
<span></span>
<span><span class="co">## Raw counts for first 3 genes in the first 5 samples</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">rse_gene_nic</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                      [,1] [,2] [,3] [,4] [,5]</span></span>
<span><span class="co">#&gt; ENSMUSG00000102693.1    0    0    0    0    0</span></span>
<span><span class="co">#&gt; ENSMUSG00000064842.1    0    0    0    0    0</span></span>
<span><span class="co">#&gt; ENSMUSG00000051951.5  811  710  812 1168  959</span></span>
<span><span class="co">## Log-normalized counts for first 3 genes in the first 5 samples</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">rse_gene_nic</span><span class="op">)</span><span class="op">$</span><span class="va">logcounts</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                           [,1]      [,2]      [,3]      [,4]      [,5]</span></span>
<span><span class="co">#&gt; ENSMUSG00000102693.1 -5.985331 -5.985331 -5.985331 -5.985331 -5.985331</span></span>
<span><span class="co">#&gt; ENSMUSG00000064842.1 -5.985331 -5.985331 -5.985331 -5.985331 -5.985331</span></span>
<span><span class="co">#&gt; ENSMUSG00000051951.5  4.509114  4.865612  4.944597  4.684371  4.813427</span></span>
<span></span>
<span><span class="co">#################################</span></span>
<span><span class="co">##     Access sample data</span></span>
<span><span class="co">#################################</span></span>
<span></span>
<span><span class="co">## Info for the first 2 samples</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse_gene_nic</span><span class="op">)</span>, <span class="fl">2</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 2 rows and 71 columns</span></span>
<span><span class="co">#&gt;     SAMPLE_ID FQCbasicStats perBaseQual perTileQual  perSeqQual perBaseContent</span></span>
<span><span class="co">#&gt;   &lt;character&gt;   &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;    &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1 Sample_2914          PASS        PASS        PASS        PASS      FAIL/WARN</span></span>
<span><span class="co">#&gt; 2 Sample_4042          PASS        PASS        PASS        PASS      FAIL/WARN</span></span>
<span><span class="co">#&gt;     GCcontent    Ncontent SeqLengthDist SeqDuplication OverrepSeqs</span></span>
<span><span class="co">#&gt;   &lt;character&gt; &lt;character&gt;   &lt;character&gt;    &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1        WARN        PASS          WARN           FAIL        PASS</span></span>
<span><span class="co">#&gt; 2        WARN        PASS          WARN           FAIL        PASS</span></span>
<span><span class="co">#&gt;   AdapterContent KmerContent SeqLength_R1 percentGC_R1 phred15-19_R1</span></span>
<span><span class="co">#&gt;      &lt;character&gt; &lt;character&gt;  &lt;character&gt;  &lt;character&gt;   &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1           PASS          NA       75-151           49          37.0</span></span>
<span><span class="co">#&gt; 2           PASS          NA       75-151           48          37.0</span></span>
<span><span class="co">#&gt;   phred65-69_R1 phred115-119_R1 phred150-151_R1 phredGT30_R1 phredGT35_R1</span></span>
<span><span class="co">#&gt;     &lt;character&gt;     &lt;character&gt;     &lt;character&gt;    &lt;numeric&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1          37.0            37.0            37.0           NA           NA</span></span>
<span><span class="co">#&gt; 2          37.0            37.0            37.0           NA           NA</span></span>
<span><span class="co">#&gt;   Adapter65-69_R1 Adapter100-104_R1 Adapter140_R1 SeqLength_R2 percentGC_R2</span></span>
<span><span class="co">#&gt;         &lt;numeric&gt;         &lt;numeric&gt;     &lt;numeric&gt;  &lt;character&gt;  &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1     0.000108294        0.00026089    0.00174971       75-151           49</span></span>
<span><span class="co">#&gt; 2     0.000210067        0.00035278    0.00154716       75-151           48</span></span>
<span><span class="co">#&gt;   phred15-19_R2 phred65-69_R2 phred115-119_R2 phred150-151_R2 phredGT30_R2</span></span>
<span><span class="co">#&gt;     &lt;character&gt;   &lt;character&gt;     &lt;character&gt;     &lt;character&gt;    &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1          37.0          37.0            37.0            37.0           NA</span></span>
<span><span class="co">#&gt; 2          37.0          37.0            37.0            37.0           NA</span></span>
<span><span class="co">#&gt;   phredGT35_R2 Adapter65-69_R2 Adapter100-104_R2 Adapter140_R2 ERCCsumLogErr</span></span>
<span><span class="co">#&gt;      &lt;numeric&gt;       &lt;numeric&gt;         &lt;numeric&gt;     &lt;numeric&gt;     &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1           NA     0.000276104       0.000486875    0.00132235      -58.2056</span></span>
<span><span class="co">#&gt; 2           NA     0.000326771       0.000574851    0.00137044      -81.6359</span></span>
<span><span class="co">#&gt;                  bamFile   trimmed  numReads numMapped numUnmapped</span></span>
<span><span class="co">#&gt;              &lt;character&gt; &lt;logical&gt; &lt;numeric&gt; &lt;numeric&gt;   &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1 Sample_2914_sorted.bam     FALSE  89386472  87621022     1765450</span></span>
<span><span class="co">#&gt; 2 Sample_4042_sorted.bam     FALSE  59980794  58967812     1012982</span></span>
<span><span class="co">#&gt;   overallMapRate concordMapRate totalMapped mitoMapped  mitoRate</span></span>
<span><span class="co">#&gt;        &lt;numeric&gt;      &lt;numeric&gt;   &lt;numeric&gt;  &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1         0.9802         0.9748    87143087   10039739  0.103308</span></span>
<span><span class="co">#&gt; 2         0.9831         0.9709    58215746    7453208  0.113497</span></span>
<span><span class="co">#&gt;   totalAssignedGene  rRNA_rate      Tissue         Age         Sex        Expt</span></span>
<span><span class="co">#&gt;           &lt;numeric&gt;  &lt;numeric&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1          0.761378 0.00396315       Brain       Adult           F    Nicotine</span></span>
<span><span class="co">#&gt; 2          0.754444 0.00301119       Brain       Adult           F    Nicotine</span></span>
<span><span class="co">#&gt;          Group    Pregnant       plate    location concentration      medium</span></span>
<span><span class="co">#&gt;    &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;character&gt;   &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">#&gt; 1 Experimental           0      Plate2         C12         165.9       Water</span></span>
<span><span class="co">#&gt; 2      Control           0      Plate1          B4         122.6       Water</span></span>
<span><span class="co">#&gt;          date   Pregnancy    flowcell       sum  detected subsets_Mito_sum</span></span>
<span><span class="co">#&gt;   &lt;character&gt; &lt;character&gt; &lt;character&gt; &lt;numeric&gt; &lt;numeric&gt;        &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1          NA          No   HKCMHDSXX  37119948     24435          2649559</span></span>
<span><span class="co">#&gt; 2          NA          No   HKCG7DSXX  24904754     23656          1913803</span></span>
<span><span class="co">#&gt;   subsets_Mito_detected subsets_Mito_percent subsets_Ribo_sum</span></span>
<span><span class="co">#&gt;               &lt;numeric&gt;            &lt;numeric&gt;        &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; 1                    26              7.13783           486678</span></span>
<span><span class="co">#&gt; 2                    31              7.68449           319445</span></span>
<span><span class="co">#&gt;   subsets_Ribo_detected subsets_Ribo_percent retained_after_QC_sample_filtering</span></span>
<span><span class="co">#&gt;               &lt;numeric&gt;            &lt;numeric&gt;                          &lt;logical&gt;</span></span>
<span><span class="co">#&gt; 1                    11              1.31110                               TRUE</span></span>
<span><span class="co">#&gt; 2                    13              1.28267                               TRUE</span></span>
<span><span class="co">#&gt;   retained_after_manual_sample_filtering</span></span>
<span><span class="co">#&gt;                                &lt;logical&gt;</span></span>
<span><span class="co">#&gt; 1                                   TRUE</span></span>
<span><span class="co">#&gt; 2                                   TRUE</span></span></code></pre></div>
</div>
</div>
<div class="section level3">
<h3 id="differential-expression-analysis">Differential Expression Analysis<a class="anchor" aria-label="anchor" href="#differential-expression-analysis"></a>
</h3>
<p>In this workshop we’ll guide you through a differential expression
analysis following the steps presented in <strong>Figure 2</strong>. The
main objective is to determine which genes are affected by nicotine
administration (in adults) or exposure (in pups) in the mouse brain.</p>
<figure><img src="figs/Analyses.png" width="800px" align="center"><figcaption style="color: gray; line-height: 0.9; text-align: justify"><font size="-1.8"> <b>Figure 2</b>: <b> Summary of the analyses. </b>
Steps in yellow will be properly performed. 1. Data Preparation 2.
Exploratory Data Analysis 3. Differential Expression Analysis 4. DEG
visualization. </font> <font size="0.8"> Abbreviations: CPM: counts per
million; QC: quality control; PC: principal component; DEG:
differentially expressed genes.</font>
</figcaption></figure><div class="section level4">
<h4 class="unnumbered" id="data-preparation">
<strong>1. Data preparation</strong><a class="anchor" aria-label="anchor" href="#data-preparation"></a>
</h4>
<p>Before exploring a dataset, normalization and filtering steps are
generally appropriate. In our case, SPEAQeasy outputs contain normalized
counts, and the example data we’ll use already has non-expressed genes
filtered out.</p>
<div class="section level5">
<h5 class="unnumbered" id="data-normalization">1.1 Data normalization<a class="anchor" aria-label="anchor" href="#data-normalization"></a>
</h5>
<p>Data normalization is a preliminary step when working with expression
data because raw counts do not necessarily reflect real expression
measures of the genes, since there are technical differences in the way
the libraries are prepared and sequenced, as well as intrinsic
differences in the genes that are translated into more or less mapped
reads. Particularly, there are <em>within-sample effects</em> that are
the differences between genes in the same sample, such as their
<strong>length</strong> (the longer the gene, the more reads it will
have) and <strong>GC content</strong>, factors that contribute to
variations in their counts. On the other hand, <em>between-sample
effects</em> are differences between samples such as the
<strong>sequencing depth</strong>, i.e., the total number of molecules
sequenced, and the <strong>library size</strong>, i.e., the total number
of reads of each sample [1].</p>
<p>These variables lead to significantly different mRNA amounts but of
course are not due to the biological or treatment conditions of interest
(such as nicotine administration in this example) so in order to remove,
or at least, to minimize this technical bias and obtain measures
comparable and consistent across samples, raw counts must be normalized
by these factors. The data that we’ll use in this case are already
normalized in <code>assays(rse_gene)$logcounts</code>. Specifically, the
assay contains counts per million (CPM), also known as reads per million
(RPM), one basic gene expression unit that only normalizes by the
sequencing depth and is computed by dividing the read counts of a gene
in a sample by a scaling factor given by the total mapping reads of the
sample per million [2]:</p>
<p><span class="math display">\[CPM = \frac{read \ \ counts \ \ of \ \
gene \ \ \times \ \ 10^6 }{Total \ \ mapping \ \ reads \ \ of \ \
sample}\]</span></p>
<p>As outlined in <a href="#id_#Data%20overview%20and%20download"><strong>Data overview and
download</strong></a>, the scaling factors were obtained with
<code>calcNormFactors()</code> applying the Trimmed Mean of M-Values
(TMM) method, the <em><a href="https://bioconductor.org/packages/3.18/edgeR" class="external-link">edgeR</a></em>
package’s default normalization method that assumes that most genes are
not differentially expressed. The effective library sizes of the samples
and the CPM of each observation were computed with the <em><a href="https://bioconductor.org/packages/3.18/edgeR" class="external-link">edgeR</a></em>
function <code><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">cpm()</a></code> setting the <code>log</code> argument to
<code>TRUE</code> and <code>prior.count</code> to 0.5 to receive values
in <span class="math inline">\(log_2(CPM+0.5)\)</span>.</p>
<p>After data normalization and scaling, we’d expect the read counts to
follow a normal distribution, something we can confirm by comparing the
counts’ distribution before and after the normalization. Note that both
datasets contain the exact same genes.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Histogram and density plot of read counts before and after normalization</span></span>
<span></span>
<span><span class="co">## Raw counts</span></span>
<span><span class="va">counts_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">rse_gene_nic</span><span class="op">)</span><span class="op">$</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">counts_data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">counts</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="st">"black"</span>, fill <span class="op">=</span> <span class="st">"lightgray"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"read counts"</span>, y <span class="op">=</span> <span class="st">"Frecuency"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2.5</span>, <span class="fl">5</span>, <span class="fl">2.5</span>, <span class="fl">5</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">plot</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/Data%20preparation-1.png" width="700"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## Normalized counts</span></span>
<span><span class="va">logcounts_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>logcounts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">rse_gene_nic</span><span class="op">)</span><span class="op">$</span><span class="va">logcounts</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">logcounts_data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logcounts</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">..density..</span><span class="op">)</span>, colour <span class="op">=</span> <span class="st">"darkgray"</span>, fill <span class="op">=</span> <span class="st">"lightgray"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"#69b3a2"</span>, alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"log(CPM+0.5)"</span>, y <span class="op">=</span> <span class="st">"Frecuency"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2.5</span>, <span class="fl">5</span>, <span class="fl">2.5</span>, <span class="fl">5</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">plot</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/Data%20preparation-2.png" width="700"></p>
<p>As presented, after data transformation, we can now see a more
widespread distribution of the counts, but note that most of them are
zeros in the first plot (the one with the raw counts) and those zeros
remain after normalization, corresponding to counts below 0 in the
second plot. That is because we haven’t filtered the low and
zero-expressed genes.</p>
</div>
<div class="section level5">
<h5 class="unnumbered" id="gene-filtering">1.2 Gene filtering<a class="anchor" aria-label="anchor" href="#gene-filtering"></a>
</h5>
<p>Low-expressed or non-expressed genes in many samples are not of
biological interest in a study of differential expression because they
don’t inform about the gene expression changes and they are, by
definition, not differentially expressed, so we have to drop them using
<code>filterByExpr()</code> from <em><a href="https://bioconductor.org/packages/3.18/edgeR" class="external-link">edgeR</a></em> that
only keeps genes with at least <em>K</em> CPM in <em>n</em> samples and
with a minimum total number of counts across all samples.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Retain genes that passed filtering step</span></span>
<span><span class="va">rse_gene_filt</span> <span class="op">&lt;-</span> <span class="va">rse_gene_nic</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">rse_gene_nic</span><span class="op">)</span><span class="op">$</span><span class="va">retained_after_feature_filtering</span> <span class="op">==</span> <span class="cn">TRUE</span>, <span class="op">]</span></span>
<span></span>
<span><span class="co">## Normalized counts and filtered genes</span></span>
<span><span class="va">filt_logcounts_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>logcounts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">rse_gene_filt</span><span class="op">)</span><span class="op">$</span><span class="va">logcounts</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Plot</span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">filt_logcounts_data</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">logcounts</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_histogram.html" class="external-link">geom_histogram</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">..density..</span><span class="op">)</span>, colour <span class="op">=</span> <span class="st">"darkgray"</span>, fill <span class="op">=</span> <span class="st">"lightgray"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html" class="external-link">geom_density</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="st">"#69b3a2"</span>, alpha <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"log(CPM+0.5)"</span>, y <span class="op">=</span> <span class="st">"Frecuency"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2.5</span>, <span class="fl">5</span>, <span class="fl">2.5</span>, <span class="fl">5</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">plot</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-6-1.png" width="700"></p>
<p>In this third plot we can observe a curve that is closer (though not
completely) to a normal distribution and with less lowly-expressed
genes.</p>
<p>With the object <code>rse_gene_filt</code> we can proceed with
downstream analyses.</p>
</div>
</div>
<div class="section level4">
<h4 class="unnumbered" id="exploratory-data-analysis">
<strong>2. Exploratory Data
Analysis</strong><a class="anchor" aria-label="anchor" href="#exploratory-data-analysis"></a>
</h4>
<p>The first formal step that we will be performing is the sample
exploration. This crucial initial part of the analysis consists of an
examination of differences and relationships between Quality-Control
(QC) metrics of the samples from different groups in order to identify
poor-quality samples that must be removed before DEA. After that, we
will evaluate the similarities and grouping (not clustering) of the
samples by their gene expression variation through a dimensionality
reduction analysis. Finally, the sample variables in the metadata also
need to be analyzed and filtered based on the percentage of gene
expression variance that they explain for each gene.</p>
<div class="section level5">
<h5 class="unnumbered" id="quality-control-analysis">2.1 Quality Control Analysis<a class="anchor" aria-label="anchor" href="#quality-control-analysis"></a>
</h5>
<p>First we have to explore and compare the quality-control metrics of
the samples in the different groups given by covariates such as age,
sex, pregnancy state, group, plate and flowcell. See <a href="#id_##Sample%20Information"><strong>Sample
Information</strong></a> for a description of these variables.</p>
<style>
p.comment {
background-color: #DBDBDB;
padding: 10px;
border: 1px solid black;
border-radius: 5px;
font-family: sans-serif;
}

</style>
<p class="comment">
❓ <strong><em>Why is that relevant?</em></strong> As you could imagine,
technical and methodological differences in all the steps that were
carried out during the experimental stages are potential sources of
variation in the quality of the samples. Just imagine all that could
have been gone wrong or unexpected while experimenting with mice, during
the sampling, in the RNA extraction using different batches, when
treating samples in different mediums, when preparing libraries in
different plates and sequencing in different flowcells. Moreover, the
inherent features of the mice from which the samples come from such as
age, tissue, sex and pregnancy state could also affect the samples’
metrics if, for example, they were separately analyzed and processed.
</p>
<p class="comment">
❓<strong><em>But why do we care about mitochondrial and ribosomal
counts as QC metrics?</em></strong> In the process of mRNA extraction,
either by mRNA enrichment (capturing polyadenylated mRNAs) or
rRNA-depletion (removing rRNA), we’d expect to have a low number of
ribosomal counts, i.e., counts that map to rDNA, and if we don’t,
something must have gone wrong with the procedures. In the case of
mitochondrial counts something similar occurs: higher mitochondrial
rates will be obtained if the cytoplasmic mRNA capture was deficient or
if the transcripts were lost by some technical issue, increasing the
proportion of mitochondrial mRNAs. As a result, high
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> mitoRate</span> </mark> and
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> rRNA_rate</span> </mark> imply poor
quality in the samples.
</p>
<div class="alert alert-info">
<p><strong>Note:</strong> the QC metrics were computed with the
unprocessed datasets (neither filtered nor normalized) to preserve the
original estimates of the samples.</p>
</div>
<div class="section level6">
<h6 class="unnumbered" id="evaluate-qc-metrics-for-groups-of-samples">
<em>2.1.1 Evaluate QC metrics for groups of
samples</em><a class="anchor" aria-label="anchor" href="#evaluate-qc-metrics-for-groups-of-samples"></a>
</h6>
<p>Fortunately, we can identify to some extent possible factors that
could have influenced on the quality of the samples, as well as isolated
samples that are problematic. To do that, we will create boxplots that
present the distribution of the samples’ metrics separating them by
sample variables.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://hbiostat.org/R/Hmisc/" class="external-link">Hmisc</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://stringr.tidyverse.org" class="external-link">stringr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://wilkelab.org/cowplot/" class="external-link">cowplot</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define QC metrics of interest</span></span>
<span><span class="va">qc_metrics</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"mitoRate"</span>, <span class="st">"overallMapRate"</span>, <span class="st">"totalAssignedGene"</span>, <span class="st">"rRNA_rate"</span>, <span class="st">"sum"</span>, <span class="st">"detected"</span>, <span class="st">"ERCCsumLogErr"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Define sample variables of interest</span></span>
<span><span class="va">sample_variables</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Group"</span>, <span class="st">"Age"</span>, <span class="st">"Sex"</span>, <span class="st">"Pregnancy"</span>, <span class="st">"plate"</span>, <span class="st">"flowcell"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="co">## Function to create boxplots of QC metrics for groups of samples</span></span>
<span></span>
<span><span class="va">QC_boxplots</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">qc_metric</span>, <span class="va">sample_var</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co">## Define sample colors depending on the sample variable</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"Group"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span> <span class="op">=</span> <span class="st">"brown2"</span>, <span class="st">"Experimental"</span> <span class="op">=</span> <span class="st">"deepskyblue3"</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"Age"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Adult"</span> <span class="op">=</span> <span class="st">"slateblue3"</span>, <span class="st">"Pup"</span> <span class="op">=</span> <span class="st">"yellow3"</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"Sex"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"F"</span> <span class="op">=</span> <span class="st">"hotpink1"</span>, <span class="st">"M"</span> <span class="op">=</span> <span class="st">"dodgerblue"</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"Pregnancy"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Yes"</span> <span class="op">=</span> <span class="st">"darkorchid3"</span>, <span class="st">"No"</span> <span class="op">=</span> <span class="st">"darkolivegreen4"</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"plate"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Plate1"</span> <span class="op">=</span> <span class="st">"darkorange"</span>, <span class="st">"Plate2"</span> <span class="op">=</span> <span class="st">"lightskyblue"</span>, <span class="st">"Plate3"</span> <span class="op">=</span> <span class="st">"deeppink1"</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"flowcell"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>            <span class="st">"HKCG7DSXX"</span> <span class="op">=</span> <span class="st">"chartreuse2"</span>, <span class="st">"HKCMHDSXX"</span> <span class="op">=</span> <span class="st">"magenta"</span>, <span class="st">"HKCNKDSXX"</span> <span class="op">=</span> <span class="st">"turquoise3"</span>,</span>
<span>            <span class="st">"HKCTMDSXX"</span> <span class="op">=</span> <span class="st">"tomato"</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span></span>
<span>    <span class="co">## Axis labels</span></span>
<span>    <span class="va">x_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/Hmisc/man/capitalize.html" class="external-link">capitalize</a></span><span class="op">(</span><span class="va">sample_var</span><span class="op">)</span></span>
<span>    <span class="va">y_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace_all</a></span><span class="op">(</span><span class="va">qc_metric</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"_"</span> <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## x-axis text angle and position</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"flowcell"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">x_axis_angle</span> <span class="op">&lt;-</span> <span class="fl">18</span></span>
<span>        <span class="va">x_axis_hjust</span> <span class="op">&lt;-</span> <span class="fl">0.5</span></span>
<span>        <span class="va">x_axis_vjust</span> <span class="op">&lt;-</span> <span class="fl">0.7</span></span>
<span>        <span class="va">x_axis_size</span> <span class="op">&lt;-</span> <span class="fl">4</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="va">x_axis_angle</span> <span class="op">&lt;-</span> <span class="fl">0</span></span>
<span>        <span class="va">x_axis_hjust</span> <span class="op">&lt;-</span> <span class="fl">0.5</span></span>
<span>        <span class="va">x_axis_vjust</span> <span class="op">&lt;-</span> <span class="fl">0.5</span></span>
<span>        <span class="va">x_axis_size</span> <span class="op">&lt;-</span> <span class="fl">6</span></span>
<span>    <span class="op">}</span></span>
<span></span>
<span>    <span class="co">## Extract sample data in colData(rse_gene_filt)</span></span>
<span>    <span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse_gene_filt</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## Sample variable separating samples in x-axis and QC metric in y-axis</span></span>
<span>    <span class="co">## (Coloring by sample variable)</span></span>
<span>    <span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">data</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">sample_var</span><span class="op">)</span>, y <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">qc_metric</span><span class="op">)</span>, color <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">sample_var</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="co">## Add violin plots</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_violin.html" class="external-link">geom_violin</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0</span>, size <span class="op">=</span> <span class="fl">0.4</span>, color <span class="op">=</span> <span class="st">"black"</span>, width <span class="op">=</span> <span class="fl">0.7</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="co">## Spread dots</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.08</span>, alpha <span class="op">=</span> <span class="fl">0.7</span>, size <span class="op">=</span> <span class="fl">1.3</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="co">## Add boxplots</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0</span>, size <span class="op">=</span> <span class="fl">0.4</span>, width <span class="op">=</span> <span class="fl">0.1</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="co">## Set colors</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">colors</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="co">## Define axis labels</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="va">y_label</span>, x <span class="op">=</span> <span class="va">x_label</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="co">## Get rid of the background</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="co">## Hide legend and define plot margins and size of axis title and text</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>            legend.position <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>            plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.4</span>, <span class="fl">0.5</span>, <span class="fl">0.4</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>            axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">7</span><span class="op">)</span>,</span>
<span>            axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="va">x_axis_size</span><span class="op">)</span>,</span>
<span>            axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="va">x_axis_angle</span>, hjust <span class="op">=</span> <span class="va">x_axis_hjust</span>, vjust <span class="op">=</span> <span class="va">x_axis_vjust</span><span class="op">)</span></span>
<span>        <span class="op">)</span></span>
<span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">plot</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co">## Plots of all QC metrics for each sample variable</span></span>
<span><span class="va">multiple_QC_boxplots</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">sample_var</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">i</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span>    <span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">qc_metric</span> <span class="kw">in</span> <span class="va">qc_metrics</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="co">## Call function to create each individual plot</span></span>
<span>        <span class="va">plots</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu">QC_boxplots</span><span class="op">(</span><span class="va">qc_metric</span>, <span class="va">sample_var</span><span class="op">)</span></span>
<span>        <span class="va">i</span> <span class="op">&lt;-</span> <span class="va">i</span> <span class="op">+</span> <span class="fl">1</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="co">## Arrange multiple plots into a grid</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">plots</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span>, <span class="va">plots</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span>, <span class="va">plots</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span>, <span class="va">plots</span><span class="op">[[</span><span class="fl">4</span><span class="op">]</span><span class="op">]</span>, <span class="va">plots</span><span class="op">[[</span><span class="fl">5</span><span class="op">]</span><span class="op">]</span>, <span class="va">plots</span><span class="op">[[</span><span class="fl">6</span><span class="op">]</span><span class="op">]</span>, <span class="va">plots</span><span class="op">[[</span><span class="fl">7</span><span class="op">]</span><span class="op">]</span>, nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">multiple_QC_boxplots</span><span class="op">(</span><span class="st">"Age"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-7-1.png" width="700"></p>
<p>Initially, when we separate samples by Age, we can appreciate a clear
segregation of adult and pup samples in
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> mitoRate</span> </mark>, with higher
mitochondrial rates for adult samples and thus, being lower quality
samples than the pup ones. We can also see that pup samples have higher
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> totalAssignedGene</span> </mark>, again
being higher quality. The samples are very similar in the rest of the QC
metrics. The former differences must be taken into account because they
guide further sample separation by Age, which is necessary to avoid
dropping most of the adult samples (that are lower quality) in the
<strong>QC-based sample filtering</strong> (see below) and to prevent
misinterpreting sample variation given by quality and not by mouse age
itself.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">multiple_QC_boxplots</span><span class="op">(</span><span class="st">"Group"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>Notably, for Group no evident contrasts are seen in the quality of
the samples, which means that both control and exposed samples have
similar metrics and therefore the differences between them won’t be
determined by technical factors but effectively by gene expression
changes.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">multiple_QC_boxplots</span><span class="op">(</span><span class="st">"plate"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
<p>In plate, no alarming differences are presented but some samples from
the 1st plate have low (though not much lower)
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> overallMapRate</span> </mark>,
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> totalAssignedGene</span> </mark> and
<mark style="background-color: #DFF0FE"><span style="font-family: monospace"> sum</span> </mark>.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">multiple_QC_boxplots</span><span class="op">(</span><span class="st">"flowcell"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-10-1.png" width="700"></p>
<p>For the flowcell, again no worrying distinctions are seen, with the
exception of a few individual samples far from the rest.</p>
</div>
<div class="section level6">
<h6 class="unnumbered" id="qc-sample-filtering">
<em>2.1.2 QC sample filtering</em><a class="anchor" aria-label="anchor" href="#qc-sample-filtering"></a>
</h6>
<p>After assessing how different or similar are the QC values between
samples, we can now proceed to sample filtering based precisely, on
these metrics. For that, we will use <code><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier()</a></code> from
<em><a href="https://bioconductor.org/packages/3.18/scater" class="external-link">scater</a></em> to
identify outlier samples only at the lower end or the higher end,
depending on the QC metric.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rlang.r-lib.org" class="external-link">rlang</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/slowkow/ggrepel" class="external-link">ggrepel</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Separate data by Age</span></span>
<span><span class="va">rse_gene_pups</span> <span class="op">&lt;-</span> <span class="va">rse_gene_filt</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">rse_gene_filt</span><span class="op">$</span><span class="va">Age</span> <span class="op">==</span> <span class="st">"Pup"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">rse_gene_adults</span> <span class="op">&lt;-</span> <span class="va">rse_gene_filt</span><span class="op">[</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">rse_gene_filt</span><span class="op">$</span><span class="va">Age</span> <span class="op">==</span> <span class="st">"Adult"</span><span class="op">)</span><span class="op">]</span></span>
<span></span>
<span></span>
<span><span class="co">## Find outlier samples based on their QC metrics (samples that are 3 median-absolute-deviations away from the median)</span></span>
<span></span>
<span></span>
<span><span class="co">## Filter all samples together</span></span>
<span></span>
<span><span class="co">## Drop samples with lower library sizes (sum), detected number of genes and totalAssignedGene</span></span>
<span><span class="va">outliers_library_size</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">rse_gene_filt</span><span class="op">$</span><span class="va">sum</span>, nmads <span class="op">=</span> <span class="fl">3</span>, type <span class="op">=</span> <span class="st">"lower"</span><span class="op">)</span></span>
<span><span class="va">outliers_detected_num</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">rse_gene_filt</span><span class="op">$</span><span class="va">detected</span>, nmads <span class="op">=</span> <span class="fl">3</span>, type <span class="op">=</span> <span class="st">"lower"</span><span class="op">)</span></span>
<span><span class="va">outliers_totalAssignedGene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">rse_gene_filt</span><span class="op">$</span><span class="va">totalAssignedGene</span>, nmads <span class="op">=</span> <span class="fl">3</span>, type <span class="op">=</span> <span class="st">"lower"</span><span class="op">)</span></span>
<span><span class="co">## Drop samples with higher mitoRates and rRNA rates</span></span>
<span><span class="va">outliers_mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">rse_gene_filt</span><span class="op">$</span><span class="va">mitoRate</span>, nmads <span class="op">=</span> <span class="fl">3</span>, type <span class="op">=</span> <span class="st">"higher"</span><span class="op">)</span></span>
<span><span class="va">outliers_rRNArate</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">rse_gene_filt</span><span class="op">$</span><span class="va">rRNA_rate</span>, nmads <span class="op">=</span> <span class="fl">3</span>, type <span class="op">=</span> <span class="st">"higher"</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Keep not outlier samples</span></span>
<span><span class="va">not_outliers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span><span class="op">(</span><span class="va">outliers_library_size</span> <span class="op">|</span> <span class="va">outliers_detected_num</span> <span class="op">|</span> <span class="va">outliers_totalAssignedGene</span> <span class="op">|</span> <span class="va">outliers_mito</span> <span class="op">|</span> <span class="va">outliers_rRNArate</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rse_gene_qc</span> <span class="op">&lt;-</span> <span class="va">rse_gene_filt</span><span class="op">[</span>, <span class="va">not_outliers</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Number of samples retained</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">rse_gene_qc</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] 39</span></span>
<span></span>
<span><span class="co">## Add new variables to rse_gene_filt with info of samples retained/dropped</span></span>
<span><span class="va">rse_gene_filt</span><span class="op">$</span><span class="va">Retention_after_QC_filtering</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">rse_gene_filt</span><span class="op">$</span><span class="va">SAMPLE_ID</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">x</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">rse_gene_qc</span><span class="op">$</span><span class="va">SAMPLE_ID</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="st">"Retained"</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="st">"Dropped"</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co">## Filter adult samples</span></span>
<span></span>
<span><span class="va">outliers_library_size</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">rse_gene_adults</span><span class="op">$</span><span class="va">sum</span>, nmads <span class="op">=</span> <span class="fl">3</span>, type <span class="op">=</span> <span class="st">"lower"</span><span class="op">)</span></span>
<span><span class="va">outliers_detected_num</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">rse_gene_adults</span><span class="op">$</span><span class="va">detected</span>, nmads <span class="op">=</span> <span class="fl">3</span>, type <span class="op">=</span> <span class="st">"lower"</span><span class="op">)</span></span>
<span><span class="va">outliers_totalAssignedGene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">rse_gene_adults</span><span class="op">$</span><span class="va">totalAssignedGene</span>, nmads <span class="op">=</span> <span class="fl">3</span>, type <span class="op">=</span> <span class="st">"lower"</span><span class="op">)</span></span>
<span><span class="va">outliers_mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">rse_gene_adults</span><span class="op">$</span><span class="va">mitoRate</span>, nmads <span class="op">=</span> <span class="fl">3</span>, type <span class="op">=</span> <span class="st">"higher"</span><span class="op">)</span></span>
<span><span class="va">outliers_rRNArate</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">rse_gene_adults</span><span class="op">$</span><span class="va">rRNA_rate</span>, nmads <span class="op">=</span> <span class="fl">3</span>, type <span class="op">=</span> <span class="st">"higher"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">not_outliers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span><span class="op">(</span><span class="va">outliers_library_size</span> <span class="op">|</span> <span class="va">outliers_detected_num</span> <span class="op">|</span> <span class="va">outliers_totalAssignedGene</span> <span class="op">|</span> <span class="va">outliers_mito</span> <span class="op">|</span> <span class="va">outliers_rRNArate</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rse_gene_adults_qc</span> <span class="op">&lt;-</span> <span class="va">rse_gene_adults</span><span class="op">[</span>, <span class="va">not_outliers</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Number of samples retained</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">rse_gene_adults_qc</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] 20</span></span>
<span></span>
<span><span class="va">rse_gene_adults</span><span class="op">$</span><span class="va">Retention_after_QC_filtering</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">rse_gene_adults</span><span class="op">$</span><span class="va">SAMPLE_ID</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">x</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">rse_gene_adults_qc</span><span class="op">$</span><span class="va">SAMPLE_ID</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="st">"Retained"</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="st">"Dropped"</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span></span>
<span></span>
<span><span class="co">## Filter pup samples</span></span>
<span></span>
<span><span class="va">outliers_library_size</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">rse_gene_pups</span><span class="op">$</span><span class="va">sum</span>, nmads <span class="op">=</span> <span class="fl">3</span>, type <span class="op">=</span> <span class="st">"lower"</span><span class="op">)</span></span>
<span><span class="va">outliers_detected_num</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">rse_gene_pups</span><span class="op">$</span><span class="va">detected</span>, nmads <span class="op">=</span> <span class="fl">3</span>, type <span class="op">=</span> <span class="st">"lower"</span><span class="op">)</span></span>
<span><span class="va">outliers_totalAssignedGene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">rse_gene_pups</span><span class="op">$</span><span class="va">totalAssignedGene</span>, nmads <span class="op">=</span> <span class="fl">3</span>, type <span class="op">=</span> <span class="st">"lower"</span><span class="op">)</span></span>
<span><span class="va">outliers_mito</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">rse_gene_pups</span><span class="op">$</span><span class="va">mitoRate</span>, nmads <span class="op">=</span> <span class="fl">3</span>, type <span class="op">=</span> <span class="st">"higher"</span><span class="op">)</span></span>
<span><span class="va">outliers_rRNArate</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/isOutlier.html" class="external-link">isOutlier</a></span><span class="op">(</span><span class="va">rse_gene_pups</span><span class="op">$</span><span class="va">rRNA_rate</span>, nmads <span class="op">=</span> <span class="fl">3</span>, type <span class="op">=</span> <span class="st">"higher"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">not_outliers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="op">!</span><span class="op">(</span><span class="va">outliers_library_size</span> <span class="op">|</span> <span class="va">outliers_detected_num</span> <span class="op">|</span> <span class="va">outliers_totalAssignedGene</span> <span class="op">|</span> <span class="va">outliers_mito</span> <span class="op">|</span> <span class="va">outliers_rRNArate</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">rse_gene_pups_qc</span> <span class="op">&lt;-</span> <span class="va">rse_gene_pups</span><span class="op">[</span>, <span class="va">not_outliers</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Number of samples retained</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">rse_gene_pups_qc</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span></span>
<span><span class="co">#&gt; [1] 41</span></span>
<span></span>
<span><span class="va">rse_gene_pups</span><span class="op">$</span><span class="va">Retention_after_QC_filtering</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.vector.html" class="external-link">as.vector</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">rse_gene_pups</span><span class="op">$</span><span class="va">SAMPLE_ID</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">x</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">rse_gene_pups_qc</span><span class="op">$</span><span class="va">SAMPLE_ID</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="st">"Retained"</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="st">"Dropped"</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">}</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>We already filtered outlier samples … but what have we removed? It is
always important to trace the QC metrics of the filtered samples to
verify that they really are poor quality. We don’t want to get rid of
useful samples! So let’s go back to the QC boxplots but now color
samples according to whether they passed the filtering step, and also
distinguishing samples’ groups by shape.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Boxplots of QC metrics after sample filtering</span></span>
<span></span>
<span><span class="co">## Boxplots</span></span>
<span><span class="va">boxplots_after_QC_filtering</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">rse_gene</span>, <span class="va">qc_metric</span>, <span class="va">sample_var</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co">## Color samples</span></span>
<span>    <span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Retained"</span> <span class="op">=</span> <span class="st">"deepskyblue"</span>, <span class="st">"Dropped"</span> <span class="op">=</span> <span class="st">"brown2"</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## Sample shape by sample variables</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"Group"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">shapes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span> <span class="op">=</span> <span class="fl">0</span>, <span class="st">"Experimental"</span> <span class="op">=</span> <span class="fl">15</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"Age"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">shapes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Adult"</span> <span class="op">=</span> <span class="fl">16</span>, <span class="st">"Pup"</span> <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"Sex"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">shapes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"F"</span> <span class="op">=</span> <span class="fl">11</span>, <span class="st">"M"</span> <span class="op">=</span> <span class="fl">19</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"Pregnancy"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">shapes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Yes"</span> <span class="op">=</span> <span class="fl">10</span>, <span class="st">"No"</span> <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"plate"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">shapes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Plate1"</span> <span class="op">=</span> <span class="fl">12</span>, <span class="st">"Plate2"</span> <span class="op">=</span> <span class="fl">5</span>, <span class="st">"Plate3"</span> <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"flowcell"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">shapes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>            <span class="st">"HKCG7DSXX"</span> <span class="op">=</span> <span class="fl">3</span>, <span class="st">"HKCMHDSXX"</span> <span class="op">=</span> <span class="fl">8</span>, <span class="st">"HKCNKDSXX"</span> <span class="op">=</span> <span class="fl">14</span>,</span>
<span>            <span class="st">"HKCTMDSXX"</span> <span class="op">=</span> <span class="fl">17</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span></span>
<span>    <span class="va">y_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stringr.tidyverse.org/reference/str_replace.html" class="external-link">str_replace_all</a></span><span class="op">(</span><span class="va">qc_metric</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"_"</span> <span class="op">=</span> <span class="st">" "</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse_gene</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## Median of the QC var values</span></span>
<span>    <span class="va">median</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/median.html" class="external-link">median</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/eval.html" class="external-link">eval</a></span><span class="op">(</span><span class="fu"><a href="https://rlang.r-lib.org/reference/parse_expr.html" class="external-link">parse_expr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"rse_gene$"</span>, <span class="va">qc_metric</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="co">## Median-absolute-deviation of the QC var values</span></span>
<span>    <span class="va">mad</span> <span class="op">&lt;-</span> <span class="fu">mad</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/eval.html" class="external-link">eval</a></span><span class="op">(</span><span class="fu"><a href="https://rlang.r-lib.org/reference/parse_expr.html" class="external-link">parse_expr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"rse_gene$"</span>, <span class="va">qc_metric</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">data</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>        x <span class="op">=</span> <span class="st">""</span>, y <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">qc_metric</span><span class="op">)</span>,</span>
<span>        color <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="st">"Retention_after_QC_filtering"</span><span class="op">)</span></span>
<span>    <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">1</span>, size <span class="op">=</span> <span class="fl">2</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/eval.html" class="external-link">eval</a></span><span class="op">(</span><span class="fu"><a href="https://rlang.r-lib.org/reference/parse_expr.html" class="external-link">parse_expr</a></span><span class="op">(</span><span class="op">(</span><span class="va">sample_var</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>alpha <span class="op">=</span> <span class="fl">0</span>, size <span class="op">=</span> <span class="fl">0.15</span>, color <span class="op">=</span> <span class="st">"black"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">colors</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_shape_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">shapes</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">""</span>, y <span class="op">=</span> <span class="va">y_label</span>, color <span class="op">=</span> <span class="st">"Retention after QC filtering"</span>, shape <span class="op">=</span> <span class="va">sample_var</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="co">## Median line</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">median</span>, size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="co">## Line of median + 3 MADs</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">median</span> <span class="op">+</span> <span class="op">(</span><span class="fl">3</span> <span class="op">*</span> <span class="va">mad</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.5</span>, linetype <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="co">## Line of median - 3 MADs</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span>yintercept <span class="op">=</span> <span class="va">median</span> <span class="op">-</span> <span class="op">(</span><span class="fl">3</span> <span class="op">*</span> <span class="va">mad</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">0.5</span>, linetype <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>            axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="op">(</span><span class="fl">9</span><span class="op">)</span><span class="op">)</span>,</span>
<span>            axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="op">(</span><span class="fl">8</span><span class="op">)</span><span class="op">)</span>,</span>
<span>            legend.position <span class="op">=</span> <span class="st">"right"</span>,</span>
<span>            legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">8</span><span class="op">)</span>,</span>
<span>            legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">9</span><span class="op">)</span></span>
<span>        <span class="op">)</span></span>
<span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">plot</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>In the following plot we can confirm that taking all samples together
(both adults and pups) all adult samples are dropped by their high
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> mitoRate</span> </mark>, but that is not
desirable because we want to analyze these samples too, so we need to
analyze pup and adult samples separately.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Plots</span></span>
<span></span>
<span><span class="co">## All samples together</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">boxplots_after_QC_filtering</span><span class="op">(</span><span class="va">rse_gene_filt</span>, <span class="st">"mitoRate"</span>, <span class="st">"Age"</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span>, <span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-12-1.png" width="700">
For adults, only 3 controls are dropped, 3 of them by their
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> mitoRate</span> </mark>.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Adult samples</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">boxplots_after_QC_filtering</span><span class="op">(</span><span class="va">rse_gene_adults</span>, <span class="st">"mitoRate"</span>, <span class="st">"Group"</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span>, <span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-13-1.png" width="700"></p>
<p>For pups, 1 experimental sample was dropped by its
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> rRNA_rate</span> </mark>.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Pup samples</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">boxplots_after_QC_filtering</span><span class="op">(</span><span class="va">rse_gene_pups</span>, <span class="st">"rRNA_rate"</span>, <span class="st">"Group"</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span>, <span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-14-1.png" width="700"><div class="alert alert-info">
<p>Try the function yourself with different QC metrics and sample
variables!</p>
</div>
</div>
</div>
<div class="section level5">
<h5 class="unnumbered" id="explore-sample-level-effects">2.2 Explore sample-level effects<a class="anchor" aria-label="anchor" href="#explore-sample-level-effects"></a>
</h5>
<p>Once data is normalized and properly filtered (features and samples),
the next step is to explore the variation in gene expression of the
samples to identify if a covariate explains a high proportion of the
variance in the expression (if samples are separated in a PC when
colored by that covariate). In few words, if a variable represents big
differences in the expression values of the samples.</p>
<p>We are gonna create some functions to do it.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">calc_PCA</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">rse</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co">## Obtain PCAs</span></span>
<span>    <span class="va">pca_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/prcomp.html" class="external-link">prcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">$</span><span class="va">logcounts</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## Join PCs and samples' info</span></span>
<span>    <span class="va">pca_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="va">pca_all</span><span class="op">$</span><span class="va">x</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">pca_data</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://rpkgs.datanovia.com/ggpubr/" class="external-link">"ggpubr"</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot_PCAs</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">sample_var</span>, <span class="va">pca_data</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co">## Define sample colors depending on the sample variable</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"Group"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span> <span class="op">=</span> <span class="st">"brown2"</span>, <span class="st">"Experimental"</span> <span class="op">=</span> <span class="st">"deepskyblue3"</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"Age"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Adult"</span> <span class="op">=</span> <span class="st">"slateblue3"</span>, <span class="st">"Pup"</span> <span class="op">=</span> <span class="st">"yellow3"</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"Sex"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"F"</span> <span class="op">=</span> <span class="st">"hotpink1"</span>, <span class="st">"M"</span> <span class="op">=</span> <span class="st">"dodgerblue"</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"Pregnancy"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Yes"</span> <span class="op">=</span> <span class="st">"darkorchid3"</span>, <span class="st">"No"</span> <span class="op">=</span> <span class="st">"darkolivegreen4"</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"plate"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Plate1"</span> <span class="op">=</span> <span class="st">"darkorange"</span>, <span class="st">"Plate2"</span> <span class="op">=</span> <span class="st">"lightskyblue"</span>, <span class="st">"Plate3"</span> <span class="op">=</span> <span class="st">"deeppink1"</span><span class="op">)</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var</span> <span class="op">==</span> <span class="st">"flowcell"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>            <span class="st">"HKCG7DSXX"</span> <span class="op">=</span> <span class="st">"chartreuse2"</span>, <span class="st">"HKCMHDSXX"</span> <span class="op">=</span> <span class="st">"magenta"</span>, <span class="st">"HKCNKDSXX"</span> <span class="op">=</span> <span class="st">"turquoise3"</span>,</span>
<span>            <span class="st">"HKCTMDSXX"</span> <span class="op">=</span> <span class="st">"tomato"</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span></span>
<span>    <span class="va">plot1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">pca_data</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">PC1</span>, <span class="va">PC2</span>, color <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">sample_var</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">colors</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">plot2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">pca_data</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">PC2</span>, <span class="va">PC3</span>, color <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">sample_var</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">colors</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">plot3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">pca_data</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">PC3</span>, <span class="va">PC4</span>, color <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">sample_var</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">colors</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">wplot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rpkgs.datanovia.com/ggpubr/reference/ggarrange.html" class="external-link">ggarrange</a></span><span class="op">(</span><span class="va">plot1</span>, <span class="va">plot2</span>, <span class="va">plot3</span>, ncol <span class="op">=</span> <span class="fl">3</span>, common.legend <span class="op">=</span> <span class="cn">TRUE</span>, legend <span class="op">=</span> <span class="st">"right"</span><span class="op">)</span></span>
<span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">wplot</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sample_variables</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Group"</span>, <span class="st">"Age"</span>, <span class="st">"Sex"</span>, <span class="st">"Pregnancy"</span>, <span class="st">"plate"</span>, <span class="st">"flowcell"</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">pca_data</span> <span class="op">&lt;-</span> <span class="fu">calc_PCA</span><span class="op">(</span><span class="va">rse_gene_filt</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pca_plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">sample_variables</span>, <span class="va">plot_PCAs</span>, pca_data <span class="op">=</span> <span class="va">pca_data</span><span class="op">)</span></span>
<span><span class="va">pca_plots</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-17-1.png" width="700" style="display: block; margin: auto;"><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-17-2.png" width="700" style="display: block; margin: auto;"><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-17-3.png" width="700" style="display: block; margin: auto;"><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-17-4.png" width="700" style="display: block; margin: auto;"><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-17-5.png" width="700" style="display: block; margin: auto;"><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-17-6.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_data</span> <span class="op">&lt;-</span> <span class="fu">calc_PCA</span><span class="op">(</span><span class="va">rse_gene_pups_qc</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pca_plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">sample_variables</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span><span class="op">]</span>, <span class="va">plot_PCAs</span>, pca_data <span class="op">=</span> <span class="va">pca_data</span><span class="op">)</span></span>
<span><span class="va">pca_plots</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-18-1.png" width="700" style="display: block; margin: auto;"><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-18-2.png" width="700" style="display: block; margin: auto;"><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-18-3.png" width="700" style="display: block; margin: auto;"><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-18-4.png" width="700" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pca_data</span> <span class="op">&lt;-</span> <span class="fu">calc_PCA</span><span class="op">(</span><span class="va">rse_gene_adults_qc</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pca_plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">sample_variables</span><span class="op">[</span><span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span><span class="op">]</span>, <span class="va">plot_PCAs</span>, pca_data <span class="op">=</span> <span class="va">pca_data</span><span class="op">)</span></span>
<span><span class="va">pca_plots</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-19-1.png" width="700" style="display: block; margin: auto;"><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-19-2.png" width="700" style="display: block; margin: auto;"><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-19-3.png" width="700" style="display: block; margin: auto;"><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-19-4.png" width="700" style="display: block; margin: auto;"></p>
</div>
<div class="section level5">
<h5 class="unnumbered" id="explore-gene-level-effects">2.3 Explore gene-level effects<a class="anchor" aria-label="anchor" href="#explore-gene-level-effects"></a>
</h5>
<p>Once the quality and the variability of the samples have been
evaluated, the next step is to explore the differences in the expression
of the genes themselves in the sample groups, or in other words, to
quantify the contribution of the multiple sample variables of the study
in the gene expression variation, which constitutes one of the
fundamental challenges when analyzing complex RNA-seq datasets [3].</p>
<p>To determine which variables are the major drivers of expression
variability, and importantly to define if the technical variability of
RNA-seq data is low enough to study nicotine effects, we can implement
an analysis of variance partition. <code>variancePartition</code> is a
package that decomposes for each gene the expression variation into
fractions of variance explained (FVE) by the sample variables of the
experimental design of high-throughput genomics studies [3].</p>
<div class="section level6">
<h6 class="unnumbered" id="canonical-correlation-analysis">
<em>2.3.1 Canonical Correlation
Analysis</em><a class="anchor" aria-label="anchor" href="#canonical-correlation-analysis"></a>
</h6>
<p>Before the analysis itself, we need to measure the correlation
between the sample variables. This is an important step because highly
correlated variables can produce unstable estimates of the variance
fractions and hinder the identification of the variables that really
contribute to the expression variation. There are at least two problems
with that:</p>
<ul>
<li>If two variables are correlated, we could incorrectly determine that
one of them contributes to gene expression changes when in reality it
was just correlated with a real contributory variable.</li>
<li>The part of variance explained by a biologically relevant variable
can be reduced by the apparent contributions of correlated variables, if
for example, they contain very similar information.</li>
</ul>
<p>Additionally, the analysis is better performed with simpler models,
specially when we have a limited number of samples in the study.</p>
<p>Hence, to drop such variables we must identify them first. Pearson
correlation can be used for comparing continuous variables but the
models can contain categorical variables as well, so in order to obtain
the correlation between a continuous and a categorical variable, or
between two categorical variables, we will perform a Canonical
Correlation Analysis (CCA) with <code><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/canCorPairs.html" class="external-link">canCorPairs()</a></code> that assesses
the degree to which the variables co-vary and contain the same
information. This function returns rho / sum(rho), the fraction of the
maximum possible correlation. Note that CCA returns correlations values
between 0 and 1 [4].</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/variancePartition" class="external-link">variancePartition</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">pheatmap</span><span class="op">)</span></span>
<span></span>
<span><span class="co">#######################   Variance Partition Analysis   #######################</span></span>
<span></span>
<span><span class="co">## Fraction of variation attributable to each variable after correcting for all other variables</span></span>
<span></span>
<span></span>
<span><span class="co">## 1. Canonical Correlation Analysis (CCA)</span></span>
<span></span>
<span><span class="co">## Asses the correlation between each pair of sample variables</span></span>
<span></span>
<span><span class="co">## Plot heatmap of correlations</span></span>
<span><span class="va">plot_CCA</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">age</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co">## Data</span></span>
<span>    <span class="va">rse_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/eval.html" class="external-link">eval</a></span><span class="op">(</span><span class="fu"><a href="https://rlang.r-lib.org/reference/parse_expr.html" class="external-link">parse_expr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"rse_gene_"</span>, <span class="va">age</span>, <span class="st">"_qc"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## Define variables to examine: remove those with single values</span></span>
<span>    <span class="co">## For adults: all are females (so we drop 'Sex' variable)</span></span>
<span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">age</span> <span class="op">==</span> <span class="st">"adults"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">formula</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="va">Group</span> <span class="op">+</span> <span class="va">Pregnancy</span> <span class="op">+</span> <span class="va">plate</span> <span class="op">+</span> <span class="va">flowcell</span> <span class="op">+</span> <span class="va">mitoRate</span> <span class="op">+</span> <span class="va">overallMapRate</span> <span class="op">+</span> <span class="va">totalAssignedGene</span> <span class="op">+</span> <span class="va">rRNA_rate</span> <span class="op">+</span> <span class="va">sum</span> <span class="op">+</span> <span class="va">detected</span> <span class="op">+</span> <span class="va">ERCCsumLogErr</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="co">## For pups: none is pregnant (so 'Pregnancy' variable is not considered)</span></span>
<span>    <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="va">formula</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="va">Group</span> <span class="op">+</span> <span class="va">Sex</span> <span class="op">+</span> <span class="va">plate</span> <span class="op">+</span> <span class="va">flowcell</span> <span class="op">+</span> <span class="va">mitoRate</span> <span class="op">+</span> <span class="va">overallMapRate</span> <span class="op">+</span> <span class="va">totalAssignedGene</span> <span class="op">+</span> <span class="va">rRNA_rate</span> <span class="op">+</span> <span class="va">sum</span> <span class="op">+</span> <span class="va">detected</span> <span class="op">+</span> <span class="va">ERCCsumLogErr</span></span>
<span>    <span class="op">}</span></span>
<span></span>
<span>    <span class="co">## Measure correlations</span></span>
<span>    <span class="va">C</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/canCorPairs.html" class="external-link">canCorPairs</a></span><span class="op">(</span><span class="va">formula</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse_gene</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="co">## Heatmap</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/pheatmap/man/pheatmap.html" class="external-link">pheatmap</a></span><span class="op">(</span></span>
<span>        <span class="va">C</span>, <span class="co">## data</span></span>
<span>        color <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grDevices/palettes.html" class="external-link">hcl.colors</a></span><span class="op">(</span><span class="fl">50</span>, <span class="st">"YlOrRd"</span>, rev <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>, <span class="co">## color scale</span></span>
<span>        fontsize <span class="op">=</span> <span class="fl">8</span>, <span class="co">## text size</span></span>
<span>        border_color <span class="op">=</span> <span class="st">"black"</span>, <span class="co">## border color for heatmap cells</span></span>
<span>        cellwidth <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0.4</span>, <span class="st">"cm"</span><span class="op">)</span>, <span class="co">## height of cells</span></span>
<span>        cellheight <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fl">0.4</span>, <span class="st">"cm"</span><span class="op">)</span> <span class="co">## width of cells</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">C</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>As you can see, for adults there’s a strong correlation between
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> mitoRate</span> </mark> and
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> totalAssignedGene</span> </mark>.
There’s also a strong correlation between
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> plate</span> </mark> and
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> flowcell </span> </mark> and between
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> plate</span> </mark> and
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> overallMapRate</span> </mark>.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Heatmap for adult samples</span></span>
<span><span class="va">CCA_adults</span> <span class="op">&lt;-</span> <span class="fu">plot_CCA</span><span class="op">(</span><span class="st">"adults"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-21-1.png" width="700"></p>
<p>In pups <mark style="background-color: #FCF3CF"><span style="font-family: monospace"> overallMapRate</span> </mark> has a
relationship with <mark style="background-color: #FCF3CF"><span style="font-family: monospace"> rRNA_rate</span> </mark>,
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> plate</span> </mark> and
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> flowcell</span> </mark>.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Heatmap for pup samples</span></span>
<span><span class="va">CCA_pups</span> <span class="op">&lt;-</span> <span class="fu">plot_CCA</span><span class="op">(</span><span class="st">"pups"</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-22-1.png" width="700"></p>
<p>Importantly, in both groups of age
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> Group</span> </mark> is not highly
correlated with any other variable. This is desirable because if this
variable, the one that separates experimental and control samples, were
correlated with another, then its individual contribution to gene
expression changes would be diminished, affecting the results of the
differential expression analysis: we would obtain gene differences
driven by other variables rather than
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> Group</span> </mark>. Also,
<mark style="background-color: #DFF0FE"><span style="font-family: monospace"> sum</span> </mark> and
<mark style="background-color: #DFF0FE"><span style="font-family: monospace"> detected</span> </mark> are
correlated.</p>
<p>Let’s explore why these variables are correlated in the following
plots.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 1.1  Barplots/Boxplots/Scatterplots for each pair of correlated variables</span></span>
<span></span>
<span><span class="va">corr_plots</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">age</span>, <span class="va">sample_var1</span>, <span class="va">sample_var2</span>, <span class="va">sample_color</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co">## Data</span></span>
<span>    <span class="va">rse_gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/eval.html" class="external-link">eval</a></span><span class="op">(</span><span class="fu"><a href="https://rlang.r-lib.org/reference/parse_expr.html" class="external-link">parse_expr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"rse_gene"</span>, <span class="va">age</span>, <span class="st">"qc"</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">CCA</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/eval.html" class="external-link">eval</a></span><span class="op">(</span><span class="fu"><a href="https://rlang.r-lib.org/reference/parse_expr.html" class="external-link">parse_expr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"CCA_"</span>, <span class="va">age</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## Sample color by one variable</span></span>
<span>    <span class="va">colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        <span class="st">"Group"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span> <span class="op">=</span> <span class="st">"brown2"</span>, <span class="st">"Experimental"</span> <span class="op">=</span> <span class="st">"deepskyblue3"</span><span class="op">)</span>,</span>
<span>        <span class="st">"Age"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Adult"</span> <span class="op">=</span> <span class="st">"slateblue3"</span>, <span class="st">"Pup"</span> <span class="op">=</span> <span class="st">"yellow3"</span><span class="op">)</span>,</span>
<span>        <span class="st">"Sex"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"F"</span> <span class="op">=</span> <span class="st">"hotpink1"</span>, <span class="st">"M"</span> <span class="op">=</span> <span class="st">"dodgerblue"</span><span class="op">)</span>,</span>
<span>        <span class="st">"Pregnancy"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Yes"</span> <span class="op">=</span> <span class="st">"darkorchid3"</span>, <span class="st">"No"</span> <span class="op">=</span> <span class="st">"darkolivegreen4"</span><span class="op">)</span>,</span>
<span>        <span class="st">"plate"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Plate1"</span> <span class="op">=</span> <span class="st">"darkorange"</span>, <span class="st">"Plate2"</span> <span class="op">=</span> <span class="st">"lightskyblue"</span>, <span class="st">"Plate3"</span> <span class="op">=</span> <span class="st">"deeppink1"</span><span class="op">)</span>,</span>
<span>        <span class="st">"flowcell"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>            <span class="st">"HKCG7DSXX"</span> <span class="op">=</span> <span class="st">"chartreuse2"</span>, <span class="st">"HKCMHDSXX"</span> <span class="op">=</span> <span class="st">"magenta"</span>, <span class="st">"HKCNKDSXX"</span> <span class="op">=</span> <span class="st">"turquoise3"</span>,</span>
<span>            <span class="st">"HKCTMDSXX"</span> <span class="op">=</span> <span class="st">"tomato"</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span></span>
<span>    <span class="va">data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">rse_gene</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## Barplots for categorical variable vs categorical variable</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span>, <span class="va">sample_var1</span><span class="op">]</span><span class="op">)</span> <span class="op">==</span> <span class="st">"character"</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span>, <span class="va">sample_var2</span><span class="op">]</span><span class="op">)</span> <span class="op">==</span> <span class="st">"character"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="co">## y-axis label</span></span>
<span>        <span class="kw">if</span> <span class="op">(</span><span class="va">sample_var2</span> <span class="op">==</span> <span class="st">"Pregnancy"</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="va">y_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Number of samples from each "</span>, <span class="va">sample_var2</span>, <span class="st">" group"</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span>        <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>            <span class="va">y_label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"Number of samples from each "</span>, <span class="va">sample_var2</span>, sep <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span></span>
<span>        <span class="co"># Stacked barplot with counts for 2nd variable</span></span>
<span>        <span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>            x <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">sample_var1</span><span class="op">)</span>,</span>
<span>            fill <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">sample_var2</span><span class="op">)</span></span>
<span>        <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span>position <span class="op">=</span> <span class="st">"stack"</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="co">## Colors by 2nd variable</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">colors</span><span class="op">[[</span><span class="va">sample_var2</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="co">## Show sample counts on stacked bars</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_text.html" class="external-link">geom_text</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>label <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes_eval.html" class="external-link">after_stat</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                stat <span class="op">=</span> <span class="st">"count"</span>,</span>
<span>                position <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_stack.html" class="external-link">position_stack</a></span><span class="op">(</span>vjust <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span>, colour <span class="op">=</span> <span class="st">"gray20"</span>, size <span class="op">=</span> <span class="fl">3</span></span>
<span>            <span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>                subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Corr: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">signif</a></span><span class="op">(</span><span class="va">CCA</span><span class="op">[</span><span class="va">sample_var1</span>, <span class="va">sample_var2</span><span class="op">]</span>, digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                y <span class="op">=</span> <span class="va">y_label</span></span>
<span>            <span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>                axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="op">(</span><span class="fl">7</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="op">(</span><span class="fl">6</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                plot.subtitle <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">7</span>, color <span class="op">=</span> <span class="st">"gray40"</span><span class="op">)</span>,</span>
<span>                legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>,</span>
<span>                legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span>
<span>            <span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span></span>
<span>    <span class="co">## Boxplots for categorical variable vs continuous variable</span></span>
<span>    <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span>, <span class="va">sample_var1</span><span class="op">]</span><span class="op">)</span> <span class="op">==</span> <span class="st">"character"</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span>, <span class="va">sample_var2</span><span class="op">]</span><span class="op">)</span> <span class="op">==</span> <span class="st">"numeric"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span>data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>, mapping <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>            x <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">sample_var1</span><span class="op">)</span>,</span>
<span>            y <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">sample_var2</span><span class="op">)</span>,</span>
<span>            color <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">sample_var1</span><span class="op">)</span></span>
<span>        <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">0.25</span>, width <span class="op">=</span> <span class="fl">0.32</span>, color <span class="op">=</span> <span class="st">"black"</span>, outlier.color <span class="op">=</span> <span class="st">"#FFFFFFFF"</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span>width <span class="op">=</span> <span class="fl">0.15</span>, alpha <span class="op">=</span> <span class="fl">1</span>, size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">stat_smooth</a></span><span class="op">(</span>geom <span class="op">=</span> <span class="st">"line"</span>, alpha <span class="op">=</span> <span class="fl">0.6</span>, size <span class="op">=</span> <span class="fl">0.4</span>, span <span class="op">=</span> <span class="fl">0.3</span>, method <span class="op">=</span> <span class="va">lm</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>group <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, color <span class="op">=</span> <span class="st">"orangered3"</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">colors</span><span class="op">[[</span><span class="va">sample_var1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>                subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Corr: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">signif</a></span><span class="op">(</span><span class="va">CCA</span><span class="op">[</span><span class="va">sample_var1</span>, <span class="va">sample_var2</span><span class="op">]</span>, digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_"</span>, <span class="st">" "</span>, <span class="va">sample_var2</span><span class="op">)</span>,</span>
<span>                x <span class="op">=</span> <span class="va">sample_var1</span></span>
<span>            <span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>                axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="op">(</span><span class="fl">7</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="op">(</span><span class="fl">6</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                plot.subtitle <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">7</span>, color <span class="op">=</span> <span class="st">"gray40"</span><span class="op">)</span>,</span>
<span>                legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>,</span>
<span>                legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span>
<span>            <span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span></span>
<span>    <span class="co">## Scatterplots for continuous variable vs continuous variable</span></span>
<span>    <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span>, <span class="va">sample_var1</span><span class="op">]</span><span class="op">)</span> <span class="op">==</span> <span class="st">"numeric"</span> <span class="op">&amp;</span> <span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">data</span><span class="op">[</span>, <span class="va">sample_var2</span><span class="op">]</span><span class="op">)</span> <span class="op">==</span> <span class="st">"numeric"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">data</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>            x <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">sample_var1</span><span class="op">)</span>,</span>
<span>            y <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">sample_var2</span><span class="op">)</span>,</span>
<span>            color <span class="op">=</span> <span class="op">!</span><span class="op">!</span><span class="fu">rlang</span><span class="fu">::</span><span class="fu"><a href="https://rlang.r-lib.org/reference/sym.html" class="external-link">sym</a></span><span class="op">(</span><span class="va">sample_color</span><span class="op">)</span></span>
<span>        <span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_smooth.html" class="external-link">stat_smooth</a></span><span class="op">(</span>geom <span class="op">=</span> <span class="st">"line"</span>, alpha <span class="op">=</span> <span class="fl">0.4</span>, size <span class="op">=</span> <span class="fl">0.4</span>, span <span class="op">=</span> <span class="fl">0.25</span>, method <span class="op">=</span> <span class="va">lm</span>, color <span class="op">=</span> <span class="st">"orangered3"</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="co">## Color by sample_color variable</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>name <span class="op">=</span> <span class="va">sample_color</span>, values <span class="op">=</span> <span class="va">colors</span><span class="op">[[</span><span class="va">sample_color</span><span class="op">]</span><span class="op">]</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"Corr: "</span>, <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">signif</a></span><span class="op">(</span><span class="va">CCA</span><span class="op">[</span><span class="va">sample_var1</span>, <span class="va">sample_var2</span><span class="op">]</span>, digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_"</span>, <span class="st">" "</span>, <span class="va">sample_var2</span><span class="op">)</span>, x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"_"</span>, <span class="st">" "</span>, <span class="va">sample_var1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>                axis.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="op">(</span><span class="fl">7</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="op">(</span><span class="fl">6</span><span class="op">)</span><span class="op">)</span>,</span>
<span>                plot.subtitle <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">7</span>, color <span class="op">=</span> <span class="st">"gray40"</span><span class="op">)</span>,</span>
<span>                legend.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">6</span><span class="op">)</span>,</span>
<span>                legend.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">7</span><span class="op">)</span></span>
<span>            <span class="op">)</span></span>
<span>    <span class="op">}</span></span>
<span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">plot</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>As shown below, the mitochondrial rate and the fraction of reads that
mapped to genes are negatively correlated in adults but control and
experimental samples are evenly distributed.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Correlation plots for adults</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">corr_plots</span><span class="op">(</span><span class="st">"adults"</span>, <span class="st">"mitoRate"</span>, <span class="st">"totalAssignedGene"</span>, <span class="st">"Group"</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span>, <span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-24-1.png" width="700">
Although not expected, the flowcell and the plate of the adult samples
were correlated, but that is due to the fact that all samples from the
first flowcell (HKCG7DSXX) were in the 1st plate, and almost all samples
from the second flowcell were in the 3rd plate.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">corr_plots</span><span class="op">(</span><span class="st">"adults"</span>, <span class="st">"flowcell"</span>, <span class="st">"plate"</span>, <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.5</span>, <span class="fl">4.5</span>, <span class="fl">1.5</span>, <span class="fl">4.5</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-25-1.png" width="700"></p>
<p>We also appreciated that the library size and the number of detected
of genes are correlated in pups and adults. Note however that in pups
control samples have bigger numbers of expressed genes.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">corr_plots</span><span class="op">(</span><span class="st">"adults"</span>, <span class="st">"sum"</span>, <span class="st">"detected"</span>, <span class="st">"Group"</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span>, <span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-26-1.png" width="700"></p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co">## Correlation plots for pups</span></span>
<span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">corr_plots</span><span class="op">(</span><span class="st">"pups"</span>, <span class="st">"sum"</span>, <span class="st">"detected"</span>, <span class="st">"Group"</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">4</span>, <span class="fl">2</span>, <span class="fl">4</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-26-2.png" width="700"></p>
<p>Plate was also slightly correlated with the overall mapping rate, but
if we look closely, the trend is given by the plate 1 samples that have
lower rates; the rates of samples from the 2nd and 3rd plates are
similar.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p</span> <span class="op">&lt;-</span> <span class="fu">corr_plots</span><span class="op">(</span><span class="st">"pups"</span>, <span class="st">"plate"</span>, <span class="st">"overallMapRate"</span>, <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="va">p</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">5.3</span>, <span class="fl">2</span>, <span class="fl">5.3</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-27-1.png" width="700"></p>
<div class="alert alert-info">
<p>Now look at the following plots. <strong>Why is it important that
experimental and control samples are distributed throughout all the
plates and flowcells?</strong></p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">corr_plots</span><span class="op">(</span><span class="st">"adults"</span>, <span class="st">"Group"</span>, <span class="st">"plate"</span>, <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">corr_plots</span><span class="op">(</span><span class="st">"pups"</span>, <span class="st">"Group"</span>, <span class="st">"plate"</span>, <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="va">p3</span> <span class="op">&lt;-</span> <span class="fu">corr_plots</span><span class="op">(</span><span class="st">"adults"</span>, <span class="st">"Group"</span>, <span class="st">"flowcell"</span>, <span class="cn">NULL</span><span class="op">)</span></span>
<span><span class="va">p4</span> <span class="op">&lt;-</span> <span class="fu">corr_plots</span><span class="op">(</span><span class="st">"pups"</span>, <span class="st">"Group"</span>, <span class="st">"flowcell"</span>, <span class="cn">NULL</span><span class="op">)</span></span>
<span></span>
<span></span>
<span><span class="va">plots</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, <span class="va">p3</span>, <span class="va">p4</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="va">plots</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2.5</span>, <span class="fl">1</span>, <span class="fl">2.5</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-28-1.png" width="700"><strong><em>Hint</em></strong>: What would happen if all experimental
samples were in one plate or flowcell and controls in another?</p>
</div>
<p>After identifying which variables are correlated and exploring the
metrics of control and experimental samples the next step is to
determine which of these variables must be removed. How do we discern
which of the correlated variables to keep and which to drop? As
recommended in the <code>variancePartition</code> user’s guide [4],
initially we can fit a linear model for each gene taking all sample
variables and then define which ones explain a higher percentage of
variance in many genes. But first let’s review how
<code>variancePartition</code> works.</p>
</div>
<div class="section level6">
<h6 class="unnumbered" id="fit-model-and-extract-fraction-of-variance-explained">
<em>2.3.2 Fit model and extract fraction of
variance explained</em><a class="anchor" aria-label="anchor" href="#fit-model-and-extract-fraction-of-variance-explained"></a>
</h6>
<p>Briefly, <code>variancePartition</code> fits a linear model for each
gene separately and <code><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/calcVarPart-method.html" class="external-link">calcVarPart()</a></code> computes the fraction of
variance in gene expression that is explained by each variable of the
study design, plus the residual variation. The effect of each variable
is assessed while jointly accounting for all others [3].</p>
<p>Basically what it does is calculate the data variance given by each
variable and that of the total model fit, summarizing the contribution
of each variable in terms of the fraction of variation explained (FVE).
Since it calculates the fraction of total variation attributable to each
aspect of the study design, these fractions naturally sum to 1 [3].</p>
<p><code>variancePartition</code> fits two types of models:</p>
<ol style="list-style-type: decimal">
<li>
<strong>Linear mixed model (LMM)</strong> where all categorical
variables are modeled as <strong>random effects</strong> and all
continuous variables are <strong>fixed effects</strong>. The function
<code>lmer()</code> from <code>lme4</code> is used to fit this
model.</li>
</ol>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Fit LMM specifying the existence of random effects with '(1| )' </span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu">lmer</span><span class="op">(</span><span class="va">expr</span> <span class="op">~</span> <span class="va">a</span> <span class="op">+</span> <span class="va">b</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span><span class="op">|</span><span class="va">c</span><span class="op">)</span>, data<span class="op">=</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<ol start="2" style="list-style-type: decimal">
<li>Fixed effects model, which is basically the <strong>standard linear
model (LM)</strong>, where all variables are modeled as fixed effects.
The function <code><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm()</a></code> is used to fit this model.</li>
</ol>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Fit LM modeling all variables as fixed effects</span></span>
<span><span class="va">fit</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/lm.html" class="external-link">lm</a></span><span class="op">(</span><span class="va">expr</span> <span class="op">~</span> <span class="va">a</span> <span class="op">+</span> <span class="va">b</span> <span class="op">+</span> <span class="va">c</span>, data<span class="op">=</span><span class="va">data</span><span class="op">)</span></span></code></pre></div>
<p>In our case, the function will be modeled a mixed model since we have
both effects.</p>
<p class="comment">
❓<strong><em>What are random and fixed effects? How to determine if a
variable is one or the other?</em></strong> <strong>Categorical
variables</strong> are usually modeled as <strong>random
effects</strong>, i.e., traits such as the batch, sex, flowcell, plate,
individual, variables ‘randomly chosen or selected from a population’
and whose specific levels are not of particular interest, only the
grouping of the samples by those variables. These are control
variables/factors that vary randomly across individuals or groups and we
use them because we must control for these effects. Think of them as
having different effects on gene expression (the dependent variable)
depending on their values. <strong>Continuous variables</strong> must be
modeled as <strong>fixed effects</strong>; they cannot be modeled as
random effects. These correspond to variables that can be measured
somehow and whose levels are themselves of interest (the QC metrics, for
instance); these effects would be the same for all genes.
</p>
<p class="comment">
❓<strong><em>Why is this effect distinction important?</em></strong>
Because when we have clustered data, like gene expression values grouped
by sex, plate, etc. we are violating the relevant assumption of
independence, making an incorrect inference when using a <strong>general
linear model (GLM)</strong>. If we have clustered data where the
variables’ values have distinct effects on gene expression, we must work
with an extension of GLM, with the <strong>linear mixed model
(LMM)</strong> that contains a mix of both fixed and random effects [5].
</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 2. Fit model</span></span>
<span></span>
<span><span class="co">## Fit a linear mixed model (LMM) that takes continuous variables as fixed effects and categorical variables as random effects</span></span>
<span></span>
<span><span class="va">varPartAnalysis</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">age</span>, <span class="va">formula</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">RSE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/eval.html" class="external-link">eval</a></span><span class="op">(</span><span class="fu"><a href="https://rlang.r-lib.org/reference/parse_expr.html" class="external-link">parse_expr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"rse_gene"</span>, <span class="va">age</span>, <span class="st">"qc"</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## Ignore genes with variance 0</span></span>
<span>    <span class="va">genes_var_zero</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">RSE</span><span class="op">)</span><span class="op">$</span><span class="va">logcounts</span>, <span class="fl">1</span>, <span class="va">var</span><span class="op">)</span> <span class="op">==</span> <span class="fl">0</span><span class="op">)</span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">genes_var_zero</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">RSE</span> <span class="op">&lt;-</span> <span class="va">RSE</span><span class="op">[</span><span class="op">-</span><span class="va">genes_var_zero</span>, <span class="op">]</span></span>
<span>    <span class="op">}</span></span>
<span></span>
<span>    <span class="co">## Loop over each gene to fit model and extract variance explained by each variable</span></span>
<span>    <span class="va">varPart</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/fitExtractVarPartModel-method.html" class="external-link">fitExtractVarPartModel</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assays</a></span><span class="op">(</span><span class="va">RSE</span><span class="op">)</span><span class="op">$</span><span class="va">logcounts</span>, <span class="va">formula</span>, <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">RSE</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># Sort variables by median fraction of variance explained</span></span>
<span>    <span class="va">vp</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/sortCols-method.html" class="external-link">sortCols</a></span><span class="op">(</span><span class="va">varPart</span><span class="op">)</span></span>
<span>    <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="http://DiseaseNeurogenomics.github.io/variancePartition/reference/plotVarPart-method.html" class="external-link">plotVarPart</a></span><span class="op">(</span><span class="va">vp</span><span class="op">)</span></span>
<span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">p</span>, <span class="va">vp</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Violin plots</span></span>
<span></span>
<span><span class="co">#####  Model with all variables  #####</span></span>
<span></span>
<span><span class="co">## Adults</span></span>
<span><span class="co">## Define variables; random effects indicated with (1| )</span></span>
<span><span class="va">formula</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">Group</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">Pregnancy</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">plate</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">flowcell</span><span class="op">)</span> <span class="op">+</span> <span class="va">mitoRate</span> <span class="op">+</span> <span class="va">overallMapRate</span> <span class="op">+</span></span>
<span>    <span class="va">totalAssignedGene</span> <span class="op">+</span> <span class="va">rRNA_rate</span> <span class="op">+</span> <span class="va">sum</span> <span class="op">+</span> <span class="va">detected</span> <span class="op">+</span> <span class="va">ERCCsumLogErr</span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu">varPartAnalysis</span><span class="op">(</span><span class="st">"adults"</span>, <span class="va">formula</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">plot</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>    axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="op">(</span><span class="fl">7</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="op">(</span><span class="fl">7.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>As presented above, in adults we can notice that
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> totalAssignedGene</span> </mark> has a
larger mean FVE than <mark style="background-color: #FCF3CF"><span style="font-family: monospace"> mitoRate</span> </mark> so we keep the
former. Same reason to remove <mark style="background-color: #FAECF8"><span style="font-family: monospace"> plate</span> </mark> that is
correlated with <mark style="background-color: #FAECF8"><span style="font-family: monospace"> flowcell</span> </mark> and with
<mark style="background-color: #FCF3CF"><span style="font-family: monospace"> overallMapRate</span> </mark>, and to
drop <mark style="background-color: #DFF0FE"><span style="font-family: monospace"> sum</span> </mark> that goes after
<mark style="background-color: #DFF0FE"><span style="font-family: monospace"> detected</span> </mark>.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#####  Model without correlated variables  #####</span></span>
<span></span>
<span><span class="co">## Adult plots without mitoRate, plate and sum</span></span>
<span><span class="va">formula</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">Group</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">Pregnancy</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">flowcell</span><span class="op">)</span> <span class="op">+</span> <span class="va">overallMapRate</span> <span class="op">+</span> <span class="va">totalAssignedGene</span> <span class="op">+</span> <span class="va">rRNA_rate</span> <span class="op">+</span> <span class="va">detected</span> <span class="op">+</span> <span class="va">ERCCsumLogErr</span></span>
<span><span class="va">varPart</span> <span class="op">&lt;-</span> <span class="fu">varPartAnalysis</span><span class="op">(</span><span class="st">"adults"</span>, <span class="va">formula</span><span class="op">)</span></span>
<span><span class="va">varPart_data_adults</span> <span class="op">&lt;-</span> <span class="va">varPart</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="va">varPart</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">plot</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>    axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="op">(</span><span class="fl">7</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="op">(</span><span class="fl">7.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Notwithstanding, in this new reduced model Group contribution doesn’t
increment much in comparison with the complete model and also note that
the % of variance explained by the residuals, i.e., the % of gene
expression variance that the model couldn’t explain, increments in this
model compared to the previous; by removing independent variables to a
regression equation, we can explain less of the variance of the
dependent variable [5]. That’s the price to pay when dropping variables
but it is convenient in this case since we don’t have many samples for
the model to determine their real unique contributions.</p>
<p>In pups, based on the model with all variables, again
<mark style="background-color: #DFF0FE"><span style="font-family: monospace"> sum</span> </mark> must be removed, as
well as <mark style="background-color: #FCF3CF"><span style="font-family: monospace"> rRNA_rate</span> </mark> and
<mark style="background-color: #FAECF8"><span style="font-family: monospace"> plate</span> </mark> that are correlated
with <mark style="background-color: #FCF3CF"><span style="font-family: monospace"> overallMapRate</span> </mark>.</p>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#####  Model with all variables  #####</span></span>
<span></span>
<span><span class="co">## Pups</span></span>
<span><span class="va">formula</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">Group</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">Sex</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">plate</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">flowcell</span><span class="op">)</span> <span class="op">+</span> <span class="va">mitoRate</span> <span class="op">+</span> <span class="va">overallMapRate</span> <span class="op">+</span></span>
<span>    <span class="va">totalAssignedGene</span> <span class="op">+</span> <span class="va">rRNA_rate</span> <span class="op">+</span> <span class="va">sum</span> <span class="op">+</span> <span class="va">detected</span> <span class="op">+</span> <span class="va">ERCCsumLogErr</span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu">varPartAnalysis</span><span class="op">(</span><span class="st">"pups"</span>, <span class="va">formula</span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">plot</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>    axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="op">(</span><span class="fl">7</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="op">(</span><span class="fl">7.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Without correlated variables, Group’s contribution increases but so
does the residual source.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#####  Model without correlated variables  #####</span></span>
<span></span>
<span><span class="co">## Pup plots without sum, rRNA_rate and plate</span></span>
<span><span class="va">formula</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">Group</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">Sex</span><span class="op">)</span> <span class="op">+</span> <span class="op">(</span><span class="fl">1</span> <span class="op">|</span> <span class="va">flowcell</span><span class="op">)</span> <span class="op">+</span> <span class="va">mitoRate</span> <span class="op">+</span> <span class="va">overallMapRate</span> <span class="op">+</span> <span class="va">totalAssignedGene</span> <span class="op">+</span> <span class="va">detected</span> <span class="op">+</span> <span class="va">ERCCsumLogErr</span></span>
<span><span class="va">varPart</span> <span class="op">&lt;-</span> <span class="fu">varPartAnalysis</span><span class="op">(</span><span class="st">"pups"</span>, <span class="va">formula</span><span class="op">)</span></span>
<span><span class="va">varPart_data_pups</span> <span class="op">&lt;-</span> <span class="va">varPart</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">plot</span> <span class="op">&lt;-</span> <span class="va">varPart</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">plot</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>    plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span>,</span>
<span>    axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="op">(</span><span class="fl">7</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    axis.text.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="op">(</span><span class="fl">7.5</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>At this point we have extracted normalized counts and filtered genes,
we have taken out poor-quality samples and we have decided which sample
variables to retain for future analyses. Now we are ready for the
DEA!</p>
</div>
</div>
</div>
<div class="section level4">
<h4 class="unnumbered" id="differential-expression-analysis-1">
<strong>3. Differential Expression
Analysis</strong><a class="anchor" aria-label="anchor" href="#differential-expression-analysis-1"></a>
</h4>
<div class="section level5">
<h5 class="unnumbered" id="modeling">3.1 Modeling<a class="anchor" aria-label="anchor" href="#modeling"></a>
</h5>
<p>So once we have decided which variables to include in the models for
DEA, we can start using <code>limma</code> to fit a linear model for
each gene and estimate the logFC’s for the contrast of interest:
Group.</p>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## DEA for nicotine vs ctrls</span></span>
<span><span class="va">DEA_expt_vs_ctl</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">age</span>, <span class="va">formula</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">RSE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/eval.html" class="external-link">eval</a></span><span class="op">(</span><span class="fu"><a href="https://rlang.r-lib.org/reference/parse_expr.html" class="external-link">parse_expr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"rse_gene"</span>, <span class="va">age</span>, <span class="st">"qc"</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">6</span>, <span class="fl">4</span>, <span class="fl">6</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## Model matrix</span></span>
<span>    <span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="va">formula</span>, data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">RSE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="co">## Comparison of interest: Group</span></span>
<span>    <span class="va">coef</span> <span class="op">&lt;-</span> <span class="st">"GroupExperimental"</span></span>
<span></span>
<span>    <span class="co">## voom():</span></span>
<span>    <span class="co">## Transform counts to log2(CPM)</span></span>
<span>    <span class="co">## Estimate mean-variance relationship for each gene</span></span>
<span>    <span class="co">## Compute weights for limma</span></span>
<span>    <span class="va">vGene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/voom.html" class="external-link">voom</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">RSE</span><span class="op">)</span>, design <span class="op">=</span> <span class="va">model</span>, plot <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## lmFit():</span></span>
<span>    <span class="co">## Fit linear model for each gene to estimate logFCs</span></span>
<span>    <span class="va">fitGene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/lmFit.html" class="external-link">lmFit</a></span><span class="op">(</span><span class="va">vGene</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## eBayes():</span></span>
<span>    <span class="co">## Compute empirical Bayes statistics</span></span>
<span>    <span class="va">eBGene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/limma/man/ebayes.html" class="external-link">eBayes</a></span><span class="op">(</span><span class="va">fitGene</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## Plot average log expression vs log2FC</span></span>
<span>    <span class="fu">limma</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/limma/man/plotma.html" class="external-link">plotMA</a></span><span class="op">(</span><span class="va">eBGene</span>,</span>
<span>        coef <span class="op">=</span> <span class="va">coef</span>, xlab <span class="op">=</span> <span class="st">"Mean of normalized counts"</span>,</span>
<span>        ylab <span class="op">=</span> <span class="st">"log2FC"</span></span>
<span>    <span class="op">)</span></span>
<span>    <span class="co">## Plot -log(p-value) vs log2FC</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/limma/man/volcanoplot.html" class="external-link">volcanoplot</a></span><span class="op">(</span><span class="va">eBGene</span>, coef <span class="op">=</span> <span class="va">coef</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## topTable():</span></span>
<span>    <span class="co">## Rank genes by logFC for Group (nicotine vs ctrl)</span></span>
<span>    <span class="va">top_genes</span> <span class="op">&lt;-</span> <span class="fu">topTable</span><span class="op">(</span><span class="va">eBGene</span>, coef <span class="op">=</span> <span class="va">coef</span>, p.value <span class="op">=</span> <span class="fl">1</span>, number <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">RSE</span><span class="op">)</span>, sort.by <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## Histogram of adjusted p-values</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">top_genes</span><span class="op">$</span><span class="va">adj.P.Val</span>, xlab <span class="op">=</span> <span class="st">"Adjusted p-value"</span>, main <span class="op">=</span> <span class="st">""</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co"># dev.off()</span></span>
<span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">top_genes</span>, <span class="va">vGene</span>, <span class="va">eBGene</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>So for the adults we can see in the histogram that there are not DEGs
(genes with FDR&lt;0.05).</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## DEA for adults</span></span>
<span><span class="va">formula</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="va">Group</span> <span class="op">+</span> <span class="va">Pregnancy</span> <span class="op">+</span> <span class="va">flowcell</span> <span class="op">+</span> <span class="va">overallMapRate</span> <span class="op">+</span> <span class="va">totalAssignedGene</span> <span class="op">+</span> <span class="va">rRNA_rate</span> <span class="op">+</span> <span class="va">detected</span> <span class="op">+</span> <span class="va">ERCCsumLogErr</span></span>
<span><span class="va">results_adults</span> <span class="op">&lt;-</span> <span class="fu">DEA_expt_vs_ctl</span><span class="op">(</span><span class="st">"adults"</span>, <span class="va">formula</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-35-1.png" width="700"></p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_genes_adults</span> <span class="op">&lt;-</span> <span class="va">results_adults</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">vGene_adults</span> <span class="op">&lt;-</span> <span class="va">results_adults</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">eBGene_adults</span> <span class="op">&lt;-</span> <span class="va">results_adults</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Number of DEGs (FDR&lt;0.05)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">top_genes_adults</span><span class="op">$</span><span class="va">adj.P.Val</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 0</span></span></code></pre></div>
<p>But for pups there are 1667 DEGs for control vs nicotine
administration.</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## DEA for pups</span></span>
<span><span class="va">formula</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="va">Group</span> <span class="op">+</span> <span class="va">Sex</span> <span class="op">+</span> <span class="va">flowcell</span> <span class="op">+</span> <span class="va">mitoRate</span> <span class="op">+</span> <span class="va">overallMapRate</span> <span class="op">+</span> <span class="va">totalAssignedGene</span> <span class="op">+</span> <span class="va">detected</span> <span class="op">+</span> <span class="va">ERCCsumLogErr</span></span>
<span><span class="va">results_pups</span> <span class="op">&lt;-</span> <span class="fu">DEA_expt_vs_ctl</span><span class="op">(</span><span class="st">"pups"</span>, <span class="va">formula</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-36-1.png" width="700"></p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">top_genes_pups</span> <span class="op">&lt;-</span> <span class="va">results_pups</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">vGene_pups</span> <span class="op">&lt;-</span> <span class="va">results_pups</span><span class="op">[[</span><span class="fl">2</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="va">eBGene_pups</span> <span class="op">&lt;-</span> <span class="va">results_pups</span><span class="op">[[</span><span class="fl">3</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co">## Number of DEGs (FDR&lt;0.05)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">top_genes_pups</span><span class="op">$</span><span class="va">adj.P.Val</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 1667</span></span></code></pre></div>
<p>In the following plots we can easily identify which genes in pups
were DE based on their logFC’s and FDR’s.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/LieberInstitute/jaffelab" class="external-link">jaffelab</a></span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Plots for DE genes</span></span>
<span><span class="va">plots_DE</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">top_genes</span>, <span class="va">vGene</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co">## Define FDR threshold for significance</span></span>
<span>    <span class="va">FDR</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span>
<span></span>
<span>    <span class="co">## NS/Down/Upregulated genes</span></span>
<span>    <span class="va">DE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">top_genes</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="kw">if</span> <span class="op">(</span><span class="va">top_genes</span><span class="op">$</span><span class="va">adj.P.Val</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&gt;</span> <span class="va">FDR</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="va">DE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="va">DE</span>, <span class="st">"ns"</span><span class="op">)</span></span>
<span>        <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>            <span class="kw">if</span> <span class="op">(</span><span class="va">top_genes</span><span class="op">$</span><span class="va">logFC</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span> <span class="op">{</span></span>
<span>                <span class="va">DE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="va">DE</span>, <span class="st">"Up"</span><span class="op">)</span></span>
<span>            <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>                <span class="va">DE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="va">DE</span>, <span class="st">"Down"</span><span class="op">)</span></span>
<span>            <span class="op">}</span></span>
<span>        <span class="op">}</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">top_genes</span><span class="op">$</span><span class="va">DE</span> <span class="op">&lt;-</span> <span class="va">DE</span></span>
<span></span>
<span>    <span class="co">## Gene symbols for top up and down DEGs</span></span>
<span>    <span class="va">up_DEGs</span> <span class="op">&lt;-</span> <span class="va">top_genes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">top_genes</span><span class="op">$</span><span class="va">logFC</span> <span class="op">&gt;</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="va">top_genes</span><span class="op">$</span><span class="va">adj.P.Val</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span>, <span class="op">]</span></span>
<span>    <span class="va">down_DEGs</span> <span class="op">&lt;-</span> <span class="va">top_genes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">top_genes</span><span class="op">$</span><span class="va">logFC</span> <span class="op">&lt;</span> <span class="fl">0</span> <span class="op">&amp;</span> <span class="va">top_genes</span><span class="op">$</span><span class="va">adj.P.Val</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span>, <span class="op">]</span></span>
<span></span>
<span>    <span class="co">## Top most significant up and down DEGs</span></span>
<span>    <span class="va">top_up_DEGs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">up_DEGs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">up_DEGs</span><span class="op">$</span><span class="va">adj.P.Val</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="op">]</span><span class="op">)</span></span>
<span>    <span class="va">top_down_DEGs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">down_DEGs</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/order.html" class="external-link">order</a></span><span class="op">(</span><span class="va">down_DEGs</span><span class="op">$</span><span class="va">adj.P.Val</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>, <span class="op">]</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">DEG_symbol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">top_genes</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">top_genes</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">top_up_DEGs</span> <span class="op">|</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">top_genes</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">top_down_DEGs</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="va">DEG_symbol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="va">DEG_symbol</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">top_genes</span><span class="op">)</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span></span>
<span>        <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>            <span class="va">DEG_symbol</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="va">DEG_symbol</span>, <span class="cn">NA</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="va">top_genes</span><span class="op">$</span><span class="va">DEG_symbol</span> <span class="op">&lt;-</span> <span class="va">DEG_symbol</span></span>
<span></span>
<span>    <span class="co">## MA plot for DE genes</span></span>
<span>    <span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Up"</span> <span class="op">=</span> <span class="st">"firebrick2"</span>, <span class="st">"Down"</span> <span class="op">=</span> <span class="st">"steelblue2"</span>, <span class="st">"ns"</span> <span class="op">=</span> <span class="st">"grey"</span><span class="op">)</span></span>
<span>    <span class="va">sizes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Up"</span> <span class="op">=</span> <span class="fl">2</span>, <span class="st">"Down"</span> <span class="op">=</span> <span class="fl">2</span>, <span class="st">"ns"</span> <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span>    <span class="va">alphas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Up"</span> <span class="op">=</span> <span class="fl">1</span>, <span class="st">"Down"</span> <span class="op">=</span> <span class="fl">1</span>, <span class="st">"ns"</span> <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></span>
<span>    <span class="va">top_genes</span><span class="op">$</span><span class="va">mean_log_expr</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/apply.html" class="external-link">apply</a></span><span class="op">(</span><span class="va">vGene</span><span class="op">$</span><span class="va">E</span>, <span class="fl">1</span>, <span class="va">mean</span><span class="op">)</span></span>
<span>    <span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">top_genes</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>            x <span class="op">=</span> <span class="va">mean_log_expr</span>, y <span class="op">=</span> <span class="va">logFC</span>,</span>
<span>            fill <span class="op">=</span> <span class="va">DE</span>,</span>
<span>            size <span class="op">=</span> <span class="va">DE</span>,</span>
<span>            alpha <span class="op">=</span> <span class="va">DE</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span></span>
<span>            shape <span class="op">=</span> <span class="fl">21</span>,</span>
<span>            colour <span class="op">=</span> <span class="st">"black"</span></span>
<span>        <span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">cols</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_size_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">sizes</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_alpha_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">alphas</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"Mean of normalized counts"</span>, y <span class="op">=</span> <span class="st">"log2FC"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">0.2</span>, <span class="fl">2</span>, <span class="fl">0.2</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## Volcano plot for DE genes</span></span>
<span>    <span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="va">top_genes</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span></span>
<span>            x <span class="op">=</span> <span class="va">logFC</span>, y <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">adj.P.Val</span><span class="op">)</span>,</span>
<span>            fill <span class="op">=</span> <span class="va">DE</span>,</span>
<span>            size <span class="op">=</span> <span class="va">DE</span>,</span>
<span>            alpha <span class="op">=</span> <span class="va">DE</span>,</span>
<span>            label <span class="op">=</span> <span class="va">DEG_symbol</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">21</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_abline.html" class="external-link">geom_hline</a></span><span class="op">(</span></span>
<span>            yintercept <span class="op">=</span> <span class="op">-</span><span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log10</a></span><span class="op">(</span><span class="va">FDR</span><span class="op">)</span>,</span>
<span>            linetype <span class="op">=</span> <span class="st">"dashed"</span></span>
<span>        <span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/pkg/ggrepel/man/geom_text_repel.html" class="external-link">geom_label_repel</a></span><span class="op">(</span></span>
<span>            fill <span class="op">=</span> <span class="st">"white"</span>, size <span class="op">=</span> <span class="fl">2</span>, max.overlaps <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>            box.padding <span class="op">=</span> <span class="fl">0.2</span>,</span>
<span>            show.legend <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>        <span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span>y <span class="op">=</span> <span class="st">"-log10(FDR)"</span>, x <span class="op">=</span> <span class="st">"log2FC"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">cols</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_size_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">sizes</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_alpha_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">alphas</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">0.2</span>, <span class="fl">2</span>, <span class="fl">0.2</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span>, ncol <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb48"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plots_DE</span><span class="op">(</span><span class="va">top_genes_adults</span>, <span class="va">vGene_adults</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-38-1.png" width="700"></p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">plots_DE</span><span class="op">(</span><span class="va">top_genes_pups</span>, <span class="va">vGene_pups</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-38-2.png" width="700"></p>
<p>We can also plot the expression counts (without batch effects) of
particular DEGs in control and experimental samples, such as the genes
labeled in the above volcano plot. We can see that for the gene
ENSMUSG00000026812.16 which was downregulated, its expression levels go
down in experimental samples compared to controls, and the opposite for
ENSMUSG00000033713.12 that was upregulated.</p>
<div class="sourceCode" id="cb50"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Boxplot of a single gene</span></span>
<span><span class="va">DE_one_boxplot</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">age</span>, <span class="va">DEgene</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">RSE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/eval.html" class="external-link">eval</a></span><span class="op">(</span><span class="fu"><a href="https://rlang.r-lib.org/reference/parse_expr.html" class="external-link">parse_expr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"rse_gene"</span>, <span class="va">age</span>, <span class="st">"qc"</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">top_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/eval.html" class="external-link">eval</a></span><span class="op">(</span><span class="fu"><a href="https://rlang.r-lib.org/reference/parse_expr.html" class="external-link">parse_expr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"top_genes"</span>, <span class="va">age</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">vGene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/eval.html" class="external-link">eval</a></span><span class="op">(</span><span class="fu"><a href="https://rlang.r-lib.org/reference/parse_expr.html" class="external-link">parse_expr</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"vGene"</span>, <span class="va">age</span>, sep <span class="op">=</span> <span class="st">"_"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="kw">if</span> <span class="op">(</span><span class="va">age</span> <span class="op">==</span> <span class="st">"pups"</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="va">formula</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="va">Group</span> <span class="op">+</span> <span class="va">Sex</span> <span class="op">+</span> <span class="va">flowcell</span> <span class="op">+</span> <span class="va">mitoRate</span> <span class="op">+</span> <span class="va">overallMapRate</span> <span class="op">+</span> <span class="va">totalAssignedGene</span> <span class="op">+</span> <span class="va">detected</span> <span class="op">+</span> <span class="va">ERCCsumLogErr</span></span>
<span>    <span class="op">}</span> <span class="kw">else</span> <span class="op">{</span></span>
<span>        <span class="va">formula</span> <span class="op">&lt;-</span> <span class="op">~</span> <span class="va">Group</span> <span class="op">+</span> <span class="va">Pregnancy</span> <span class="op">+</span> <span class="va">flowcell</span> <span class="op">+</span> <span class="va">overallMapRate</span> <span class="op">+</span> <span class="va">totalAssignedGene</span> <span class="op">+</span> <span class="va">rRNA_rate</span> <span class="op">+</span> <span class="va">detected</span> <span class="op">+</span> <span class="va">ERCCsumLogErr</span></span>
<span>    <span class="op">}</span></span>
<span></span>
<span>    <span class="co">## q-value for the gene</span></span>
<span>    <span class="va">q_value</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">signif</a></span><span class="op">(</span><span class="va">top_genes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">top_genes</span><span class="op">)</span> <span class="op">==</span> <span class="va">DEgene</span><span class="op">)</span>, <span class="st">"adj.P.Val"</span><span class="op">]</span>, digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span>    <span class="co">## logFC for the gene</span></span>
<span>    <span class="va">logFC</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">signif</a></span><span class="op">(</span><span class="va">top_genes</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/which.html" class="external-link">which</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">top_genes</span><span class="op">)</span> <span class="op">==</span> <span class="va">DEgene</span><span class="op">)</span>, <span class="st">"logFC"</span><span class="op">]</span>, digits <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## Regress out residuals to remove batch effects</span></span>
<span>    <span class="va">model</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/model.matrix.html" class="external-link">model.matrix</a></span><span class="op">(</span><span class="va">formula</span>, data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">RSE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="va">vGene</span><span class="op">$</span><span class="va">E</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/jaffelab/man/cleaningY.html" class="external-link">cleaningY</a></span><span class="op">(</span><span class="va">vGene</span><span class="op">$</span><span class="va">E</span>, <span class="va">model</span>, P <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span>
<span>    <span class="co">## Samples as rows and genes as columns</span></span>
<span>    <span class="va">lognorm_DE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">vGene</span><span class="op">$</span><span class="va">E</span><span class="op">)</span></span>
<span>    <span class="co">## Add samples' Group information</span></span>
<span>    <span class="va">lognorm_DE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="va">lognorm_DE</span>, <span class="st">"Group"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">RSE</span><span class="op">)</span><span class="op">$</span><span class="va">Group</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">p</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span></span>
<span>        data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">lognorm_DE</span><span class="op">)</span>,</span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">Group</span>, y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/eval.html" class="external-link">eval</a></span><span class="op">(</span><span class="fu"><a href="https://rlang.r-lib.org/reference/parse_expr.html" class="external-link">parse_expr</a></span><span class="op">(</span><span class="va">DEgene</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="co">## Hide outliers</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_boxplot.html" class="external-link">geom_boxplot</a></span><span class="op">(</span>outlier.color <span class="op">=</span> <span class="st">"#FFFFFFFF"</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="co">## Samples colored by Group + noise</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_jitter.html" class="external-link">geom_jitter</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>colour <span class="op">=</span> <span class="va">Group</span><span class="op">)</span>,</span>
<span>            shape <span class="op">=</span> <span class="fl">16</span>,</span>
<span>            position <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_jitter.html" class="external-link">position_jitter</a></span><span class="op">(</span><span class="fl">0.2</span><span class="op">)</span>, size <span class="op">=</span> <span class="fl">2</span></span>
<span>        <span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_classic</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Control"</span> <span class="op">=</span> <span class="st">"brown2"</span>, <span class="st">"Experimental"</span> <span class="op">=</span> <span class="st">"deepskyblue3"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>            x <span class="op">=</span> <span class="st">"Group"</span>, y <span class="op">=</span> <span class="st">"norm counts - covariates"</span>,</span>
<span>            title <span class="op">=</span> <span class="va">DEgene</span>,</span>
<span>            subtitle <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">" FDR:"</span>, <span class="va">q_value</span>, <span class="st">"\n"</span>, <span class="st">"logFC:"</span>, <span class="va">logFC</span><span class="op">)</span></span>
<span>        <span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span></span>
<span>            plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.5</span>, <span class="fl">2.5</span>, <span class="fl">1.5</span>, <span class="fl">2.5</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span>, legend.position <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>            plot.title <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>hjust <span class="op">=</span> <span class="fl">0.5</span>, size <span class="op">=</span> <span class="fl">10</span>, face <span class="op">=</span> <span class="st">"bold"</span><span class="op">)</span>,</span>
<span>            plot.subtitle <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">9</span><span class="op">)</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">p</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Boxplots for DEGs in pups</span></span>
<span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu">DE_one_boxplot</span><span class="op">(</span><span class="st">"pups"</span>, <span class="st">"ENSMUSG00000026812.16"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.3</span>, <span class="fl">1.4</span>, <span class="fl">1.3</span>, <span class="fl">1.4</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu">DE_one_boxplot</span><span class="op">(</span><span class="st">"pups"</span>, <span class="st">"ENSMUSG00000033713.12"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.3</span>, <span class="fl">1.4</span>, <span class="fl">1.3</span>, <span class="fl">1.4</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://wilkelab.org/cowplot/reference/plot_grid.html" class="external-link">plot_grid</a></span><span class="op">(</span><span class="va">p1</span>, <span class="va">p2</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-40-1.png" width="700"></p>
</div>
<div class="section level5">
<h5 class="unnumbered" id="comparisons">3.2 Comparisons<a class="anchor" aria-label="anchor" href="#comparisons"></a>
</h5>
<p>Now that we have computed statistics for DE we can compare the
t-stats of the genes in pups and adults to examine the signals for DE in
both ages.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 1. t-stats plots</span></span>
<span></span>
<span><span class="co">## Function to add DE info of genes in both groups</span></span>
<span><span class="va">add_DE_info</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">top_genes1</span>, <span class="va">top_genes2</span>, <span class="va">name_1</span>, <span class="va">name_2</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">DE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/vector.html" class="external-link">vector</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">top_genes1</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span> <span class="op">{</span></span>
<span>        <span class="co">## DE genes in both groups</span></span>
<span>        <span class="kw">if</span> <span class="op">(</span><span class="va">top_genes1</span><span class="op">$</span><span class="va">adj.P.Val</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;&amp;</span> <span class="va">top_genes2</span><span class="op">$</span><span class="va">adj.P.Val</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="va">DE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="va">DE</span>, <span class="st">"sig Both"</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>        <span class="co">## DE genes in only one of the groups</span></span>
<span>        <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">top_genes1</span><span class="op">$</span><span class="va">adj.P.Val</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;&amp;</span> <span class="op">!</span><span class="va">top_genes2</span><span class="op">$</span><span class="va">adj.P.Val</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="va">DE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="va">DE</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"sig"</span>, <span class="va">name_1</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="va">top_genes2</span><span class="op">$</span><span class="va">adj.P.Val</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.05</span> <span class="op">&amp;&amp;</span> <span class="op">!</span><span class="va">top_genes1</span><span class="op">$</span><span class="va">adj.P.Val</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">{</span></span>
<span>            <span class="va">DE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="va">DE</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"sig"</span>, <span class="va">name_2</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>        <span class="co">## ns genes in both groups</span></span>
<span>        <span class="kw">else</span> <span class="op">{</span></span>
<span>            <span class="va">DE</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/append.html" class="external-link">append</a></span><span class="op">(</span><span class="va">DE</span>, <span class="st">"n.s."</span><span class="op">)</span></span>
<span>        <span class="op">}</span></span>
<span>    <span class="op">}</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">DE</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span></span>
<span><span class="co">## Compare t-stats of genes from different groups of samples</span></span>
<span><span class="va">t_stat_plot</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">top_genes1</span>, <span class="va">top_genes2</span>, <span class="va">name_1</span>, <span class="va">name_2</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="co">## Correlation coeff</span></span>
<span>    <span class="va">rho</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cor</a></span><span class="op">(</span><span class="va">top_genes1</span><span class="op">$</span><span class="va">t</span>, <span class="va">top_genes2</span><span class="op">$</span><span class="va">t</span>, method <span class="op">=</span> <span class="st">"spearman"</span><span class="op">)</span></span>
<span>    <span class="va">rho_anno</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"rho = "</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/format.html" class="external-link">format</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="va">rho</span>, <span class="fl">2</span><span class="op">)</span>, nsmall <span class="op">=</span> <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span>    <span class="co">## Merge data</span></span>
<span>    <span class="va">t_stats</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>t1 <span class="op">=</span> <span class="va">top_genes1</span><span class="op">$</span><span class="va">t</span>, t2 <span class="op">=</span> <span class="va">top_genes2</span><span class="op">$</span><span class="va">t</span><span class="op">)</span></span>
<span>    <span class="co">## Add DE info for both groups</span></span>
<span>    <span class="va">t_stats</span><span class="op">$</span><span class="va">DEG</span> <span class="op">&lt;-</span> <span class="fu">add_DE_info</span><span class="op">(</span><span class="va">top_genes1</span>, <span class="va">top_genes2</span>, <span class="va">name_1</span>, <span class="va">name_2</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">cols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"red"</span>, <span class="st">"#ffad73"</span>, <span class="st">"#26b3ff"</span>, <span class="st">"dark grey"</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cols</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sig Both"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"sig "</span>, <span class="va">name_1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"sig "</span>, <span class="va">name_2</span><span class="op">)</span>, <span class="st">"n.s."</span><span class="op">)</span></span>
<span>    <span class="va">alphas</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">1</span>, <span class="fl">0.5</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">alphas</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"sig Both"</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"sig "</span>, <span class="va">name_1</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"sig "</span>, <span class="va">name_2</span><span class="op">)</span>, <span class="st">"n.s."</span><span class="op">)</span></span>
<span></span>
<span>    <span class="va">plot</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">t_stats</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">t1</span>, y <span class="op">=</span> <span class="va">t2</span>, color <span class="op">=</span> <span class="va">DEG</span>, alpha <span class="op">=</span> <span class="va">DEG</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/labs.html" class="external-link">labs</a></span><span class="op">(</span></span>
<span>            x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"t-stats"</span>, <span class="va">name_1</span><span class="op">)</span>,</span>
<span>            y <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/paste.html" class="external-link">paste</a></span><span class="op">(</span><span class="st">"t-stats"</span>, <span class="va">name_2</span><span class="op">)</span>,</span>
<span>            title <span class="op">=</span> <span class="st">"Pups vs Adults"</span>,</span>
<span>            subtitle <span class="op">=</span> <span class="va">rho_anno</span>,</span>
<span>            parse <span class="op">=</span> <span class="cn">T</span></span>
<span>        <span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html" class="external-link">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">cols</span><span class="op">)</span> <span class="op">+</span></span>
<span>        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_alpha_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">alphas</span><span class="op">)</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">plot</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="co">## Compare t-stats of genes in Pups vs Adults</span></span>
<span></span>
<span><span class="fu">t_stat_plot</span><span class="op">(</span><span class="va">top_genes_pups</span>, <span class="va">top_genes_adults</span>, <span class="st">"Nicotine pups"</span>, <span class="st">"Nicotine adults"</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>plot.margin <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/grid/unit.html" class="external-link">unit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1.5</span>, <span class="fl">3.5</span>, <span class="fl">1.5</span>, <span class="fl">3.5</span><span class="op">)</span>, <span class="st">"cm"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="SPEAQeasyWorkshop2023_files/figure-html/unnamed-chunk-41-1.png" width="700">
As shown above, there’s no correlation between the t-stats of the genes
in pups and adults, which could be interpreted as the genes having
different signals for differential expression in both ages.</p>
</div>
</div>
<div class="section level4">
<h4 class="unnumbered" id="deg-visualization">
<strong>4. DEG visualization</strong><a class="anchor" aria-label="anchor" href="#deg-visualization"></a>
</h4>
<p>ToDO Renee</p>
</div>
<div class="section level4">
<h4 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h4>
<ol style="list-style-type: decimal">
<li><p>Evans, C., Hardin, J., &amp; Stoebel, D. M. (2018). Selecting
between-sample RNA-Seq normalization methods from the perspective of
their assumptions. <em>Briefings in bioinformatics</em>, 19(5),
776-792.</p></li>
<li><p>Bedre, R. (2023). Gene expression units explained: RPM, RPKM,
FPKM, TPM, <em>DESeq</em>, TMM, SCnorm, GeTMM, and ComBat-Seq. Web site:
<a href="https://www.reneshbedre.com/blog/expression_units.html" class="external-link uri">https://www.reneshbedre.com/blog/expression_units.html</a></p></li>
<li><p>Hoffman, G. E., &amp; Schadt, E. E. (2016). variancePartition:
interpreting drivers of variation in complex gene expression studies.
BMC bioinformatics, 17(1), 1-13.</p></li>
<li><p>Hoffman, G. (2022). variancePartition: Quantifying and
interpreting drivers of variation in multilevel gene expression
experiments.</p></li>
<li><p>van den Berg, S. M. (2022). Analysing data using linear models.
Web site: <a href="https://bookdown.org/pingapang9/linear_models_bookdown/" class="external-link uri">https://bookdown.org/pingapang9/linear_models_bookdown/</a></p></li>
</ol>
</div>
<div class="section level4">
<h4 class="unnumbered" id="additional-resources">Additional resources<a class="anchor" aria-label="anchor" href="#additional-resources"></a>
</h4>
<p>Check the material for the course of <a href="https://lcolladotor.github.io/cshl_rstats_genome_scale_2023" class="external-link">Statistical
Analysis of Genome Scale Data at CSHL 2023</a> for more exercises and
explanations of the analyses for differential expression.</p>
</div>
</div>
</div>
<div class="section level2">
<h2 id="reproducibility">Reproducibility<a class="anchor" aria-label="anchor" href="#reproducibility"></a>
</h2>
<p>Code for creating the vignette</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## Create the vignette</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/rstudio/rmarkdown" class="external-link">"rmarkdown"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="fu"><a href="https://pkgs.rstudio.com/rmarkdown/reference/render.html" class="external-link">render</a></span><span class="op">(</span><span class="st">"SPEAQeasyWorkshop2023.Rmd"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co">## Extract the R code</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://yihui.org/knitr/" class="external-link">"knitr"</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/knit.html" class="external-link">knit</a></span><span class="op">(</span><span class="st">"SPEAQeasyWorkshop2023.Rmd"</span>, tangle <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Date the vignette was generated.</p>
<pre><code><span><span class="co">#&gt; [1] "2023-07-24 18:02:15 UTC"</span></span></code></pre>
<p>Wallclock time spent generating the vignette.</p>
<pre><code><span><span class="co">#&gt; Time difference of 1.354 mins</span></span></code></pre>
<p><code>R</code> session information.</p>
<pre><code><span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">setting </span> <span style="color: #555555; font-style: italic;">value</span></span></span>
<span><span class="co">#&gt;  version  R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt;  os       Ubuntu 22.04.2 LTS</span></span>
<span><span class="co">#&gt;  system   x86_64, linux-gnu</span></span>
<span><span class="co">#&gt;  ui       X11</span></span>
<span><span class="co">#&gt;  language en</span></span>
<span><span class="co">#&gt;  collate  en_US.UTF-8</span></span>
<span><span class="co">#&gt;  ctype    en_US.UTF-8</span></span>
<span><span class="co">#&gt;  tz       UTC</span></span>
<span><span class="co">#&gt;  date     2023-07-24</span></span>
<span><span class="co">#&gt;  pandoc   3.1.1 @ /usr/local/bin/ (via rmarkdown)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;  <span style="color: #555555; font-style: italic;">package               </span> <span style="color: #555555; font-style: italic;">*</span> <span style="color: #555555; font-style: italic;">version   </span> <span style="color: #555555; font-style: italic;">date (UTC)</span> <span style="color: #555555; font-style: italic;">lib</span> <span style="color: #555555; font-style: italic;">source</span></span></span>
<span><span class="co">#&gt;  abind                    1.4-5      <span style="color: #555555;">2016-07-21</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  AnnotationDbi            1.63.2     <span style="color: #555555;">2023-07-02</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  AnnotationHub          * 3.9.1      <span style="color: #555555;">2023-05-01</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  aod                      1.3.2      <span style="color: #555555;">2022-04-02</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  backports                1.4.1      <span style="color: #555555;">2021-12-13</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  base64enc                0.1-3      <span style="color: #555555;">2015-07-28</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  beachmat                 2.17.13    <span style="color: #555555;">2023-07-09</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  beeswarm                 0.4.0      <span style="color: #555555;">2021-06-01</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  bibtex                   0.5.1      <span style="color: #555555;">2023-01-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  Biobase                * 2.61.0     <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  BiocFileCache          * 2.9.1      <span style="color: #555555;">2023-07-12</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  BiocGenerics           * 0.47.0     <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  BiocManager              1.30.21.1  <span style="color: #555555;">2023-07-18</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  BiocNeighbors            1.19.0     <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  BiocParallel           * 1.35.3     <span style="color: #555555;">2023-07-07</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  BiocSingular             1.17.1     <span style="color: #555555;">2023-07-09</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  BiocStyle              * 2.29.1     <span style="color: #555555;">2023-07-19</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  BiocVersion              3.18.0     <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  Biostrings               2.69.2     <span style="color: #555555;">2023-07-02</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  bit                      4.0.5      <span style="color: #555555;">2022-11-15</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  bit64                    4.0.5      <span style="color: #555555;">2020-08-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  bitops                   1.0-7      <span style="color: #555555;">2021-04-24</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  blob                     1.2.4      <span style="color: #555555;">2023-03-17</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  bookdown                 0.34       <span style="color: #555555;">2023-05-09</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  boot                     1.3-28.1   <span style="color: #555555;">2022-11-22</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  broom                    1.0.5      <span style="color: #555555;">2023-06-09</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  bslib                    0.5.0      <span style="color: #555555;">2023-06-09</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  cachem                   1.0.8      <span style="color: #555555;">2023-05-01</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  car                      3.1-2      <span style="color: #555555;">2023-03-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  carData                  3.0-5      <span style="color: #555555;">2022-01-06</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  caTools                  1.18.2     <span style="color: #555555;">2021-03-28</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  checkmate                2.2.0      <span style="color: #555555;">2023-04-27</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  cli                      3.6.1      <span style="color: #555555;">2023-03-23</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  cluster                  2.1.4      <span style="color: #555555;">2022-08-22</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  codetools                0.2-19     <span style="color: #555555;">2023-02-01</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  colorspace               2.1-0      <span style="color: #555555;">2023-01-23</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  corpcor                  1.6.10     <span style="color: #555555;">2021-09-16</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  cowplot                * 1.1.1      <span style="color: #555555;">2020-12-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  crayon                   1.5.2      <span style="color: #555555;">2022-09-29</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  curl                     5.0.1      <span style="color: #555555;">2023-06-07</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  data.table               1.14.8     <span style="color: #555555;">2023-02-17</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  DBI                      1.1.3      <span style="color: #555555;">2022-06-18</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  dbplyr                 * 2.3.3      <span style="color: #555555;">2023-07-07</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  DelayedArray             0.27.9     <span style="color: #555555;">2023-07-04</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  DelayedMatrixStats       1.23.0     <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  desc                     1.4.2      <span style="color: #555555;">2022-09-08</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  digest                   0.6.33     <span style="color: #555555;">2023-07-07</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  dplyr                    1.1.2      <span style="color: #555555;">2023-04-20</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  ellipsis                 0.3.2      <span style="color: #555555;">2021-04-29</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  EnvStats                 2.8.0      <span style="color: #555555;">2023-07-08</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  evaluate                 0.21       <span style="color: #555555;">2023-05-05</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  ExperimentHub          * 2.9.1      <span style="color: #555555;">2023-07-12</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  fANCOVA                  0.6-1      <span style="color: #555555;">2020-11-13</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  fansi                    1.0.4      <span style="color: #555555;">2023-01-22</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  farver                   2.1.1      <span style="color: #555555;">2022-07-06</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  fastmap                  1.1.1      <span style="color: #555555;">2023-02-24</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  filelock                 1.0.2      <span style="color: #555555;">2018-10-05</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  foreign                  0.8-84     <span style="color: #555555;">2022-12-06</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  Formula                  1.2-5      <span style="color: #555555;">2023-02-24</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  fs                       1.6.3      <span style="color: #555555;">2023-07-20</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  gargle                   1.5.2      <span style="color: #555555;">2023-07-20</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  generics                 0.1.3      <span style="color: #555555;">2022-07-05</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  GenomeInfoDb           * 1.37.2     <span style="color: #555555;">2023-06-21</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  GenomeInfoDbData         1.2.10     <span style="color: #555555;">2023-07-10</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  GenomicRanges          * 1.53.1     <span style="color: #555555;">2023-05-04</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  ggbeeswarm               0.7.2      <span style="color: #555555;">2023-04-29</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  ggplot2                * 3.4.2      <span style="color: #555555;">2023-04-03</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  ggpubr                 * 0.6.0      <span style="color: #555555;">2023-02-10</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  ggrepel                * 0.9.3      <span style="color: #555555;">2023-02-03</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  ggsignif                 0.6.4      <span style="color: #555555;">2022-10-13</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  glue                     1.6.2      <span style="color: #555555;">2022-02-24</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  googledrive              2.1.1      <span style="color: #555555;">2023-06-11</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  gplots                   3.1.3      <span style="color: #555555;">2022-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  gridExtra                2.3        <span style="color: #555555;">2017-09-09</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  gtable                   0.3.3      <span style="color: #555555;">2023-03-21</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  gtools                   3.9.4      <span style="color: #555555;">2022-11-27</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  highr                    0.10       <span style="color: #555555;">2022-12-22</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  Hmisc                  * 5.1-0      <span style="color: #555555;">2023-05-08</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  htmlTable                2.4.1      <span style="color: #555555;">2022-07-07</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  htmltools                0.5.5      <span style="color: #555555;">2023-03-23</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  htmlwidgets              1.6.2      <span style="color: #555555;">2023-03-17</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  httpuv                   1.6.11     <span style="color: #555555;">2023-05-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  httr                     1.4.6      <span style="color: #555555;">2023-05-08</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  interactiveDisplayBase   1.39.0     <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  IRanges                * 2.35.2     <span style="color: #555555;">2023-06-22</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  irlba                    2.3.5.1    <span style="color: #555555;">2022-10-03</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  iterators                1.0.14     <span style="color: #555555;">2022-02-05</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  jaffelab               * 0.99.32    <span style="color: #555555;">2023-07-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Github (LieberInstitute/jaffelab@21e6574)</span></span></span>
<span><span class="co">#&gt;  jquerylib                0.1.4      <span style="color: #555555;">2021-04-26</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  jsonlite                 1.8.7      <span style="color: #555555;">2023-06-29</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  KEGGREST                 1.41.0     <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  KernSmooth               2.23-22    <span style="color: #555555;">2023-07-10</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  knitr                    1.43       <span style="color: #555555;">2023-05-25</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  labeling                 0.4.2      <span style="color: #555555;">2020-10-20</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  later                    1.3.1      <span style="color: #555555;">2023-05-02</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  lattice                  0.21-8     <span style="color: #555555;">2023-04-05</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  lifecycle                1.0.3      <span style="color: #555555;">2022-10-07</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  limma                  * 3.57.6     <span style="color: #555555;">2023-06-19</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  lme4                     1.1-34     <span style="color: #555555;">2023-07-04</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  lmerTest                 3.1-3      <span style="color: #555555;">2020-10-23</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  lubridate                1.9.2      <span style="color: #555555;">2023-02-10</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  magrittr                 2.0.3      <span style="color: #555555;">2022-03-30</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  MASS                     7.3-60     <span style="color: #555555;">2023-05-04</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  Matrix                   1.6-0      <span style="color: #555555;">2023-07-08</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  MatrixGenerics         * 1.13.0     <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  matrixStats            * 1.0.0      <span style="color: #555555;">2023-06-02</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  memoise                  2.0.1      <span style="color: #555555;">2021-11-26</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  mgcv                     1.9-0      <span style="color: #555555;">2023-07-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  mime                     0.12       <span style="color: #555555;">2021-09-28</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  minqa                    1.2.5      <span style="color: #555555;">2022-10-19</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  munsell                  0.5.0      <span style="color: #555555;">2018-06-12</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  mvtnorm                  1.2-2      <span style="color: #555555;">2023-06-08</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  nlme                     3.1-162    <span style="color: #555555;">2023-01-31</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  nloptr                   2.0.3      <span style="color: #555555;">2022-05-26</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  nnet                     7.3-19     <span style="color: #555555;">2023-05-03</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  numDeriv                 <span style="color: #BB00BB; font-weight: bold;">2016.8-1.1</span> <span style="color: #555555;">2019-06-06</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  pbkrtest                 0.5.2      <span style="color: #555555;">2023-01-19</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  pheatmap               * 1.0.12     <span style="color: #555555;">2019-01-04</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  pillar                   1.9.0      <span style="color: #555555;">2023-03-22</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  pkgconfig                2.0.3      <span style="color: #555555;">2019-09-22</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  pkgdown                  <span style="color: #BB00BB; font-weight: bold;">2.0.7.9000</span> <span style="color: #555555;">2023-07-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Github (r-lib/pkgdown@c920680)</span></span></span>
<span><span class="co">#&gt;  plyr                     1.8.8      <span style="color: #555555;">2022-11-11</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  png                      0.1-8      <span style="color: #555555;">2022-11-29</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  promises                 1.2.0.1    <span style="color: #555555;">2021-02-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  purrr                    1.0.1      <span style="color: #555555;">2023-01-10</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  R6                       2.5.1      <span style="color: #555555;">2021-08-19</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  rafalib                * 1.0.0      <span style="color: #555555;">2015-08-09</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  ragg                     1.2.5      <span style="color: #555555;">2023-01-12</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  rappdirs                 0.3.3      <span style="color: #555555;">2021-01-31</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  rbibutils                2.2.13     <span style="color: #555555;">2023-01-13</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  RColorBrewer             1.1-3      <span style="color: #555555;">2022-04-03</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  Rcpp                     1.0.11     <span style="color: #555555;">2023-07-06</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  RCurl                    1.98-1.12  <span style="color: #555555;">2023-03-27</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  Rdpack                   2.4        <span style="color: #555555;">2022-07-20</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  RefManageR             * 1.4.0      <span style="color: #555555;">2022-09-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  remaCor                  0.0.16     <span style="color: #555555;">2023-06-21</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  reshape2                 1.4.4      <span style="color: #555555;">2020-04-09</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  RhpcBLASctl              0.23-42    <span style="color: #555555;">2023-02-11</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  rlang                  * 1.1.1      <span style="color: #555555;">2023-04-28</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  rmarkdown                2.23       <span style="color: #555555;">2023-07-01</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  rpart                    4.1.19     <span style="color: #555555;">2022-10-21</span> <span style="color: #555555;">[3]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  rprojroot                2.0.3      <span style="color: #555555;">2022-04-02</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  RSQLite                  2.3.1      <span style="color: #555555;">2023-04-03</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  rstatix                  0.7.2      <span style="color: #555555;">2023-02-01</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  rstudioapi               0.15.0     <span style="color: #555555;">2023-07-07</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  rsvd                     1.0.5      <span style="color: #555555;">2021-04-16</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  S4Arrays                 1.1.4      <span style="color: #555555;">2023-05-14</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  S4Vectors              * 0.39.1     <span style="color: #555555;">2023-05-03</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  sass                     0.4.7      <span style="color: #555555;">2023-07-15</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  ScaledMatrix             1.9.1      <span style="color: #555555;">2023-05-03</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  scales                   1.2.1      <span style="color: #555555;">2022-08-20</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  scater                 * 1.29.0     <span style="color: #555555;">2023-05-19</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  scuttle                * 1.11.0     <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  segmented                1.6-4      <span style="color: #555555;">2023-04-13</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  sessioninfo            * 1.2.2      <span style="color: #555555;">2021-12-06</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  shiny                    1.7.4.1    <span style="color: #555555;">2023-07-06</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  SingleCellExperiment   * 1.23.0     <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  smokingMouse           * 0.99.9     <span style="color: #555555;">2023-07-24</span> <span style="color: #555555;">[1]</span> <span style="color: #BB00BB; font-weight: bold;">Github (LieberInstitute/smokingMouse@8ae6660)</span></span></span>
<span><span class="co">#&gt;  SparseArray              1.1.10     <span style="color: #555555;">2023-06-11</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  sparseMatrixStats        1.13.0     <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  stringi                  1.7.12     <span style="color: #555555;">2023-01-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  stringr                * 1.5.0      <span style="color: #555555;">2022-12-02</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  SummarizedExperiment   * 1.31.1     <span style="color: #555555;">2023-05-01</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  systemfonts              1.0.4      <span style="color: #555555;">2022-02-11</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  textshaping              0.3.6      <span style="color: #555555;">2021-10-13</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  tibble                   3.2.1      <span style="color: #555555;">2023-03-20</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  tidyr                    1.3.0      <span style="color: #555555;">2023-01-24</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  tidyselect               1.2.0      <span style="color: #555555;">2022-10-10</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  timechange               0.2.0      <span style="color: #555555;">2023-01-11</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  utf8                     1.2.3      <span style="color: #555555;">2023-01-31</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  variancePartition      * 1.31.11    <span style="color: #555555;">2023-06-30</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  vctrs                    0.6.3      <span style="color: #555555;">2023-06-14</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  vipor                    0.4.5      <span style="color: #555555;">2017-03-22</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  viridis                  0.6.4      <span style="color: #555555;">2023-07-22</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  viridisLite              0.4.2      <span style="color: #555555;">2023-05-02</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  withr                    2.5.0      <span style="color: #555555;">2022-03-03</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  xfun                     0.39       <span style="color: #555555;">2023-04-20</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  xml2                     1.3.5      <span style="color: #555555;">2023-07-06</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  xtable                   1.8-4      <span style="color: #555555;">2019-04-21</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  XVector                  0.41.1     <span style="color: #555555;">2023-05-03</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt;  yaml                     2.3.7      <span style="color: #555555;">2023-01-23</span> <span style="color: #555555;">[2]</span> <span style="color: #555555;">CRAN (R 4.3.1)</span></span></span>
<span><span class="co">#&gt;  zlibbioc                 1.47.0     <span style="color: #555555;">2023-04-25</span> <span style="color: #555555;">[1]</span> <span style="color: #555555;">Bioconductor</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [1] /__w/_temp/Library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [2] /usr/local/lib/R/site-library</span></span></span>
<span><span class="co">#&gt; <span style="color: #555555;"> [3] /usr/local/lib/R/library</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB; font-weight: bold;">──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></span></code></pre>
</div>
<div class="section level2">
<h2 id="bibliography">Bibliography<a class="anchor" aria-label="anchor" href="#bibliography"></a>
</h2>
<p>This vignette was generated using <em><a href="https://bioconductor.org/packages/3.18/BiocStyle" class="external-link">BiocStyle</a></em>
(Oleś, 2023), <em><a href="https://CRAN.R-project.org/package=knitr" class="external-link">knitr</a></em> (Xie,
2023) and <em><a href="https://CRAN.R-project.org/package=rmarkdown" class="external-link">rmarkdown</a></em>
(Allaire, Xie, Dervieux, McPherson, Luraschi, Ushey, Atkins, Wickham,
Cheng, Chang, and Iannone, 2023) running behind the scenes.</p>
<p>Citations made with <em><a href="https://CRAN.R-project.org/package=RefManageR" class="external-link">RefManageR</a></em>
(McLean, 2017).</p>
<p>
<a id="bib-allaire2023rmarkdown"></a><a href="#cite-allaire2023rmarkdown">[1]</a><cite>
J. Allaire, Y. Xie, C. Dervieux, et al. <em>rmarkdown: Dynamic Documents
for R</em>. R package version 2.23. 2023. URL:
<a href="https://github.com/rstudio/rmarkdown" class="external-link">https://github.com/rstudio/rmarkdown</a>.</cite>
</p>
<p>
<a id="bib-eagles2021speaqeasy"></a><a href="#cite-eagles2021speaqeasy">[2]</a><cite>
N. J. Eagles, E. E. Burke, J. Leonard, et al. “SPEAQeasy: a scalable
pipeline for expression analysis and quantification for
R/bioconductor-powered RNA-seq analyses”. In: <em>BMC
Bioinformatics</em> (2021). DOI:
<a href="https://doi.org/10.1186/s12859-021-04142-3" class="external-link">10.1186/s12859-021-04142-3</a>.
URL:
<a href="https://doi.org/10.1186/s12859-021-04142-3" class="external-link">https://doi.org/10.1186/s12859-021-04142-3</a>.</cite>
</p>
<p>
<a id="bib-gonzalezpadilla2023provides"></a><a href="#cite-gonzalezpadilla2023provides">[3]</a><cite>
D. Gonzalez-Padilla and L. Collado-Torres. <em>Provides access to
smokingMouse project data </em>. <a href="https://github.com/LieberInstitute/smokingMouse/smokingMouse" class="external-link uri">https://github.com/LieberInstitute/smokingMouse/smokingMouse</a>
- R package version 0.99.9. 2023. DOI:
<a href="https://doi.org/10.18129/B9.bioc.smokingMouse" class="external-link">10.18129/B9.bioc.smokingMouse</a>.
URL:
<a href="http://www.bioconductor.org/packages/smokingMouse" class="external-link">http://www.bioconductor.org/packages/smokingMouse</a>.</cite>
</p>
<p>
<a id="bib-mclean2017refmanager"></a><a href="#cite-mclean2017refmanager">[4]</a><cite>
M. W. McLean. “RefManageR: Import and Manage BibTeX and BibLaTeX
References in R”. In: <em>The Journal of Open Source Software</em>
(2017). DOI:
<a href="https://doi.org/10.21105/joss.00338" class="external-link">10.21105/joss.00338</a>.</cite>
</p>
<p>
<a id="bib-ole2023biocstyle"></a><a href="#cite-ole2023biocstyle">[5]</a><cite>
A. Oleś. <em>BiocStyle: Standard styles for vignettes and other
Bioconductor documents</em>. R package version 2.29.1. 2023. DOI:
<a href="https://doi.org/10.18129/B9.bioc.BiocStyle" class="external-link">10.18129/B9.bioc.BiocStyle</a>.
URL:
<a href="https://bioconductor.org/packages/BiocStyle" class="external-link">https://bioconductor.org/packages/BiocStyle</a>.</cite>
</p>
<p>
<a id="bib-xie2023knitr"></a><a href="#cite-xie2023knitr">[6]</a><cite>
Y. Xie. <em>knitr: A General-Purpose Package for Dynamic Report
Generation in R</em>. R package version 1.43. 2023. URL:
<a href="https://yihui.org/knitr/" class="external-link">https://yihui.org/knitr/</a>.</cite>
</p>
</div>
<div class="section level2">
<h2 id="acknowledgements">Acknowledgements<a class="anchor" aria-label="anchor" href="#acknowledgements"></a>
</h2>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Nick J. Eagles, Renee Garcia-Flores, Daianna Gonzalez-Padilla.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
